<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Pipelines for studying the adaptive immune repertoire of T cells 
    and B cells via network analysis based on receptor sequence similarity. 
    Relate clinical outcomes to immune repertoires based on their network 
    properties, or to particular clusters and clones within a repertoire. 
    Yang et al. (2023) &lt;doi:10.3389/fimmu.2023.1181825&gt;.">
<title>Network Analysis of Immune Repertoire • NAIR</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Network Analysis of Immune Repertoire">
<meta property="og:description" content="Pipelines for studying the adaptive immune repertoire of T cells 
    and B cells via network analysis based on receptor sequence similarity. 
    Relate clinical outcomes to immune repertoires based on their network 
    properties, or to particular clusters and clones within a repertoire. 
    Yang et al. (2023) &lt;doi:10.3389/fimmu.2023.1181825&gt;.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">NAIR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.3.9002</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="articles/NAIR.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="articles/buildRepSeqNetwork.html">buildRepSeqNetwork()/buildNet()</a>
    <a class="dropdown-item" href="articles/node_properties.html">Node-Level Network Properties</a>
    <a class="dropdown-item" href="articles/cluster_analysis.html">Cluster Analysis</a>
    <a class="dropdown-item" href="articles/network_visualization.html">Network Visualization</a>
    <a class="dropdown-item" href="articles/supplementary.html">Supplementary Functions</a>
    <a class="dropdown-item" href="articles/dual_chain.html">Dual-Chain Network Analysis</a>
    <a class="dropdown-item" href="articles/associated_clusters.html">Searching for Associated TCR/BCR Clusters</a>
    <a class="dropdown-item" href="articles/public_clusters.html">Searching for Public TCR/BCR Clusters</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/mlizhangx/Network-Analysis-for-Repertoire-Sequencing-/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="nair-network-analysis-of-immune-repertoire">NAIR: Network Analysis of Immune Repertoire<a class="anchor" aria-label="anchor" href="#nair-network-analysis-of-immune-repertoire"></a>
</h1></div>
<p><code>NAIR</code> is an R package for analyzing the adaptive immune repertoire using network analysis based on similarities among receptor sequences. It implements methods from the following paper:</p>
<p><a href="https://www.frontiersin.org/articles/10.3389/fimmu.2023.1181825/full" class="external-link">Hai Yang, Jason Cham, Brian Neal, Zenghua Fan, Tao He and Li Zhang. (2023). NAIR: Network Analysis of Immune Repertoire. <em>Frontiers in Immunology</em>, vol. 14. https://doi.org/10.3389/fimmu.2023.1181825</a></p>
<p><code>NAIR</code> allows users to perform network analysis on Adaptive Immune Receptor Repertoire Sequencing (AIRR-Seq) data, including computing local and global network properties of nodes and clusters, which can provide insights into the structural organization of the immune repertoire network.</p>
<p><code>NAIR</code> also enables users to search across multiple AIRR-Seq samples for clones/clusters associated with subject characteristics, disease conditions or clinical outcomes, as well as identify public clones/clusters. This can help researchers identify potentially important TCR/BCR clones.</p>
<p>To aid in interpretation of the immune repertoire network, <code>NAIR</code> includes convenient functionality for generating customized network visualizations.</p>
<div class="section level4">
<h4 id="what-data-does-nair-support">What data does <code>NAIR</code> support?<a class="anchor" aria-label="anchor" href="#what-data-does-nair-support"></a>
</h4>
<p><code>NAIR</code> supports bulk and single-cell immune repertoire sequence data for T-cell or B-cell receptors (TCR or BCR).</p>
<ul>
<li>
<strong>Single-cell data:</strong> Each row is a single cell</li>
<li>
<strong>Bulk data:</strong> Each row is a distinct TCR/BCR clone (unique combination of V-D-J genes and nucleotide sequence) and typically includes a corresponding measurement of clonal abundance (e.g., clone count and clone frequency/fraction)</li>
</ul>
</div>
<div class="section level4">
<h4 id="how-does-nair-model-the-immune-repertoire-as-a-network">How does <code>NAIR</code> model the immune repertoire as a network?<a class="anchor" aria-label="anchor" href="#how-does-nair-model-the-immune-repertoire-as-a-network"></a>
</h4>
<ul>
<li>Each cell (single-cell data) or clone (bulk data) is modeled as a node (vertex) in the network</li>
<li>For each node, we consider the corresponding receptor sequence (nucleotide or amino acid)</li>
<li>For each pair of nodes, we measure the similarity in their receptor sequences (using the Hamming or Levenshtein distance)</li>
<li>An edge is drawn between two nodes if the distance is below a specified threshold
<ul>
<li>For single-cell data, sequences from two chains (e.g., alpha chain and beta chain) can be jointly used to determine similarity between cells, considering cells as similar when the sequences for both chains are similar (i.e., when the distance for each chain is below the threshold)</li>
</ul>
</li>
<li>Clustering analysis is used to partition the network graph into clusters (densely-connected subgraphs)
<ul>
<li>Many clustering algorithms are available, with each seeking to identify the “best” configuration of clusters according to different graph criteria</li>
</ul>
</li>
<li>Network statistics characterize the repertoire in terms of the local and global structural properties of its graph</li>
<li>Customized visual plots of the network graph are generated, with nodes colored according to desired metadata (e.g., disease status, sample, cluster, clonal abundance, etc.)</li>
</ul>
</div>
</div>
<div class="section level1">
<h1 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h1>
<p>To install the latest release version of <code>NAIR</code>, use the following command:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"NAIR"</span><span class="op">)</span></span></code></pre></div>
<p>To install the latest development version of <code>NAIR</code> from source (which requires compilation), use the following command:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span></span>
<span>  <span class="st">"mlizhangx/Network-Analysis-for-Repertoire-Sequencing-"</span>,</span>
<span>  dependencies <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>  build_vignettes <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<!-- Installing the development version requires a toolchain compiler. On Windows, this means downloading and installing Rtools. On MacOS, this entails installing XCode Command Line Tools ("XCode CLI") and the correct version of gfortran for your macOS version (instructions [here](https://thecoatlessprofessor.com/programming/cpp/r-compiler-tools-for-rcpp-on-macos/)). -->
</div>
<div class="section level1">
<h1 id="getting-started">Getting Started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h1>
<div class="section level3">
<h3 id="main-function">Main Function<a class="anchor" aria-label="anchor" href="#main-function"></a>
</h3>
<p>General network analysis on AIRR-Seq data is performed using <code><a href="reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code> or its convenient alias <code><a href="reference/buildRepSeqNetwork.html">buildNet()</a></code>. This function does the following:</p>
<ul>
<li>Filters the AIRR-Seq data according to user specifications</li>
<li>Builds the network graph for the immune repertoire</li>
<li>Performs additional network analysis, which can include:
<ul>
<li>Cluster analysis</li>
<li>Network properties</li>
<li>Customizable visual plots of the network graph</li>
</ul>
</li>
<li>Returns (and optionally saves) the following output:
<ul>
<li>The network graph (as <code>igraph</code> and adjacency matrix)</li>
<li>Metadata for the network</li>
<li>Metadata for the nodes in the network</li>
<li>Metadata for the clusters in the network</li>
<li>Plots of the network graph</li>
</ul>
</li>
</ul>
<p>See <a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/articles/buildRepSeqNetwork.html">this vignette</a> for a tutorial.</p>
</div>
<div class="section level3">
<h3 id="searching-for-associated-clusters">Searching for Associated Clusters<a class="anchor" aria-label="anchor" href="#searching-for-associated-clusters"></a>
</h3>
<p>Given multiple samples of bulk AIRR-Seq data, <code>NAIR</code> can be used to search for TCR/BCR clusters associated with a binary variable of interest, such as a disease condition, treatment or clinical outcome. See <a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/articles/associated_clusters.html">this article</a> for a tutorial.</p>
</div>
<div class="section level3">
<h3 id="searching-for-public-clusters">Searching for Public Clusters<a class="anchor" aria-label="anchor" href="#searching-for-public-clusters"></a>
</h3>
<p>The <code>NAIR</code> package includes a set of functions that facilitate searching for public TCR/BCR clusters across multiple samples of bulk AIRR-seq data. In this context, a public cluster consists of similar TCR/BCR clones (e.g., those whose CDR3 amino acid sequences differ by at most one amino acid) that are shared across samples (e.g., across individuals or across time points for a single individual). See <a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/articles/public_clusters.html">this article</a> for a tutorial.</p>
</div>
</div>
<div class="section level1">
<h1 id="additional-resources">Additional Resources<a class="anchor" aria-label="anchor" href="#additional-resources"></a>
</h1>
<div class="section level3">
<h3 id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a>
</h3>
<p><a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/articles/network_visualization.html">This article</a> provides an introduction to the creation and customization of network visualizations using <code>NAIR</code>.</p>
</div>
<div class="section level3">
<h3 id="network-properties-and-cluster-analysis">Network Properties and Cluster Analysis<a class="anchor" aria-label="anchor" href="#network-properties-and-cluster-analysis"></a>
</h3>
<p><a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/articles/node_properties.html">This vignette</a> provides an introduction to computing node-level network properties with <code>NAIR</code>.</p>
<p><a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/articles/cluster_analysis.html">This vignette</a> explains how to perform cluster analysis with <code>NAIR</code>.</p>
</div>
<div class="section level3">
<h3 id="supplementary-functions">Supplementary Functions<a class="anchor" aria-label="anchor" href="#supplementary-functions"></a>
</h3>
<p><a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/articles/supplementary.html">This vignette</a> provides an overview of <code>NAIR</code> utility functions that supplement the main function <code><a href="reference/buildRepSeqNetwork.html">buildNet()</a></code>.</p>
</div>
</div>

  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=NAIR" class="external-link">View on CRAN</a></li>
<li><a href="https://github.com/mlizhangx/Network-Analysis-for-Repertoire-Sequencing-/" class="external-link">Browse source code</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small>GPL (&gt;= 3)</small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing NAIR</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Brian Neal <br><small class="roles"> Author, maintainer </small>  </li>
<li>Hai Yang <br><small class="roles"> Author </small>  </li>
<li>Daniil Matveev <br><small class="roles"> Author </small>  </li>
<li>Phi Long Le <br><small class="roles"> Author </small>  </li>
<li>Li Zhang <br><small class="roles"> Copyright holder, author </small>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://lifecycle.r-lib.org/articles/stages.html#experimental" class="external-link"><img src="https://img.shields.io/badge/lifecycle-experimental-orange.svg" alt="Lifecycle: experimental"></a></li>
<li><a href="https://github.com/mlizhangx/Network-Analysis-for-Repertoire-Sequencing-/actions/workflows/check-standard.yaml" class="external-link"><img src="https://github.com/mlizhangx/Network-Analysis-for-Repertoire-Sequencing-/actions/workflows/check-standard.yaml/badge.svg" alt="R-CMD-check"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Brian Neal, Hai Yang, Daniil Matveev, Phi Long Le, Li Zhang.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
