<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Network Analysis of Immune Repertoire — buildRepSeqNetwork • NAIR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Network Analysis of Immune Repertoire — buildRepSeqNetwork"><meta name="description" content="Given Adaptive Immune Receptor Repertoire Sequencing (AIRR-Seq) data, builds
    the network graph for the immune repertoire based on sequence similarity,
    computes specified network properties and generates customized visualizations.
buildNet() is identical to buildRepSeqNetwork(), existing as
    an alias for convenience."><meta property="og:description" content="Given Adaptive Immune Receptor Repertoire Sequencing (AIRR-Seq) data, builds
    the network graph for the immune repertoire based on sequence similarity,
    computes specified network properties and generates customized visualizations.
buildNet() is identical to buildRepSeqNetwork(), existing as
    an alias for convenience."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">NAIR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.4.9001</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/NAIR.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/buildRepSeqNetwork.html">buildRepSeqNetwork()/buildNet()</a></li>
    <li><a class="dropdown-item" href="../articles/node_properties.html">Node-Level Network Properties</a></li>
    <li><a class="dropdown-item" href="../articles/cluster_analysis.html">Cluster Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/network_visualization.html">Network Visualization</a></li>
    <li><a class="dropdown-item" href="../articles/supplementary.html">Supplementary Functions</a></li>
    <li><a class="dropdown-item" href="../articles/dual_chain.html">Dual-Chain Network Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/associated_clusters.html">Searching for Associated TCR/BCR Clusters</a></li>
    <li><a class="dropdown-item" href="../articles/public_clusters.html">Searching for Public TCR/BCR Clusters</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/mlizhangx/Network-Analysis-for-Repertoire-Sequencing-/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Network Analysis of Immune Repertoire</h1>

      <div class="d-none name"><code>buildRepSeqNetwork.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Given Adaptive Immune Receptor Repertoire Sequencing (AIRR-Seq) data, builds
    the network graph for the immune repertoire based on sequence similarity,
    computes specified network properties and generates customized visualizations.</p>
<p><code>buildNet()</code> is identical to <code>buildRepSeqNetwork()</code>, existing as
    an alias for convenience.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">buildRepSeqNetwork</span><span class="op">(</span></span>
<span></span>
<span>  <span class="co">## Input ##</span></span>
<span>  <span class="va">data</span>,</span>
<span>  <span class="va">seq_col</span>,</span>
<span>  count_col <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  subset_cols <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  min_seq_length <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  drop_matches <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span></span>
<span>  <span class="co">## Network ##</span></span>
<span>  dist_type <span class="op">=</span> <span class="st">"hamming"</span>,</span>
<span>  dist_cutoff <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  drop_isolated_nodes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  net_build_method <span class="op">=</span> <span class="st">"default"</span>,</span>
<span>  node_stats <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  stats_to_include <span class="op">=</span> <span class="fu"><a href="chooseNodeStats.html">chooseNodeStats</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  cluster_stats <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  cluster_fun <span class="op">=</span> <span class="st">"fast_greedy"</span>,</span>
<span>  cluster_id_name <span class="op">=</span> <span class="st">"cluster_id"</span>,</span>
<span></span>
<span>  <span class="co">## Visualization ##</span></span>
<span>  plots <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  print_plots <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  plot_title <span class="op">=</span> <span class="st">"auto"</span>,</span>
<span>  plot_subtitle <span class="op">=</span> <span class="st">"auto"</span>,</span>
<span>  color_nodes_by <span class="op">=</span> <span class="st">"auto"</span>,</span>
<span>  <span class="va">...</span>,</span>
<span></span>
<span>  <span class="co">## Output ##</span></span>
<span>  output_dir <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  output_type <span class="op">=</span> <span class="st">"rds"</span>,</span>
<span>  output_name <span class="op">=</span> <span class="st">"MyRepSeqNetwork"</span>,</span>
<span>  pdf_width <span class="op">=</span> <span class="fl">12</span>,</span>
<span>  pdf_height <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Alias for buildRepSeqNetwork()</span></span>
<span><span class="fu">buildNet</span><span class="op">(</span></span>
<span>  <span class="va">data</span>,</span>
<span>  <span class="va">seq_col</span>,</span>
<span>  count_col <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  subset_cols <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  min_seq_length <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  drop_matches <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  dist_type <span class="op">=</span> <span class="st">"hamming"</span>,</span>
<span>  dist_cutoff <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  drop_isolated_nodes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  net_build_method <span class="op">=</span> <span class="st">"default"</span>,</span>
<span>  node_stats <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  stats_to_include <span class="op">=</span> <span class="fu"><a href="chooseNodeStats.html">chooseNodeStats</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  cluster_stats <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  cluster_fun <span class="op">=</span> <span class="st">"fast_greedy"</span>,</span>
<span>  cluster_id_name <span class="op">=</span> <span class="st">"cluster_id"</span>,</span>
<span>  plots <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  print_plots <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  plot_title <span class="op">=</span> <span class="st">"auto"</span>,</span>
<span>  plot_subtitle <span class="op">=</span> <span class="st">"auto"</span>,</span>
<span>  color_nodes_by <span class="op">=</span> <span class="st">"auto"</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  output_dir <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  output_type <span class="op">=</span> <span class="st">"rds"</span>,</span>
<span>  output_name <span class="op">=</span> <span class="st">"MyRepSeqNetwork"</span>,</span>
<span>  pdf_width <span class="op">=</span> <span class="fl">12</span>,</span>
<span>  pdf_height <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <p></p>
<dl><dt id="arg-data">data<a class="anchor" aria-label="anchor" href="#arg-data"></a></dt>
<dd><p>A data frame containing the AIRR-Seq data, with variables indexed by column
    and observations (e.g., clones or cells) indexed by row.</p></dd>

  <dt id="arg-seq-col">seq_col<a class="anchor" aria-label="anchor" href="#arg-seq-col"></a></dt>
<dd><p>Specifies the column(s) of <code>data</code> containing
    the receptor sequences to be used as the basis of similarity between rows.
    Accepts a character string containing the column name
      or a numeric scalar containing the column index.
    Also accepts a vector of length 2 specifying distinct sequence columns
    (e.g., alpha chain and beta chain), in which case
    similarity between rows depends on similarity in both sequence columns
    (see details).</p></dd>

  <dt id="arg-count-col">count_col<a class="anchor" aria-label="anchor" href="#arg-count-col"></a></dt>
<dd><p>Optional. Specifies the column of <code>data</code> containing a measure of abundance,
    e.g., clone count or unique molecular identifier (UMI) count. Accepts either
    the column name or column index. If provided, data rows with <code>NA</code> count
    values will be removed prior to network building. Count values are used to
    compute certain cluster properties when <code>cluster_stats = TRUE</code>; see
    <code><a href="addClusterStats.html">addClusterStats()</a></code> for details.</p></dd>

  <dt id="arg-subset-cols">subset_cols<a class="anchor" aria-label="anchor" href="#arg-subset-cols"></a></dt>
<dd><p>Specifies which columns of the AIRR-Seq data are included in the output.
    Accepts a vector of column names or a vector of column indices. The default
    <code>NULL</code> includes all columns. The receptor sequence column is always
    included regardless of this argument's value.
    Passed to <code><a href="filterInputData.html">filterInputData()</a></code>.</p></dd>

  <dt id="arg-min-seq-length">min_seq_length<a class="anchor" aria-label="anchor" href="#arg-min-seq-length"></a></dt>
<dd><p>A numeric scalar, or <code>NULL</code>. Observations whose receptor sequences have
    fewer than <code>min_seq_length</code> characters are removed prior to network analysis.</p></dd>

  <dt id="arg-drop-matches">drop_matches<a class="anchor" aria-label="anchor" href="#arg-drop-matches"></a></dt>
<dd><p>Optional. Passed to <code><a href="filterInputData.html">filterInputData()</a></code>.
    Accepts a character string
    containing a regular expression (see <code><a href="https://rdrr.io/r/base/regex.html" class="external-link">regex</a></code>).
    Checks receptor sequences for a pattern match using <code><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep()</a></code>.
    Those returning a match are removed prior to network analysis.</p></dd>

  <dt id="arg-dist-type">dist_type<a class="anchor" aria-label="anchor" href="#arg-dist-type"></a></dt>
<dd><p>Specifies the function used to quantify the similarity between sequences.
    The similarity between two sequences determines the pairwise distance between
    their respective nodes in the network graph, with greater similarity corresponding
    to shorter distance. Valid options are <code>"hamming"</code> (the default), which
    uses <code><a href="hamDistBounded.html">hamDistBounded()</a></code>,
    and <code>"levenshtein"</code>, which uses
    <code><a href="levDistBounded.html">levDistBounded()</a></code>.</p></dd>

  <dt id="arg-dist-cutoff">dist_cutoff<a class="anchor" aria-label="anchor" href="#arg-dist-cutoff"></a></dt>
<dd><p>A nonnegative scalar. Specifies the maximum pairwise distance (based on
    <code>dist_type</code>) for an edge connection to exist between two nodes. Pairs of
    nodes whose distance is less than or equal to this value will be joined by an
    edge connection in the network graph. Controls the stringency of the network
    construction and affects the number and density of edges in the network. A lower
    cutoff value requires greater similarity between sequences in order for their
    respective nodes to be joined by an edge connection. A value of <code>0</code>
    requires two sequences to be identical in order for their nodes to be joined
    by an edge.</p></dd>

  <dt id="arg-drop-isolated-nodes">drop_isolated_nodes<a class="anchor" aria-label="anchor" href="#arg-drop-isolated-nodes"></a></dt>
<dd><p>A logical scalar. When <code>TRUE</code>, removes each node that is not joined by an
    edge connection to any other node in the network graph.</p></dd>

  <dt id="arg-net-build-method">net_build_method<a class="anchor" aria-label="anchor" href="#arg-net-build-method"></a></dt>
<dd><p>Passed to the <code>method</code> argument of <code><a href="generateAdjacencyMatrix.html">generateAdjacencyMatrix()</a></code>. Specifies the algorithm used to compute the network adjacency matrix.</p></dd>

  <dt id="arg-node-stats">node_stats<a class="anchor" aria-label="anchor" href="#arg-node-stats"></a></dt>
<dd><p>A logical scalar. Specifies whether node-level network properties are computed.</p></dd>

  <dt id="arg-stats-to-include">stats_to_include<a class="anchor" aria-label="anchor" href="#arg-stats-to-include"></a></dt>
<dd><p>A named logical vector returned by
    <code><a href="chooseNodeStats.html">chooseNodeStats()</a></code> or
    <code><a href="chooseNodeStats.html">exclusiveNodeStats()</a></code>.
    Specifies the node-level network properties
    to compute. Also accepts the value <code>"all"</code>.
    Only relevant if <code>node_stats = TRUE</code>.</p></dd>

  <dt id="arg-cluster-stats">cluster_stats<a class="anchor" aria-label="anchor" href="#arg-cluster-stats"></a></dt>
<dd><p>A logical scalar. Specifies whether to compute cluster-level network properties.</p></dd>

  <dt id="arg-cluster-fun">cluster_fun<a class="anchor" aria-label="anchor" href="#arg-cluster-fun"></a></dt>
<dd><p>Passed to <code><a href="addClusterMembership.html">addClusterMembership()</a></code>.
    Specifies the clustering algorithm
    used when cluster analysis is performed. Cluster analysis is performed when
    <code>cluster_stats = TRUE</code> or when <code>node_stats = TRUE</code> with the
    <code>cluster_id</code> property enabled via the <code>stats_to_include</code> argument.</p></dd>

  <dt id="arg-cluster-id-name">cluster_id_name<a class="anchor" aria-label="anchor" href="#arg-cluster-id-name"></a></dt>
<dd><p>Passed to <code><a href="addClusterMembership.html">addClusterMembership()</a></code>.
    Specifies the name of the cluster membership variable added to the node metadata
    when cluster analysis is performed (see <code>cluster_fun</code>).</p></dd>

  <dt id="arg-plots">plots<a class="anchor" aria-label="anchor" href="#arg-plots"></a></dt>
<dd><p>A logical scalar. Specifies whether to generate plots of the network graph.</p></dd>

  <dt id="arg-print-plots">print_plots<a class="anchor" aria-label="anchor" href="#arg-print-plots"></a></dt>
<dd><p>A logical scalar. If <code>plots = TRUE</code>, specifies whether the plots should
    be printed to the R plotting window.</p></dd>

  <dt id="arg-plot-title">plot_title<a class="anchor" aria-label="anchor" href="#arg-plot-title"></a></dt>
<dd><p>A character string or <code>NULL</code>. If <code>plots = TRUE</code>, this is the title
    used for each plot. The default value <code>"auto"</code> generates the title based
    on the value of the <code>output_name</code> argument.</p></dd>

  <dt id="arg-plot-subtitle">plot_subtitle<a class="anchor" aria-label="anchor" href="#arg-plot-subtitle"></a></dt>
<dd><p>A character string or <code>NULL</code>. If <code>plots = TRUE</code>, this is the subtitle
    used for each plot. The default value <code>"auto"</code> generates a subtitle based
    on the values of the <code>dist_type</code> and <code>dist_cutoff</code> arguments.</p></dd>

  <dt id="arg-color-nodes-by">color_nodes_by<a class="anchor" aria-label="anchor" href="#arg-color-nodes-by"></a></dt>
<dd><p>Optional. Specifies a variable to be used as metadata for coloring the nodes
    in the network graph plot. Accepts a character string. This can be a column
    name of <code>data</code> or (if <code>node_stats = TRUE</code>) the name of a computed
    node-level network property (based on <code>stats_to_include</code>). Also accepts
    a character vector specifying multiple variables, in which case one plot will
    be generated for each variable. The default value <code>"auto"</code> attempts to use
    one of several potential variables to color the nodes, depending on what is
    available. A value of <code>NULL</code> leaves the nodes uncolored.</p></dd>

  <dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Other named arguments to <code><a href="addPlots.html">addPlots()</a></code>.</p></dd>

  <dt id="arg-output-dir">output_dir<a class="anchor" aria-label="anchor" href="#arg-output-dir"></a></dt>
<dd><p>A file path specifying the directory for saving the output. The directory will
  be created if it does not exist. If <code>NULL</code>, output will be returned but
  not saved.</p></dd>

  <dt id="arg-output-type">output_type<a class="anchor" aria-label="anchor" href="#arg-output-type"></a></dt>
<dd><p>A character string specifying the file format to use when saving the output.
  The default value <code>"individual"</code> saves each element of the returned list
  as an individual uncompressed file, with data frames saved in csv format. For
  better compression, the values <code>"rda"</code> and <code>"rds"</code> save the returned
  list as a single file using the rda and rds format, respectively (in the former
  case, the list will be named <code>net</code> within the rda file). Regardless of the
  argument value, any plots generated will saved to a pdf file containing one plot
  per page.</p></dd>

  <dt id="arg-output-name">output_name<a class="anchor" aria-label="anchor" href="#arg-output-name"></a></dt>
<dd><p>A character string. All files saved will have file names beginning with this value.</p></dd>

  <dt id="arg-pdf-width">pdf_width<a class="anchor" aria-label="anchor" href="#arg-pdf-width"></a></dt>
<dd><p>Sets the width of each plot when writing to pdf.
    Passed to <code><a href="saveNetwork.html">saveNetwork()</a></code>.</p></dd>

  <dt id="arg-pdf-height">pdf_height<a class="anchor" aria-label="anchor" href="#arg-pdf-height"></a></dt>
<dd><p>Sets the height of each plot when writing to pdf.
    Passed to <code><a href="saveNetwork.html">saveNetwork()</a></code>.</p></dd>

  <dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Logical. If <code>TRUE</code>, generates messages about the tasks
      performed and their progress, as well as relevant properties of intermediate
      outputs. Messages are sent to <code><a href="https://rdrr.io/r/base/showConnections.html" class="external-link">stderr()</a></code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>To construct the immune repertoire network, each TCR/BCR clone (bulk data)
or cell (single-cell data) is modeled as a node in the network graph,
corresponding to a single row of the AIRR-Seq data. For each node, the
corresponding receptor sequence is considered. Both nucleotide and amino acid
sequences are supported for this purpose. The receptor sequence is used as the
basis of similarity and distance between nodes in the network.</p>
<p>Similarity between sequences is measured using either the Hamming distance
or Levenshtein (edit) distance. The similarity determines the pairwise distance
between nodes in the network graph. The more similar two sequences are, the
shorter the distance between their respective nodes. Two nodes in the graph
are joined by an edge if the distance between them is sufficiently small,
i.e., if their receptor sequences are sufficiently similar.</p>
<p>For single-cell data, edge connections between nodes can be based on similarity
in both the alpha chain and beta chain sequences.
This is done by providing a vector of length 2 to <code>seq_cols</code>
specifying the two sequence columns in <code>data</code>.
The distance between two nodes is then the greater of the two distances between
sequences in corresponding chains.
Two nodes will be joined by an edge if their alpha chain sequences are sufficiently
similar and their beta chain sequences are sufficiently similar.</p>
<p>See the
<a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/articles/buildRepSeqNetwork.html">buildRepSeqNetwork package vignette</a>
for more details. The vignette can be accessed offline using
<code><a href="../articles/buildRepSeqNetwork.html">vignette("buildRepSeqNetwork")</a></code>.</p>
    </div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>If the constructed network contains no nodes, the function will return
    <code>NULL</code>, invisibly, with a warning. Otherwise, the function invisibly
    returns a list containing the following items:</p>
<p></p>
<dl><dt>details</dt>
<dd><p>A list containing information about the network and the
    settings used during its construction.</p></dd>

    <dt>igraph</dt>
<dd><p>An object of class <code><a href="https://r.igraph.org/reference/aaa-igraph-package.html" class="external-link">igraph</a></code> containing the
    list of nodes and edges for the network graph.</p></dd>

    <dt>adjacency_matrix</dt>
<dd><p>The network graph adjacency matrix, stored as a sparse
    matrix of class <code>dgCMatrix</code> from the <code>Matrix</code> package.
    See <code><a href="https://rdrr.io/pkg/Matrix/man/dgCMatrix-class.html" class="external-link">dgCMatrix-class</a></code>.</p></dd>

    <dt>node_data</dt>
<dd><p>A data frame containing containing metadata for the network
    nodes, where each row corresponds to a node in the network graph. This data
    frame contains all variables from <code>data</code> (unless otherwise specified via
    <code>subset_cols</code>) in addition to the computed node-level network properties
    if <code>node_stats = TRUE</code>. Each row's name is the name of the corresponding
    row from <code>data</code>.</p></dd>

    <dt>cluster_data</dt>
<dd><p>A data frame containing network properties for the clusters,
    where each row corresponds to a cluster in the network graph. Only included if
    <code>cluster_stats = TRUE</code>.</p></dd>

    <dt>plots</dt>
<dd><p>A list containing one element for each plot generated
    as well as an additional element for the matrix that specifies the graph layout.
    Each plot is an object of class <code><a href="https://ggraph.data-imaginist.com/reference/ggraph.html" class="external-link">ggraph</a></code>. Only included
    if <code>plots = TRUE</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Hai Yang, Jason Cham, Brian Neal, Zenghua Fan, Tao He and Li Zhang. (2023).
NAIR: Network Analysis of Immune Repertoire. <em>Frontiers in Immunology</em>, vol. 14.
<a href="https://www.frontiersin.org/articles/10.3389/fimmu.2023.1181825/full" class="external-link">doi: 10.3389/fimmu.2023.1181825</a></p>
<p><a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/index.html">Webpage for the NAIR package</a></p>
<p><a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/articles/buildRepSeqNetwork.html">buildRepSeqNetwork vignette</a></p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Brian Neal (<a href="mailto:Brian.Neal@ucsf.edu">Brian.Neal@ucsf.edu</a>)</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">42</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">toy_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="simulateToyData.html">simulateToyData</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Simple call</span></span></span>
<span class="r-in"><span><span class="va">network</span> <span class="op">=</span> <span class="fu">buildNet</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">toy_data</span>,</span></span>
<span class="r-in"><span>  seq_col <span class="op">=</span> <span class="st">"CloneSeq"</span>,</span></span>
<span class="r-in"><span>  print_plots <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="buildRepSeqNetwork-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Customized:</span></span></span>
<span class="r-in"><span><span class="va">network</span> <span class="op">&lt;-</span> <span class="fu">buildNet</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">toy_data</span>, <span class="st">"CloneSeq"</span>,</span></span>
<span class="r-in"><span>  dist_type <span class="op">=</span> <span class="st">"levenshtein"</span>,</span></span>
<span class="r-in"><span>  node_stats <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>  cluster_stats <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>  cluster_fun <span class="op">=</span> <span class="st">"louvain"</span>,</span></span>
<span class="r-in"><span>  cluster_id_name <span class="op">=</span> <span class="st">"cluster_membership"</span>,</span></span>
<span class="r-in"><span>  count_col <span class="op">=</span> <span class="st">"CloneCount"</span>,</span></span>
<span class="r-in"><span>  color_nodes_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SampleID"</span>, <span class="st">"cluster_membership"</span>, <span class="st">"coreness"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  color_scheme <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"default"</span>, <span class="st">"Viridis"</span>, <span class="st">"plasma-1"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  size_nodes_by <span class="op">=</span> <span class="st">"degree"</span>,</span></span>
<span class="r-in"><span>  node_size_limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">1.5</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  plot_title <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>  plot_subtitle <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>  print_plots <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Input data contains 200 rows.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Removing sequences with length fewer than 3 characters...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  Done. 200 rows remaining.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Computing network edges based on a max levenshtein distance of 1...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  Done.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Network contains 124 nodes (after removing isolated nodes).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Network objects and node metadata stored in a list</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Obtaining cluster properties...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Partitioning the network graph into clusters...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  Done.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cluster membership variable ‘cluster_membership’ added to node metadata.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Data frame ‘cluster_data’ added to list of network objects.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Generating graph plot with nodes colored by SampleID...</span>
<span class="r-plt img"><img src="buildRepSeqNetwork-2.png" alt="" width="700" height="433"></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  Done.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Generating graph plot with nodes colored by cluster_membership...</span>
<span class="r-plt img"><img src="buildRepSeqNetwork-3.png" alt="" width="700" height="433"></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  Done.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Generating graph plot with nodes colored by coreness...</span>
<span class="r-plt img"><img src="buildRepSeqNetwork-4.png" alt="" width="700" height="433"></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  Done.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> New plots added to ‘net$plots’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ‘output_dir’ is “NULL”. Output will not be saved.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/typeof.html" class="external-link">typeof</a></span><span class="op">(</span><span class="va">network</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "list"</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">network</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "details"          "igraph"           "adjacency_matrix" "node_data"       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [5] "cluster_data"     "plots"           </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">network</span><span class="op">$</span><span class="va">details</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $seq_col</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "CloneSeq"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $dist_type</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "levenshtein"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $dist_cutoff</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $drop_isolated_nodes</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $nodes_in_network</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 124</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $clusters_in_network</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> louvain </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      19 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $cluster_id_variable</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              louvain </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> "cluster_membership" </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $cluster_data_goes_with</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "cluster_membership"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $count_col_for_cluster_data</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "CloneCount"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $min_seq_length</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> $drop_matches</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "NULL"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">network</span><span class="op">$</span><span class="va">node_data</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         CloneSeq CloneFrequency CloneCount SampleID degree transitivity</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2  GGAGATGAATCGG    0.007777102       3057  Sample1      1          NaN</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5  GAAAGAGAATTCG    0.009336593       3670  Sample1      3    0.3333333</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8  GGGGAGAAATTGG    0.006220155       2445  Sample1      2    1.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11 GGGGGAGAATTGC    0.012969469       5098  Sample1      4    0.6666667</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 12 GGGGGGGAATTGC    0.009079646       3569  Sample1     10    0.3555556</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 13 AGGGGGAAATTGG    0.014941093       5873  Sample1      5    0.1000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    eigen_centrality centrality_by_eigen betweenness centrality_by_betweenness</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2        0.00000000          0.00000000    0.000000                  0.000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5        0.00000000          0.00000000   48.619048                 48.619048</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8        0.04558284          0.04558284    0.000000                  0.000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11       0.15054597          0.15054597    2.156593                  2.156593</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 12       0.52695160          0.52695160  111.034076                111.034076</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 13       0.14681104          0.14681104   77.439560                 77.439560</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    authority_score coreness   page_rank cluster_membership</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2     1.210694e-18        1 0.008064516                  1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5     6.945373e-18        2 0.009136977                  2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8     4.558284e-02        2 0.003869458                  3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 11    1.505460e-01        4 0.004979346                  3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 12    5.269516e-01        6 0.011562784                  3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 13    1.468110e-01        3 0.008554352                  3</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">network</span><span class="op">$</span><span class="va">cluster_data</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   cluster_id node_count eigen_centrality_eigenvalue eigen_centrality_index</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1          1          2                    1.000000                    NaN</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2          2          7                    2.342923              0.5678500</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3          3         33                   12.159339              0.6082397</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4          4          2                    1.000000                    NaN</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5          5          5                    3.086130              0.3046233</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6          6          3                    1.414214              0.5857864</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   closeness_centrality_index degree_centrality_index edge_density</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1                        NaN               0.0000000    1.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2                  0.3650918               0.1666667    0.3333333</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3                  0.4521644               0.3238636    0.2386364</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4                        NaN               0.0000000    1.0000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5                  0.6000000               0.3000000    0.7000000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6                  1.0000000               0.3333333    0.6666667</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   global_transitivity assortativity diameter_length max_degree mean_degree</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1                 NaN           NaN               2          1        1.00</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2           0.3333333   -0.16666667               6          3        2.29</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3           0.5759124    0.01763639               7         18        7.79</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4                 NaN           NaN               2          1        1.00</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5           0.8000000   -0.55555556               3          7        3.60</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6           0.0000000   -1.00000000               3          2        1.33</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   mean_seq_length seq_w_max_degree max_count agg_count seq_w_max_count</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1           13.00    GGAGATGAATCGG      5242      8299   GGAGCTGAATCGG</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2           12.86    GAAAGAGAATTCG      4422     23255   GAAATAGAATTCG</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3           12.97    GGGGGGGAATTGG      6526    136306   GGGGGGGAATTGG</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4           13.00    GGTTAGGAATTCG      4553      7630   GGTTAGGAATTCG</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5           13.00    AGGGGAGAATTGG      4422     17971   AGCGGAGAATTGG</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6           13.00    GGAAAGAAATTGG      4990     13219   GGAAAAAAATTGG</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Brian Neal, Hai Yang, Daniil Matveev, Phi Long Le, Li Zhang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

