<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Given Adaptive Immune Receptor Repertoire Sequencing (AIRR-Seq) data, builds the network graph for the immune repertoire based on sequence similarity, computes specified network properties and generates customized visualizations."><title>Network Analysis of Immune Repertoire — buildRepSeqNetwork • NAIR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Network Analysis of Immune Repertoire — buildRepSeqNetwork"><meta property="og:description" content="Given Adaptive Immune Receptor Repertoire Sequencing (AIRR-Seq) data, builds the network graph for the immune repertoire based on sequence similarity, computes specified network properties and generates customized visualizations."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">NAIR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.9043</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/NAIR.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/associated_clusters.html">Searching for Associated TCR/BCR Clusters</a>
    <a class="dropdown-item" href="../articles/buildRepSeqNetwork.html">buildRepSeqNetwork</a>
    <a class="dropdown-item" href="../articles/dual_chain.html">Dual-Chain Network Analysis</a>
    <a class="dropdown-item" href="../articles/network_visualization.html">Network Visualization</a>
    <a class="dropdown-item" href="../articles/public_clusters.html">Searching for Public TCR/BCR Clusters</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/mlizhangx/Network-Analysis-for-Repertoire-Sequencing-/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Network Analysis of Immune Repertoire</h1>
      
      <div class="d-none name"><code>buildRepSeqNetwork.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Given Adaptive Immune Receptor Repertoire Sequencing (AIRR-Seq) data, builds the network graph for the immune repertoire based on sequence similarity, computes specified network properties and generates customized visualizations.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">buildRepSeqNetwork</span><span class="op">(</span></span>
<span></span>
<span>  <span class="co">## Input ##</span></span>
<span>  <span class="va">data</span>,</span>
<span>  <span class="va">seq_col</span>,</span>
<span>  count_col <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  subset_cols <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  min_seq_length <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  drop_matches <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span></span>
<span>  <span class="co">## Network ##</span></span>
<span>  dist_type <span class="op">=</span> <span class="st">"hamming"</span>,</span>
<span>  dist_cutoff <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  drop_isolated_nodes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  node_stats <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  stats_to_include <span class="op">=</span> <span class="fu"><a href="chooseNodeStats.html">chooseNodeStats</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  cluster_stats <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  cluster_fun <span class="op">=</span> <span class="va">cluster_fast_greedy</span>,</span>
<span></span>
<span>  <span class="co">## Visualization ##</span></span>
<span>  plots <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  print_plots <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  plot_title <span class="op">=</span> <span class="st">"auto"</span>,</span>
<span>  plot_subtitle <span class="op">=</span> <span class="st">"auto"</span>,</span>
<span>  color_nodes_by <span class="op">=</span> <span class="st">"auto"</span>,</span>
<span>  <span class="va">...</span>,</span>
<span></span>
<span>  <span class="co">## Output ##</span></span>
<span>  output_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  output_type <span class="op">=</span> <span class="st">"individual"</span>,</span>
<span>  output_name <span class="op">=</span> <span class="st">"MyRepSeqNetwork"</span>,</span>
<span>  pdf_width <span class="op">=</span> <span class="fl">12</span>,</span>
<span>  pdf_height <span class="op">=</span> <span class="fl">10</span></span>
<span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>data</dt>
<dd><p>A data frame containing the AIRR-Seq data, with variables indexed by column and observations (e.g., clones or cells) indexed by row.</p></dd>

  <dt>seq_col</dt>
<dd><p>Specifies the column of <code>data</code> containing the receptor sequences to be used as the basis of similarity between rows. Accepts either the column name or column index. Alternatively, a vector of length 2 specifying two columns with distinct sequences (e.g., alpha chain and beta chain), in which case similarity between rows depends on both types of sequences being similar (e.g., similar alpha chain sequences and similar beta chain sequences).</p></dd>

  <dt>count_col</dt>
<dd><p>Optional. Specifies the column of <code>data</code> containing a measure of abundance, e.g., clone count or unique molecular identifier (UMI) count. Accepts either the column name or column index. Passed to <code><a href="getClusterStats.html">getClusterStats</a></code>; only relevant if <code>cluster_stats = TRUE</code>.</p></dd>

  <dt>subset_cols</dt>
<dd><p>Specifies which columns of the AIRR-Seq data are included in the output. Accepts a vector of column names or a vector of column indices. The default <code>NULL</code> includes all columns. The receptor sequence column is always included regardless of this argument's value. Passed to <code><a href="filterInputData.html">filterInputData</a></code>.</p></dd>

  <dt>min_seq_length</dt>
<dd><p>A numeric scalar, or <code>NULL</code>. Observations whose receptor sequences have fewer than <code>min_seq_length</code> characters are removed prior to network analysis.</p></dd>

  <dt>drop_matches</dt>
<dd><p>Optional. Passed to <code><a href="filterInputData.html">filterInputData</a></code>. Accepts a regular expression (see <code><a href="https://rdrr.io/r/base/regex.html" class="external-link">regex</a></code>). Checks receptor sequences for a pattern match using <code><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></code>. Those returning a match are removed prior to network analysis.</p></dd>

  <dt>dist_type</dt>
<dd><p>Specifies the function used to quantify the similarity between sequences. The similarity between two sequences determines the pairwise distance between their respective nodes in the network graph, with greater similarity corresponding to shorter distance. Valid options are <code>"hamming"</code> (the default), which uses <code><a href="hamDistBounded.html">hamDistBounded</a></code>, and <code>"levenshtein"</code>, which uses <code><a href="levDistBounded.html">levDistBounded</a></code>. An additional option <code>"euclidean_on_atchley"</code> exists for TCR CDR3 amino acid sequences (see <code><a href="adjacencyMatAtchleyFromSeqs.html">adjacencyMatAtchleyFromSeqs</a></code>).</p></dd>

  <dt>dist_cutoff</dt>
<dd><p>A nonnegative scalar. Specifies the maximum pairwise distance (based on <code>dist_type</code>) for an edge connection to exist between two nodes. Pairs of nodes whose distance is less than or equal to this value will be joined by an edge connection in the network graph. Controls the stringency of the network construction and affects the number and density of edges in the network. A lower cutoff value requires greater similarity between sequences in order for their respective nodes to be joined by an edge connection. A value of <code>0</code> requires two sequences to be identical in order for their nodes to be joined by an edge.</p></dd>

  <dt>drop_isolated_nodes</dt>
<dd><p>A logical scalar. When <code>TRUE</code>, removes each node that is not joined by an edge connection to any other node in the network graph.</p></dd>

  <dt>node_stats</dt>
<dd><p>A logical scalar. Specifies whether node-level network properties are computed.</p></dd>

  <dt>stats_to_include</dt>
<dd><p>A named logical vector returned by <code><a href="chooseNodeStats.html">chooseNodeStats</a></code> or <code><a href="chooseNodeStats.html">exclusiveNodeStats</a></code>. Specifies the node-level network properties to compute. Also accepts the values <code>"all"</code> and <code>"cluster_id_only"</code>. Only relevant if <code>node_stats = TRUE</code>.</p></dd>

  <dt>cluster_stats</dt>
<dd><p>A logical scalar. Specifies whether to compute cluster-level network properties.</p></dd>

  <dt>cluster_fun</dt>
<dd><p>Passed to <code><a href="addClusterMembership.html">addClusterMembership</a></code>. Specifies the clustering algorithm used when cluster analysis is performed. Cluster analysis is performed when <code>cluster_stats = TRUE</code> or when <code>node_stats = TRUE</code> with the <code>cluster_id</code> property enabled via the <code>stats_to_include</code> argument.</p></dd>

  <dt>plots</dt>
<dd><p>A logical scalar. Specifies whether to generate plots of the network graph.</p></dd>

  <dt>print_plots</dt>
<dd><p>A logical scalar. If <code>plots = TRUE</code>, specifies whether the plots should be printed to the R plotting window.</p></dd>

  <dt>plot_title</dt>
<dd><p>A character string or <code>NULL</code>. If <code>plots = TRUE</code>, this is the title used for each plot. The default value <code>"auto"</code> generates the title based on the value of the <code>output_name</code> argument.</p></dd>

  <dt>plot_subtitle</dt>
<dd><p>A character string or <code>NULL</code>. If <code>plots = TRUE</code>, this is the subtitle used for each plot. The default value <code>"auto"</code> generates a subtitle based on the values of the <code>dist_type</code> and <code>dist_cutoff</code> arguments.</p></dd>

  <dt>color_nodes_by</dt>
<dd><p>Optional. Specifies a variable to be used as metadata for coloring the nodes in the network graph plot. Accepts a character string. This can be a column name of <code>data</code> or (if <code>node_stats = TRUE</code>) the name of a computed node-level network property (based on <code>stats_to_include</code>). Also accepts a character vector specifying multiple variables, in which case one plot will be generated for each variable. The default value <code>"auto"</code> attempts to use one of several potential variables to color the nodes, depending on what is available.</p></dd>

  <dt>...</dt>
<dd><p>Other arguments to <code><a href="generateNetworkGraphPlots.html">generateNetworkGraphPlots</a></code>.</p></dd>

  <dt>output_dir</dt>
<dd><p>A file path specifying the directory for saving the output. The directory will be created if it does not exist. If <code>NULL</code>, output will be returned but not saved.</p></dd>

  <dt>output_type</dt>
<dd><p>A character string specifying the file format to use when saving the output. The default value <code>"individual"</code> saves each element of the returned list as an individual uncompressed file, with data frames saved in csv format. For better compression, the values <code>"rda"</code> and <code>"rds"</code> save the returned list as a single file using the rda and rds format, respectively (in the former case, the list will be named <code>net</code> within the rda file). Regardless of the argument value, any plots generated will saved to a pdf file containing one plot per page.</p></dd>

  <dt>output_name</dt>
<dd><p>A character string. All files saved will have file names beginning with this value.</p></dd>

  <dt>pdf_width</dt>
<dd><p>Sets the width of each plot when writing to pdf. Passed to <code><a href="saveNetwork.html">saveNetwork</a></code>.</p></dd>

  <dt>pdf_height</dt>
<dd><p>Sets the height of each plot when writing to pdf. Passed to <code><a href="saveNetwork.html">saveNetwork</a></code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>To construct the immune repertoire network, each TCR/BCR clone (bulk data) or cell (single-cell data) is modeled as a node in the network graph, corresponding to a single row of the AIRR-Seq data. For each node, the corresponding receptor sequence is considered. Both nucleotide and amino acid sequences are supported for this purpose. The receptor sequence is used as the basis of similarity and distance between nodes in the network.</p>
<p>Similarity between sequences is measured using either the Hamming distance or Levenshtein (edit) distance. The similarity determines the pairwise distance between nodes in the network graph. The more similar two sequences are, the shorter the distance between their respective nodes. Two nodes in the graph are joined by an edge if the distance between them is sufficiently small, i.e., if their receptor sequences are sufficiently similar.</p>
<p>See the <a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/articles/buildRepSeqNetwork.html">buildRepSeqNetwork package vignette</a> for more details. The vignette can be accessed offline using <code><a href="../articles/buildRepSeqNetwork.html">vignette("buildRepSeqNetwork")</a></code>.</p>
    </div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>If the constructed network contains no nodes, the function will return <code>NULL</code> with a warning. Otherwise, the function invisibly returns a list containing the following items:</p>
<dl><dt>igraph</dt>
<dd><p>An object of class <code><a href="https://r.igraph.org/reference/aaa-igraph-package.html" class="external-link">igraph</a></code> containing the list of nodes and edges for the network graph.</p></dd>

    <dt>adjacency_matrix</dt>
<dd><p>The network graph adjacency matrix, stored as a sparse matrix of class <code>dgCMatrix</code> from the <code>Matrix</code> package. See <code><a href="https://rdrr.io/pkg/Matrix/man/dgCMatrix-class.html" class="external-link">dgCMatrix-class</a></code>.</p></dd>

    <dt>node_data</dt>
<dd><p>A data frame containing containing metadata for the network nodes, where each row corresponds to a node in the network graph. This data frame contains all variables from <code>data</code> (unless otherwise specified via <code>subset_cols</code>) in addition to the computed node-level network properties if <code>node_stats = TRUE</code>. Each row's name is the name of the corresponding row from <code>data</code>.</p></dd>

    <dt>cluster_data</dt>
<dd><p>A data frame containing network properties for the clusters, where each row corresponds to a cluster in the network graph. Only included if <code>cluster_stats = TRUE</code>.</p></dd>

    <dt>plots</dt>
<dd><p>A list containing each plot generated for the network graph. Each plot is an object of class <code><a href="https://ggraph.data-imaginist.com/reference/ggraph.html" class="external-link">ggraph</a></code>. Only included if <code>plots = TRUE</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Hai Yang, Jason Cham, Brian Neal, Zenghua Fan, Tao He and Li Zhang. (2023). NAIR: Network Analysis of Immune Repertoire. <em>Frontiers in Immunology</em>, vol. 14. <a href="https://www.frontiersin.org/articles/10.3389/fimmu.2023.1181825/full" class="external-link">doi: 10.3389/fimmu.2023.1181825</a></p>
<p><a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/index.html">Webpage for the NAIR package</a></p>
<p><a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/articles/buildRepSeqNetwork.html">buildRepSeqNetwork vignette</a></p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Brian Neal (<a href="mailto:Brian.Neal@ucsf.edu">Brian.Neal@ucsf.edu</a>)</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">toy_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="simulateToyData.html">simulateToyData</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">network</span> <span class="op">&lt;-</span> <span class="fu">buildRepSeqNetwork</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">toy_data</span>, seq_col <span class="op">=</span> <span class="st">"CloneSeq"</span>,</span></span>
<span class="r-in"><span>  node_stats <span class="op">=</span> <span class="cn">TRUE</span>, cluster_stats <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>  color_nodes_by <span class="op">=</span> <span class="st">"transitivity"</span>, color_scheme <span class="op">=</span> <span class="st">"plasma-1"</span>,</span></span>
<span class="r-in"><span>  size_nodes_by <span class="op">=</span> <span class="st">"degree"</span>, node_size_limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">1.5</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  output_dir <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Input data contains 200 rows.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Removing sequences with length fewer than 3 characters... Done. 200 rows remaining.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Computing network edges based on a max hamming distance of 1... Done.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Network contains 122 nodes (after removing isolated nodes).</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Computing node-level network statistics... Done.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Computing cluster membership within the network... Done.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Computing statistics for the 20 clusters in the network... Done.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Generating graph plot with nodes colored by transitivity...</span>
<span class="r-plt img"><img src="buildRepSeqNetwork-1.png" alt="" width="700" height="433"></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  Done.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Brian Neal, Hai Yang, Phi Long Le, Li Zhang.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

