<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="NAIR">
<title>Searching for Associated TCR/BCR Clusters • NAIR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Searching for Associated TCR/BCR Clusters">
<meta property="og:description" content="NAIR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">NAIR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.9045</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/NAIR.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/buildRepSeqNetwork.html">buildRepSeqNetwork()/buildNet()</a>
    <a class="dropdown-item" href="../articles/node_properties.html">Node-Level Network Properties</a>
    <a class="dropdown-item" href="../articles/cluster_analysis.html">Cluster Analysis</a>
    <a class="dropdown-item" href="../articles/network_visualization.html">Network Visualization</a>
    <a class="dropdown-item" href="../articles/supplementary.html">Supplementary Functions</a>
    <a class="dropdown-item" href="../articles/dual_chain.html">Dual-Chain Network Analysis</a>
    <a class="dropdown-item" href="../articles/associated_clusters.html">Searching for Associated TCR/BCR Clusters</a>
    <a class="dropdown-item" href="../articles/public_clusters.html">Searching for Public TCR/BCR Clusters</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/mlizhangx/Network-Analysis-for-Repertoire-Sequencing-/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Searching for Associated TCR/BCR Clusters</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/mlizhangx/Network-Analysis-for-Repertoire-Sequencing-/blob/HEAD/vignettes/associated_clusters.Rmd" class="external-link"><code>vignettes/associated_clusters.Rmd</code></a></small>
      <div class="d-none name"><code>associated_clusters.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Given multiple samples of bulk Adaptive Immune Receptor Repertoire
Sequencing (AIRR-Seq) data, the <code>NAIR</code> package can be used to
search for TCR/BCR clusters associated with a binary variable of
interest, such as a disease condition, treatment or clinical
outcome.</p>
<div class="section level5">
<h5 id="overview-of-process">Overview of Process<a class="anchor" aria-label="anchor" href="#overview-of-process"></a>
</h5>
<ol style="list-style-type: decimal">
<li>
<a href="#step1"><strong>Identify associated sequences.</strong></a>
Divide the subjects into two groups based on the two levels of the
binary variable. Identify TCR/BCR sequences that exhibit a statistically
significant difference in frequency between the two groups using
Fisher’s exact test.</li>
<li>
<a href="#step2"><strong>Identify clones with sequences similar to
the associated sequences.</strong></a> For a single associated sequence,
all sequences that fall within a certain distance (e.g., those that
differ by at most one amino acid) comprise its neighborhood. From all
samples, identify all clones whose sequences belong to this
neighborhood. Repeat for each associated sequence.</li>
<li>
<a href="#step3"><strong>Construct global network using identified
clones.</strong></a> Combine the clones from step 2 into a single global
network. Use cluster analysis to partition the global network into
clusters, which are considered as the associated clusters.</li>
<li>
<a href="#step4">Perform additional tasks</a> such as labeling the
global clusters in the visual plot and analyzing individual clusters of
interest.</li>
</ol>
</div>
<div class="section level5">
<h5 id="simulate-data-for-demonstration">Simulate Data for Demonstration<a class="anchor" aria-label="anchor" href="#simulate-data-for-demonstration"></a>
</h5>
<p>We simulate some toy data for demonstration.</p>
<p>As our binary variable of interest, we consider a single treatment
factor with two levels, labeled treatment and control.</p>
<p>In each of the two groups, we simulate 15 samples, each containing 30
observations. The generation probabilities of the possible sequences are
fixed within each group. In order to simulate the treatment effect, the
generation probabilities of certain sequences differ substantially
between the two groups.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">42</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/">NAIR</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Welcome to NAIR: Network Analysis of Immune Repertoire.</span></span>
<span><span class="co">#&gt; Get started using `vignette("NAIR", package = "NAIR")`, or by visiting</span></span>
<span><span class="co">#&gt; https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/</span></span>
<span><span class="va">data_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">dir_input_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"input_samples"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">dir_input_samples</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Number of samples by control/treatment group</span></span>
<span><span class="va">n_control</span> <span class="op">&lt;-</span> <span class="va">n_treatment</span> <span class="op">&lt;-</span> <span class="fl">15</span> </span>
<span><span class="va">n_samples</span> <span class="op">&lt;-</span> <span class="va">n_control</span> <span class="op">+</span> <span class="va">n_treatment</span></span>
<span><span class="va">sample_size</span> <span class="op">&lt;-</span> <span class="fl">30</span> <span class="co"># (seqs per sample)          </span></span>
<span></span>
<span><span class="co"># sequences (first five are chosen to be associated with treatment)</span></span>
<span><span class="va">base_seqs</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CASSGAYEQYF"</span>, <span class="st">"CSVDLGKGNNEQFF"</span>, </span>
<span>    <span class="st">"CASSIEGQLSTDTQYF"</span>, <span class="co"># 3 of the associated</span></span>
<span>    <span class="st">"CASSEEGQLSTDTQYF"</span>, <span class="co"># sequences differ by</span></span>
<span>    <span class="st">"CASSPEGQLSTDTQYF"</span>, <span class="co"># only one amino acid</span></span>
<span>    <span class="st">"RASSLAGNTEAFF"</span>, <span class="st">"CASSHRGTDTQYF"</span>, <span class="st">"CASDAGVFQPQHF"</span></span>
<span>  <span class="op">)</span> </span>
<span></span>
<span><span class="co"># relative generation probabilities by control/treatment group</span></span>
<span><span class="va">pgen_c</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">30</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span>, times <span class="op">=</span> <span class="va">n_control</span><span class="op">)</span>, </span>
<span>                 nrow <span class="op">=</span> <span class="va">n_control</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">pgen_t</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">/</span><span class="fl">3</span>, <span class="fl">3</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span>, times <span class="op">=</span> <span class="va">n_treatment</span><span class="op">)</span>, </span>
<span>                 nrow <span class="op">=</span> <span class="va">n_treatment</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">pgen</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">pgen_c</span>, <span class="va">pgen_t</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/simulateToyData.html">simulateToyData</a></span><span class="op">(</span>    </span>
<span>  samples <span class="op">=</span> <span class="va">n_samples</span>, </span>
<span>  sample_size <span class="op">=</span> <span class="va">sample_size</span>,</span>
<span>  prefix_length <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  prefix_chars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">""</span>, <span class="st">""</span><span class="op">)</span>,</span>
<span>  prefix_probs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>, <span class="va">n_samples</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">n_samples</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  affixes <span class="op">=</span> <span class="va">base_seqs</span>, </span>
<span>  affix_probs <span class="op">=</span> <span class="va">pgen</span>, </span>
<span>  num_edits <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  output_dir <span class="op">=</span> <span class="va">dir_input_samples</span>, </span>
<span>  no_return <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
<p>Each sample’s data frame is saved to its own file using the RDS file
format. The files are named “<code>Sample1.rds</code>”,
“<code>Sample2.rds</code>”, etc. A character string containing the
directory path is assigned to the R environment variable
<code>dir_input_samples</code> for later reference.</p>
<p>The first few rows of the data for the first sample appear as
follows:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># View first few rows of data for sample 1</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir_input_samples</span>, <span class="st">"Sample1.rds"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;        CloneSeq CloneFrequency CloneCount SampleID</span></span>
<span><span class="co">#&gt; 1 CASDAGVFQPQHF     0.02606559       2832  Sample1</span></span>
<span><span class="co">#&gt; 2 CASDAGVFQPQHF     0.03718396       4040  Sample1</span></span>
<span><span class="co">#&gt; 3 CASSHRGTDTQYF     0.03182726       3458  Sample1</span></span>
<span><span class="co">#&gt; 4 CASDAGVFQPQHF     0.04615781       5015  Sample1</span></span>
<span><span class="co">#&gt; 5 RASSLAGNTEAFF     0.06006498       6526  Sample1</span></span>
<span><span class="co">#&gt; 6 CASDAGVFQPQHF     0.03363123       3654  Sample1</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 class="tabset" id="step-1-find-associated-sequences">
<a name="step1"></a> Step 1: Find Associated
Sequences<a class="anchor" aria-label="anchor" href="#step-1-find-associated-sequences"></a>
</h2>
<p>First, we use <code><a href="../reference/findAssociatedSeqs.html">findAssociatedSeqs()</a></code> to conduct a
systematic search for associated sequences within the provided
samples.</p>
<p>This search is a two-stage procedure.</p>
<p>The unique receptor sequences are first filtered according to basic
criteria in order to narrow the list of candidates.</p>
<p>For each candidate sequence, we then perform Fisher’s exact test of
independence between the binary variable of interest and the observed
presence of the sequence within a sample/subject. Significant
associations are identified based on a specified P-value threshold.</p>
<p>Below, we explain how to use <code><a href="../reference/findAssociatedSeqs.html">findAssociatedSeqs()</a></code>.</p>
<div class="section level3">
<h3 class="tabset" id="input-data-for-step-1">
<a name="input_settings"></a> Input Data for Step
1<a class="anchor" aria-label="anchor" href="#input-data-for-step-1"></a>
</h3>
<p>Each sample’s AIRR-Seq data must be contained in a separate file,
with observations indexed by row, and with the same columns across
samples.</p>
<div class="section level5">
<h5 id="file-paths-of-sample-data">
<a name="input_files"></a> File Paths of Sample Data<a class="anchor" aria-label="anchor" href="#file-paths-of-sample-data"></a>
</h5>
<p>The <code>file_list</code> parameter accepts a character vector
containing file paths (or a list containing file paths and connections),
where each element corresponds to a file containing a single sample.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create vector of input file paths for step 1 (one per sample)</span></span>
<span><span class="va">input_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir_input_samples</span>,</span>
<span>                         <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Sample"</span>, <span class="fl">1</span><span class="op">:</span><span class="va">n_samples</span>, <span class="st">".rds"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">input_files</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 30</span></span>
<span><span class="va">input_files</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="co">#&gt; [1] "/tmp/RtmpVKpOuB/input_samples/Sample1.rds"</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="file-format-of-sample-data">
<a name="input_type"></a> File Format of Sample Data<a class="anchor" aria-label="anchor" href="#file-format-of-sample-data"></a>
</h5>
<p>The file format of the input files is specified by the
<code>input_type</code> parameter. The supported values are
<code>"rds"</code>, <code>"rda"</code>, <code>"csv"</code>,
<code>"csv2"</code>, <code>"tsv"</code> and <code>"table"</code>.
Depending on the input type, further options are specified with
<code>data_symbols</code> or <code>read.args</code>.</p>
<p>Refer <a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/articles/supplementary.html#load">here</a>
and to <code><a href="../reference/combineSamples.html">loadDataFromFileList()</a></code> for details and examples.</p>
</div>
<div class="section level5">
<h5 id="sequence-column-in-sample-data">
<a name="seq_col"></a> Sequence Column in Sample Data<a class="anchor" aria-label="anchor" href="#sequence-column-in-sample-data"></a>
</h5>
<p>The <code>seq_col</code> parameter specifies the column containing
the TCR/BCR sequences within each sample. It accepts the column name (as
a character string) or the column position index.</p>
</div>
<div class="section level5">
<h5 id="assigning-samples-to-groups">
<a name="group_ids"></a> Assigning Samples to Groups<a class="anchor" aria-label="anchor" href="#assigning-samples-to-groups"></a>
</h5>
<p>The <code>group_ids</code> parameter assigns each sample to one of
two groups, representing the two levels of the binary variable of
interest. It accepts a vector of the same length as
<code>file_list</code>, where each entry is the group label for the
corresponding sample. Any two distinct values may be used for the group
labels.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create vector of group labels for the samples</span></span>
<span><span class="va">group_labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"reference"</span>, <span class="va">n_control</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"comparison"</span>, <span class="va">n_treatment</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="assigning-samples-to-subjects-if-applicable">
<a name="subject_ids"></a> Assigning Samples to Subjects (If
Applicable)<a class="anchor" aria-label="anchor" href="#assigning-samples-to-subjects-if-applicable"></a>
</h5>
<p>The optional <code>subject_ids</code> parameter assigns each sample
to a subject. It accepts a vector of the same length as
<code>file_list</code>, where each entry is the subject ID for the
corresponding sample.</p>
<p>Without subjects, each sample is an independent observational unit.
The contingency table for Fisher’s exact test then counts
<strong>samples</strong> possessing the given sequence.</p>
<p>With subjects, all samples from the same subject form one
observational unit. The contingency table then counts
<strong>subjects</strong> possessing the given sequence. A sequence is
counted at most once per subject, and each subject is counted exactly
once in each margin.</p>
<p>Subject IDs should be provided when the binary variable of interest
is subject-specific <strong>and</strong> the data contains multiple
samples from a single subject.</p>
</div>
<div class="section level5">
<h5 id="clone-frequency-column-optional">
<a name="freq_col"></a> Clone Frequency Column (Optional)<a class="anchor" aria-label="anchor" href="#clone-frequency-column-optional"></a>
</h5>
<p>The optional <code>freq_col</code> parameter specifies the column
containing the clone frequencies (normalized measure of abundance)
within each sample. It accepts the column name (as a character string)
or the column position index. If provided, the maximum clone count
(across all samples) of each associated sequence will be included in the
output.</p>
</div>
</div>
<div class="section level3">
<h3 class="tabset" id="search-criteria">Search Criteria<a class="anchor" aria-label="anchor" href="#search-criteria"></a>
</h3>
<p>The following parameters control the criteria used when searching for
associated sequences. These include filter criteria that determine which
sequences are considered for testing, as well as the P-value cutoff
below which an association is detected by Fisher’s exact test.</p>
<div class="section level5">
<h5 id="sample-membership">Sample Membership<a class="anchor" aria-label="anchor" href="#sample-membership"></a>
</h5>
<p>By default, only sequences that appear in at least five samples will
be considered. This can be changed by setting the
<code>min_sample_membership</code> parameter to a different value. An
argument value of <code>NULL</code> bypasses this check.</p>
</div>
<div class="section level5">
<h5 id="sequence-length">Sequence Length<a class="anchor" aria-label="anchor" href="#sequence-length"></a>
</h5>
<p>By default, only sequences with a length of at least seven characters
will be considered. This can be changed by setting the
<code>min_seq_length</code> parameter to a different value. An argument
value of <code>NULL</code> bypasses this check.</p>
</div>
<div class="section level5">
<h5 id="sequence-content">Sequence Content<a class="anchor" aria-label="anchor" href="#sequence-content"></a>
</h5>
<p>Sequences containing any of the characters <code>*</code>,
<code>|</code> or <code>_</code> will be excluded from consideration.
This can be changed using the <code>drop_matches</code> parameter, which
accepts a character string containing a regular expression that
specifies a pattern of content to search for. Each sequence is checked
for a pattern match using <code><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep()</a></code>. If a match is found, the
sequence is excluded. An argument value of <code>NULL</code> bypasses
this check.</p>
</div>
<div class="section level5">
<h5 id="p-value-cutoff">P-value Cutoff<a class="anchor" aria-label="anchor" href="#p-value-cutoff"></a>
</h5>
<p>By default, sequences with a P-value below 0.05 on Fisher’s exact
test are included in the output of <code><a href="../reference/findAssociatedSeqs.html">findAssociatedSeqs()</a></code>.
The cutoff can be set to a different value using the
<code>pval_cutoff</code> parameter. The lower the cutoff value, the
stronger the evidence of an association is required for a sequence to be
included in the output.</p>
<p>It should be noted that the output includes the P-value for each
associated sequence; furthermore, the sequences are ordered by P-value,
and any subset of them can be used in <a href="#step2">step2</a>. Thus
it is not necessary to rerun <code><a href="../reference/findAssociatedSeqs.html">findAssociatedSeqs()</a></code> in order
to impose a stricter P-value cutoff. The user can instead subset the
results in order to use only those sequences below a specified
P-value.</p>
</div>
</div>
<div class="section level3">
<h3 id="output-settings-for-step-1">Output Settings for Step 1<a class="anchor" aria-label="anchor" href="#output-settings-for-step-1"></a>
</h3>
<p><code><a href="../reference/findAssociatedSeqs.html">findAssociatedSeqs()</a></code> returns a data frame (see
demonstration below). The data frame can be written to a csv file by
providing a file path to the <code>outfile</code> parameter.</p>
</div>
<div class="section level3">
<h3 id="demonstration-step-1">
<a name="demo1"></a> Demonstration, Step 1<a class="anchor" aria-label="anchor" href="#demonstration-step-1"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">associated_seqs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/findAssociatedSeqs.html">findAssociatedSeqs</a></span><span class="op">(</span>file_list <span class="op">=</span> <span class="va">input_files</span>, </span>
<span>                                      input_type <span class="op">=</span> <span class="st">"rds"</span>, </span>
<span>                                      group_ids <span class="op">=</span> <span class="va">group_labels</span>, </span>
<span>                                      seq_col <span class="op">=</span> <span class="st">"CloneSeq"</span>, </span>
<span>                                      min_seq_length <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>                                      drop_matches <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>                                      min_sample_membership <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>                                      pval_cutoff <span class="op">=</span> <span class="fl">0.1</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><code><a href="../reference/findAssociatedSeqs.html">findAssociatedSeqs()</a></code> returns a data frame containing the
associated sequences based on the specified search criteria.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">associated_seqs</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;        ReceptorSeq fisher_pvalue shared_by_n_samples samples_g0 samples_g1</span></span>
<span><span class="co">#&gt; 8   CSVDLGKGNNEQFF  1.052106e-05                  18          3         15</span></span>
<span><span class="co">#&gt; 7      CASSGAYEQYF  1.157316e-04                  17          3         14</span></span>
<span><span class="co">#&gt; 4 CASSEEGQLSTDTQYF  5.197401e-03                  10          1          9</span></span>
<span><span class="co">#&gt; 5 CASSIEGQLSTDTQYF  6.559548e-02                  16          5         11</span></span></code></pre></div>
<p>Each row corresponds to a unique receptor sequence. The rows are
ordered by <span class="math inline">\(P\)</span>-value.</p>
<p>For the remaining steps, we will only need to use the
<code>ReceptorSeq</code> column. The documentation for
<code><a href="../reference/findAssociatedSeqs.html">findAssociatedSeqs()</a></code> describes the other variables.</p>
</div>
</div>
<div class="section level2">
<h2 class="tabset" id="step-2-find-associated-clones">
<a name="step2"></a> Step 2: Find Associated
Clones<a class="anchor" aria-label="anchor" href="#step-2-find-associated-clones"></a>
</h2>
<p><code><a href="../reference/findAssociatedClones.html">findAssociatedClones()</a></code> searches across samples and
identifies clones with sequences similar to any of the associated
sequences identified in step 1.</p>
<p>For each associated sequence, we define its neighborhood as the
collection of all sequences that fall within a specified distance (e.g.,
a maximum Hamming distance of 1). We then identify the clones (from all
samples) whose sequences belong to this neighborhood.</p>
<div class="section level3">
<h3 id="associated-sequences">
<a name="assoc_seqs"></a> Associated Sequences<a class="anchor" aria-label="anchor" href="#associated-sequences"></a>
</h3>
<p>The <code>assoc_seqs</code> parameter specifies the associated
sequences. It accepts a character vector, which is typically the
<code>ReceptorSeq</code> column from the output of
<code><a href="../reference/findAssociatedSeqs.html">findAssociatedSeqs()</a></code>, or a subset of this column if not all
associated sequences from step 1 are to be considered (the column is
ordered by P-value to facilitate subsetting).</p>
</div>
<div class="section level3">
<h3 class="tabset" id="input-data-for-step-2">Input Data for Step 2<a class="anchor" aria-label="anchor" href="#input-data-for-step-2"></a>
</h3>
<div class="section level5">
<h5 id="argument-values-from-step-1">
<a name="files2"></a> Argument Values from Step 1<a class="anchor" aria-label="anchor" href="#argument-values-from-step-1"></a>
</h5>
<p>With <code><a href="../reference/findAssociatedClones.html">findAssociatedClones()</a></code>, we specify our sample data
using the same parameters from <code><a href="../reference/findAssociatedSeqs.html">findAssociatedSeqs()</a></code>:</p>
<ul>
<li><a href="#input_files"><code>file_list</code></a></li>
<li><a href="#input_type"><code>input_type</code></a></li>
<li><a href="#seq_col"><code>seq_col</code></a></li>
<li><a href="#group_ids"><code>group_ids</code></a></li>
<li>
<a href="#subject_ids"><code>subject_ids</code></a> (if
applicable)</li>
</ul>
<p>The user should provide the same argument values used in the call to
<code><a href="../reference/findAssociatedSeqs.html">findAssociatedSeqs()</a></code> during <a href="#step1">step
1</a>.</p>
</div>
<div class="section level5">
<h5 id="custom-sample-ids-optional">Custom Sample IDs (Optional)<a class="anchor" aria-label="anchor" href="#custom-sample-ids-optional"></a>
</h5>
<p>Each clone’s sample ID is included in the output. By default, these
are <code>"Sample1"</code>, <code>"Sample2"</code>, etc., according to
the order in <code>file_list</code>.</p>
<p>The optional <code>sample_ids</code> parameter assigns custom sample
IDs. It accepts a vector of the same length as <code>file_list</code>,
where each entry is the corresponding sample ID.</p>
</div>
<div class="section level5">
<h5 id="filtering-the-sample-data">Filtering the Sample Data<a class="anchor" aria-label="anchor" href="#filtering-the-sample-data"></a>
</h5>
<p>The clones from each sample are filtered to remove any irrelevant
data. By default, clones with sequences that are less than six
characters in length, as well as sequences containing any of the
characters <code>*</code>, <code>_</code> or <code>|</code>, will be
excluded. The <code>min_seq_length</code> and <code>drop_matches</code>
parameters control the filter settings. Refer <a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/articles/buildRepSeqNetwork.html#input_options">here</a>
for details.</p>
</div>
</div>
<div class="section level3">
<h3 class="tabset" id="defining-the-neighborhoods">Defining the Neighborhoods<a class="anchor" aria-label="anchor" href="#defining-the-neighborhoods"></a>
</h3>
<p>The neighborhood of an associated sequence includes all sequences
that are sufficiently similar to it. By default, similarity is measured
using the <a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/reference/hamDistBounded.html">Hamming
distance</a>, and “sufficiently similar” means a Hamming distance of at
most 1.</p>
<p>The <a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/reference/levDistBounded.html">Levenshtein
distance</a> can be used by specifying <code>dist_type = "lev"</code>.
As described <a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/articles/buildRepSeqNetwork.html#distance-function">here</a>,
it is preferable for comparing nucleotide sequences, but has greater
computational cost.</p>
<p>The maximum distance between an associated sequence other sequences
in its neighborhood is specified by the <code>nbd_radius</code>
parameter (the default is 1). Increasing the value increases the
potential number of clones in each neighborhood. A value of 0 means each
neighborhood includes only clones with the exact associated
sequence.</p>
</div>
<div class="section level3">
<h3 class="tabset" id="output-settings-for-step-2">Output Settings for Step 2<a class="anchor" aria-label="anchor" href="#output-settings-for-step-2"></a>
</h3>
<p><code><a href="../reference/findAssociatedClones.html">findAssociatedClones()</a></code> does not return any direct
output. Instead, the AIRR-Seq data for the neighborhoods is saved to
files to be used as inputs in <a href="#step3">step 3</a>. The following
parameters control the output settings.</p>
<div class="section level5">
<h5 id="variables-to-keep-from-sample-data">
<a name="nbd_variables"></a> Variables to Keep From Sample Data<a class="anchor" aria-label="anchor" href="#variables-to-keep-from-sample-data"></a>
</h5>
<p>By default, each neighborhood’s <a href="#nbd_output">AIRR-Seq
data</a> includes all variables from the original sample data. These
variables can be <a href="#vis">used as metadata in visualizations</a>
of the global network in <a href="#step3">step 3</a>.</p>
<p>To keep only a subset of the original variables, specify the
variables to keep using the <code>subset_cols</code> parameter, which
accepts a character vector of column names or a vector of column
indices. The sequence column is always included.</p>
</div>
<div class="section level5">
<h5 id="output-directory-for-step-2">
<a name="output2"></a> Output Directory for Step 2<a class="anchor" aria-label="anchor" href="#output-directory-for-step-2"></a>
</h5>
<p>The <code>output_dir</code> parameter specifies the output directory.
It accepts a character string containing the directory path. The
directory will be created if it does not exist.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># output directory for current step</span></span>
<span><span class="va">dir_nbds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"assoc_seq_nbds"</span><span class="op">)</span></span></code></pre></div>
<p>Using an empty directory simplifies the task of <a href="#input_files3">inputting the files during step 3</a>.</p>
</div>
<div class="section level5">
<h5 id="output-file-format-for-step-2">Output File Format for Step 2<a class="anchor" aria-label="anchor" href="#output-file-format-for-step-2"></a>
</h5>
<p>By default, each neighborhood’s data frame is saved as an RDS file.
This can be changed using the <code>output_type</code> parameter. Other
accepted values are <code>"rda"</code>, <code>"csv"</code>,
<code>"csv2"</code>, <code>"tsv"</code> and <code>"table"</code>. For
<code>"rda"</code>, data frames are named <code>data</code> in the R
environment. For the remaining options, <code><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table()</a></code> is
called with <code>row.names = TRUE</code>.</p>
</div>
</div>
<div class="section level3">
<h3 id="demonstration-step-2">
<a name="demo2"></a> Demonstration, Step 2<a class="anchor" aria-label="anchor" href="#demonstration-step-2"></a>
</h3>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Identify clones in a neighborhood around each associated sequence</span></span>
<span><span class="fu"><a href="../reference/findAssociatedClones.html">findAssociatedClones</a></span><span class="op">(</span>file_list <span class="op">=</span> <span class="va">input_files</span>, </span>
<span>                     input_type <span class="op">=</span> <span class="st">"rds"</span>, </span>
<span>                     group_ids <span class="op">=</span> <span class="va">group_labels</span>, </span>
<span>                     seq_col <span class="op">=</span> <span class="st">"CloneSeq"</span>, </span>
<span>                     assoc_seqs <span class="op">=</span> <span class="va">associated_seqs</span><span class="op">$</span><span class="va">ReceptorSeq</span>,</span>
<span>                     min_seq_length <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>                     drop_matches <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                     output_dir <span class="op">=</span> <span class="va">dir_nbds</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The following files are created in the output directory:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Files created by findAssociatedClones()</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">dir_nbds</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "assoc_nbd_1_CSVDLGKGNNEQFF.rds"   "assoc_nbd_2_CASSGAYEQYF.rds"     </span></span>
<span><span class="co">#&gt; [3] "assoc_nbd_3_CASSEEGQLSTDTQYF.rds" "assoc_nbd_4_CASSIEGQLSTDTQYF.rds"</span></span></code></pre></div>
<p><a name="nbd_output"></a></p>
<p>Each file contains neighborhood data for a single associated
sequence.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># first neighborhood data file</span></span>
<span><span class="va">one_nbd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">dir_nbds</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># first few rows</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">one_nbd</span><span class="op">)</span></span>
<span><span class="co">#&gt;                   CloneSeq CloneFrequency CloneCount SampleID    GroupID</span></span>
<span><span class="co">#&gt; Sample5.14  CSVDLGKGNNEQFF     0.02668662       2924  Sample5  reference</span></span>
<span><span class="co">#&gt; Sample7.17  CSVDLGKGNNEQFF     0.01957279       2113  Sample7  reference</span></span>
<span><span class="co">#&gt; Sample10.17 CSVDLGKGNNEQFF     0.03594648       4097 Sample10  reference</span></span>
<span><span class="co">#&gt; Sample16.10 CSVDLGKGNNEQFF     0.03404553       3802 Sample16 comparison</span></span>
<span><span class="co">#&gt; Sample16.28 CSVDLGKGNNEQFF     0.02736537       3056 Sample16 comparison</span></span>
<span><span class="co">#&gt; Sample17.3  CSVDLGKGNNEQFF     0.04114818       4603 Sample17 comparison</span></span>
<span><span class="co">#&gt;                   AssocSeq</span></span>
<span><span class="co">#&gt; Sample5.14  CSVDLGKGNNEQFF</span></span>
<span><span class="co">#&gt; Sample7.17  CSVDLGKGNNEQFF</span></span>
<span><span class="co">#&gt; Sample10.17 CSVDLGKGNNEQFF</span></span>
<span><span class="co">#&gt; Sample16.10 CSVDLGKGNNEQFF</span></span>
<span><span class="co">#&gt; Sample16.28 CSVDLGKGNNEQFF</span></span>
<span><span class="co">#&gt; Sample17.3  CSVDLGKGNNEQFF</span></span></code></pre></div>
<p>The row names identify the sample and original row of each clone.</p>
<p>In addition to the <a href="#nbd_variables">variables kept</a> from
the original data, the following variables have been added:</p>
<ul>
<li><code>SampleID</code></li>
<li><code>GroupID</code></li>
<li>
<code>AssocSeq</code> (the associated sequence for the
neighborhood)</li>
</ul>
<p>If <code>subject_ids</code> was specified, a variable
<code>SubjectID</code> will also be present.</p>
</div>
</div>
<div class="section level2">
<h2 class="tabset" id="step-3-global-network-of-associated-clusters">
<a name="step3"></a> Step 3: Global Network of
Associated Clusters<a class="anchor" aria-label="anchor" href="#step-3-global-network-of-associated-clusters"></a>
</h2>
<p><code><a href="../reference/buildAssociatedClusterNetwork.html">buildAssociatedClusterNetwork()</a></code> combines the clones from
all neighborhoods into a single global network. Cluster analysis is used
to partition the global network into clusters, which are considered as
the associated clusters.</p>
<div class="section level3">
<h3 class="tabset" id="input-data-for-step-3">Input Data for Step 3<a class="anchor" aria-label="anchor" href="#input-data-for-step-3"></a>
</h3>
<p>The input files for <code><a href="../reference/buildAssociatedClusterNetwork.html">buildAssociatedClusterNetwork()</a></code> are
the <a href="#demo2">output files from step 2</a>. Each file contains
the AIRR-Seq data for the clones belonging to a particular
neighborhood.</p>
<div class="section level5">
<h5 id="file-paths-of-neighborhood-data">
<a name="input_files3"></a> File Paths of Neighborhood Data<a class="anchor" aria-label="anchor" href="#file-paths-of-neighborhood-data"></a>
</h5>
<p>As in previous steps, the <code>file_list</code> parameter specifies
the input files, which are located in the <a href="#output2">output
directory from step 2</a>. If an empty output directory was used, they
can be easily referenced using <code><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files()</a></code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create vector of file paths to files from step 2</span></span>
<span><span class="va">nbd_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">dir_nbds</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="file-format-of-neighborhood-data">File Format of Neighborhood Data<a class="anchor" aria-label="anchor" href="#file-format-of-neighborhood-data"></a>
</h5>
<p>If <code><a href="../reference/findAssociatedClones.html">findAssociatedClones()</a></code> was called with a non-default
value of <code>output_type</code>, this value must be passed to the
<code>input_type</code> parameter of
<code><a href="../reference/buildAssociatedClusterNetwork.html">buildAssociatedClusterNetwork()</a></code>.</p>
</div>
</div>
<div class="section level3">
<h3 class="tabset" id="global-network-analysis">Global Network Analysis<a class="anchor" aria-label="anchor" href="#global-network-analysis"></a>
</h3>
<div class="section level5">
<h5 id="network-construction">Network Construction<a class="anchor" aria-label="anchor" href="#network-construction"></a>
</h5>
<p>The parameters that control construction of the global network are
shown below along with their default values.</p>
<ul>
<li><code>dist_type = "hamming"</code></li>
<li><code>dist_cutoff = 1</code></li>
<li><code>drop_isolated_nodes = FALSE</code></li>
</ul>
<p>Refer <a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/articles/buildRepSeqNetwork.html#network_settings">here</a>
for their meaning and usage.</p>
</div>
<div class="section level5">
<h5 id="clustering-algorithm">Clustering Algorithm<a class="anchor" aria-label="anchor" href="#clustering-algorithm"></a>
</h5>
<p>A clustering algorithm is used to partition the global network graph
into densely-connected subgraphs (clusters). Each cluster represents a
distinct group of clones that are associated with the binary variable of
interest.</p>
<p>By default, clustering is performed using
<code><a href="https://r.igraph.org/reference/cluster_fast_greedy.html" class="external-link">igraph::cluster_fast_greedy()</a></code>. A different algorithm can be
specified using the <code>cluster_fun</code> parameter, as described <a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/articles/cluster_analysis.html#clustering-algorithm">here</a>.</p>
</div>
<div class="section level5">
<h5 id="network-properties">
<a name="global_network_properties"></a> Network Properties<a class="anchor" aria-label="anchor" href="#network-properties"></a>
</h5>
<p>By default, <code><a href="../reference/buildAssociatedClusterNetwork.html">buildAssociatedClusterNetwork()</a></code> computes <a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/articles/buildRepSeqNetwork.html#node-level-network-properties">node-level</a>
and <a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/articles/cluster_analysis#cluster-properties">cluster-level</a>
properties of the global network. These can be excluded by setting
<code>node_stats = FALSE</code> and <code>cluster_stats = FALSE</code>,
respectively. The set of node-level properties to be computed can be
specified using the <code>stats_to_include</code> parameter, as
described <a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/articles/node_properties.html#choose_node_stats">here</a>.</p>
</div>
<div class="section level5">
<h5 id="visualization-of-global-network">
<a name="vis"></a> Visualization of Global Network<a class="anchor" aria-label="anchor" href="#visualization-of-global-network"></a>
</h5>
<p>By default, <code><a href="../reference/buildAssociatedClusterNetwork.html">buildAssociatedClusterNetwork()</a></code> produces a
visual plot of the global network graph with the nodes colored according
to the binary variable of interest.</p>
<p>The <code>color_nodes_by</code> parameter specifies the variable used
to color the nodes. It accepts a character string naming a <a href="#nbd_output">variable present in the neighborhood data</a> (e.g.,
<code>"SampleID"</code>) or a <a href="#global_network_properties">computed node-level property</a>
(e.g., <code>"degree"</code>).</p>
<p><code>color_nodes_by</code> also accepts a vector naming multiple
variables. <a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/articles/network_visualization.html#generating-multiple-plots">One
plot will be created for each entry, with the nodes colored according to
the respective variable</a>. The name of the binary variable of interest
is <code>"GroupID"</code>.</p>
<p>Refer <a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/articles/network_visualization.html">here</a>
to learn about other parameters for customizing the visualization.</p>
</div>
</div>
<div class="section level3">
<h3 class="tabset" id="output-settings-for-step-3">Output Settings for Step 3<a class="anchor" aria-label="anchor" href="#output-settings-for-step-3"></a>
</h3>
<p><code><a href="../reference/buildAssociatedClusterNetwork.html">buildAssociatedClusterNetwork()</a></code> returns a list
containing plots, metadata and other network objects, with the <a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/articles/buildRepSeqNetwork.html#base-output">same
structure</a> as the output of <code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code>.</p>
<p>The output can be saved to a local directory using the parameters
<code>output_dir</code>, <code>output_type</code> and
<code>output_name</code>, whose usage is described <a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/articles/buildRepSeqNetwork.html#output-directory">here</a>.</p>
</div>
<div class="section level3">
<h3 id="demonstration-step-3">
<a name="demo3"></a> Demonstration, Step 3<a class="anchor" aria-label="anchor" href="#demonstration-step-3"></a>
</h3>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Combine neighborhoods and perform network analysis</span></span>
<span><span class="va">all_clusters</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/buildAssociatedClusterNetwork.html">buildAssociatedClusterNetwork</a></span><span class="op">(</span></span>
<span>    file_list <span class="op">=</span> <span class="va">nbd_files</span>, </span>
<span>    seq_col <span class="op">=</span> <span class="st">"CloneSeq"</span>,</span>
<span>    size_nodes_by <span class="op">=</span> <span class="fl">1.5</span>,</span>
<span>    print_plots <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="associated_clusters_files/figure-html/unnamed-chunk-12-1.png" width="100%" style="display: block; margin: auto;"></p>
<p>The returned list contains the following elements:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">all_clusters</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "details"          "igraph"           "adjacency_matrix" "node_data"       </span></span>
<span><span class="co">#&gt; [5] "cluster_data"     "plots"</span></span></code></pre></div>
<p>The elements are described <a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/articles/buildRepSeqNetwork.html#base-output">here</a>.
We inspect the node metadata and cluster metadata.</p>
<div class="section level5">
<h5 id="node-metadata-for-global-network">
<a name="node_variables"></a> Node Metadata for Global Network<a class="anchor" aria-label="anchor" href="#node-metadata-for-global-network"></a>
</h5>
<p>The list element <code>node_data</code> is a data frame containing
metadata for the network nodes, where each row represents a distinct
clone corresponding to a node in the global network graph.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">all_clusters</span><span class="op">$</span><span class="va">node_data</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">8</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;                   CloneSeq CloneFrequency CloneCount SampleID    GroupID</span></span>
<span><span class="co">#&gt; Sample5.14  CSVDLGKGNNEQFF     0.02668662       2924  Sample5  reference</span></span>
<span><span class="co">#&gt; Sample7.17  CSVDLGKGNNEQFF     0.01957279       2113  Sample7  reference</span></span>
<span><span class="co">#&gt; Sample10.17 CSVDLGKGNNEQFF     0.03594648       4097 Sample10  reference</span></span>
<span><span class="co">#&gt; Sample16.10 CSVDLGKGNNEQFF     0.03404553       3802 Sample16 comparison</span></span>
<span><span class="co">#&gt; Sample16.28 CSVDLGKGNNEQFF     0.02736537       3056 Sample16 comparison</span></span>
<span><span class="co">#&gt; Sample17.3  CSVDLGKGNNEQFF     0.04114818       4603 Sample17 comparison</span></span>
<span><span class="co">#&gt;                   AssocSeq degree cluster_id</span></span>
<span><span class="co">#&gt; Sample5.14  CSVDLGKGNNEQFF     53          3</span></span>
<span><span class="co">#&gt; Sample7.17  CSVDLGKGNNEQFF     53          3</span></span>
<span><span class="co">#&gt; Sample10.17 CSVDLGKGNNEQFF     53          3</span></span>
<span><span class="co">#&gt; Sample16.10 CSVDLGKGNNEQFF     53          3</span></span>
<span><span class="co">#&gt; Sample16.28 CSVDLGKGNNEQFF     53          3</span></span>
<span><span class="co">#&gt; Sample17.3  CSVDLGKGNNEQFF     53          3</span></span></code></pre></div>
<p>The row names identify the sample and original row of each clone.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># variables in the node metadata</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">all_clusters</span><span class="op">$</span><span class="va">node_data</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "CloneSeq"                  "CloneFrequency"           </span></span>
<span><span class="co">#&gt;  [3] "CloneCount"                "SampleID"                 </span></span>
<span><span class="co">#&gt;  [5] "GroupID"                   "AssocSeq"                 </span></span>
<span><span class="co">#&gt;  [7] "degree"                    "cluster_id"               </span></span>
<span><span class="co">#&gt;  [9] "transitivity"              "eigen_centrality"         </span></span>
<span><span class="co">#&gt; [11] "centrality_by_eigen"       "betweenness"              </span></span>
<span><span class="co">#&gt; [13] "centrality_by_betweenness" "authority_score"          </span></span>
<span><span class="co">#&gt; [15] "coreness"                  "page_rank"</span></span></code></pre></div>
<p><a href="#nbd_variables">By default</a>, all variables from the
original sample data are present. Node-level network properties are also
present, which are described <a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/articles/buildRepSeqNetwork.html#node-level-network-properties">here</a>.</p>
</div>
<div class="section level5">
<h5 id="cluster-metadata-for-global-network">
<a name="cluster_variables"></a> Cluster Metadata for Global
Network<a class="anchor" aria-label="anchor" href="#cluster-metadata-for-global-network"></a>
</h5>
<p>The list element <code>cluster_data</code> is a data frame containing
metadata for the associated clusters, where each row corresponds to a
cluster in the global network. It contains the following variables:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># variables in the cluster metadata</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">all_clusters</span><span class="op">$</span><span class="va">cluster_data</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "cluster_id"                  "node_count"                 </span></span>
<span><span class="co">#&gt;  [3] "mean_seq_length"             "mean_degree"                </span></span>
<span><span class="co">#&gt;  [5] "max_degree"                  "seq_w_max_degree"           </span></span>
<span><span class="co">#&gt;  [7] "agg_count"                   "max_count"                  </span></span>
<span><span class="co">#&gt;  [9] "seq_w_max_count"             "diameter_length"            </span></span>
<span><span class="co">#&gt; [11] "global_transitivity"         "assortativity"              </span></span>
<span><span class="co">#&gt; [13] "edge_density"                "degree_centrality_index"    </span></span>
<span><span class="co">#&gt; [15] "closeness_centrality_index"  "eigen_centrality_index"     </span></span>
<span><span class="co">#&gt; [17] "eigen_centrality_eigenvalue"</span></span></code></pre></div>
<p>These cluster properties are described <a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/articles/buildRepSeqNetwork.html#cluster-level-network-properties">here</a>.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 class="tabset" id="step-4-additional-tasks">
<a name="step4"></a> Step 4: Additional Tasks<a class="anchor" aria-label="anchor" href="#step-4-additional-tasks"></a>
</h2>
<p>After calling <code><a href="../reference/buildAssociatedClusterNetwork.html">buildAssociatedClusterNetwork()</a></code>, the
following tasks can be performed using the returned output.</p>
<div class="section level3">
<h3 id="additional-plots">Additional Plots<a class="anchor" aria-label="anchor" href="#additional-plots"></a>
</h3>
<p>We may wish to generate additional plots of the global network graph,
coloring the nodes according to different metadata to gain further
insight into the associated clusters. This can be done using
<code><a href="../reference/addPlots.html">addPlots()</a></code> as described <a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/articles/network_visualization.html#generating-new-plots">here</a>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># global network plot, nodes colored by associated sequence</span></span>
<span><span class="va">all_clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addPlots.html">addPlots</a></span><span class="op">(</span><span class="va">all_clusters</span>,</span>
<span>                         color_nodes_by <span class="op">=</span> <span class="st">"AssocSeq"</span>,</span>
<span>                         color_title <span class="op">=</span> <span class="st">"Neighborhood Sequence"</span>,</span>
<span>                         size_nodes_by <span class="op">=</span> <span class="fl">1.5</span>,</span>
<span>                         print_plots <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="associated_clusters_files/figure-html/unnamed-chunk-17-1.png" width="100%" style="display: block; margin: auto;"></p>
<p>In the cluster on the right, we notice that the neighborhoods for two
similar associated sequences were grouped into a single cluster.</p>
</div>
<div class="section level3">
<h3 id="labeling-the-global-clusters">Labeling the Global Clusters<a class="anchor" aria-label="anchor" href="#labeling-the-global-clusters"></a>
</h3>
<p>In order to more easily cross-reference the clusters in the visual
plot with the clusters in the data, we can label the clusters with their
ID numbers.</p>
<p>This is accomplished using the <code><a href="../reference/addClusterLabels.html">labelClusters()</a></code> function
as described <a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/articles/cluster_analysis.html#labeling-clusters">here</a>.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">all_clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addClusterLabels.html">labelClusters</a></span><span class="op">(</span><span class="va">all_clusters</span>, size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="va">all_clusters</span><span class="op">$</span><span class="va">plots</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#&gt; Warning: Removed 172 rows containing missing values (`geom_text()`).</span></span></code></pre></div>
<p><img src="associated_clusters_files/figure-html/unnamed-chunk-18-1.png" width="100%" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="focusing-on-individual-clusters">Focusing on Individual Clusters<a class="anchor" aria-label="anchor" href="#focusing-on-individual-clusters"></a>
</h3>
<p>To focus on a particular cluster, we can subset the node metadata
based on the value of <code>cluster_id</code> and use
<code><a href="../reference/buildRepSeqNetwork.html">buildNet()</a></code> to produce plots of the cluster’s graph.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># focus on the second cluster</span></span>
<span><span class="fu"><a href="../reference/buildRepSeqNetwork.html">buildNet</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">all_clusters</span><span class="op">$</span><span class="va">node_data</span><span class="op">[</span><span class="va">all_clusters</span><span class="op">$</span><span class="va">node_data</span><span class="op">$</span><span class="va">cluster_id</span> <span class="op">==</span> <span class="fl">2</span>, <span class="op">]</span>,</span>
<span>  seq_col <span class="op">=</span> <span class="st">"CloneSeq"</span>, </span>
<span>  color_nodes_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CloneSeq"</span>, <span class="st">"SampleID"</span><span class="op">)</span>, </span>
<span>  color_scheme <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"plasma"</span>, <span class="st">"turbo"</span><span class="op">)</span>,</span>
<span>  size_nodes_by <span class="op">=</span> <span class="fl">3</span>, </span>
<span>  plot_title <span class="op">=</span> <span class="st">"Cluster 2"</span>,</span>
<span>  print_plots <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="associated_clusters_files/figure-html/unnamed-chunk-19-1.png" width="100%" style="display: block; margin: auto;"><img src="associated_clusters_files/figure-html/unnamed-chunk-19-2.png" width="100%" style="display: block; margin: auto;"></p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Brian Neal, Hai Yang, Phi Long Le, Li Zhang.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
