<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="NAIR">
<title>buildRepSeqNetwork()/buildNet() â€¢ NAIR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="buildRepSeqNetwork()/buildNet()">
<meta property="og:description" content="NAIR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">NAIR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.3</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/NAIR.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/buildRepSeqNetwork.html">buildRepSeqNetwork()/buildNet()</a>
    <a class="dropdown-item" href="../articles/node_properties.html">Node-Level Network Properties</a>
    <a class="dropdown-item" href="../articles/cluster_analysis.html">Cluster Analysis</a>
    <a class="dropdown-item" href="../articles/network_visualization.html">Network Visualization</a>
    <a class="dropdown-item" href="../articles/supplementary.html">Supplementary Functions</a>
    <a class="dropdown-item" href="../articles/dual_chain.html">Dual-Chain Network Analysis</a>
    <a class="dropdown-item" href="../articles/associated_clusters.html">Searching for Associated TCR/BCR Clusters</a>
    <a class="dropdown-item" href="../articles/public_clusters.html">Searching for Public TCR/BCR Clusters</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/mlizhangx/Network-Analysis-for-Repertoire-Sequencing-/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>buildRepSeqNetwork()/buildNet()</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/mlizhangx/Network-Analysis-for-Repertoire-Sequencing-/blob/HEAD/vignettes/buildRepSeqNetwork.Rmd" class="external-link"><code>vignettes/buildRepSeqNetwork.Rmd</code></a></small>
      <div class="d-none name"><code>buildRepSeqNetwork.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>General network analysis on Adaptive Immune Receptor Repertoire
Sequencing (AIRR-Seq) data is performed using
<code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code> or its convenient alias
<code><a href="../reference/buildRepSeqNetwork.html">buildNet()</a></code>. This function does the following:</p>
<ul>
<li>Filters the AIRR-Seq data according to user specifications</li>
<li>Builds the network graph for the immune repertoire</li>
<li>Performs additional network analysis, which can include:
<ul>
<li>Cluster analysis</li>
<li>Network properties</li>
<li>Customizable visual plots of the network graph</li>
</ul>
</li>
<li>Returns (and optionally saves) the following output:
<ul>
<li>The network graph (as <code>igraph</code> and adjacency matrix)</li>
<li>Metadata for the network</li>
<li>Metadata for the nodes in the network</li>
<li>Metadata for the clusters in the network</li>
<li>Plots of the network graph</li>
</ul>
</li>
</ul>
<div class="section level5">
<h5 id="simulate-data-for-demonstration">Simulate Data for Demonstration<a class="anchor" aria-label="anchor" href="#simulate-data-for-demonstration"></a>
</h5>
<p>We simulate some toy data for demonstration.</p>
<p>We simulate data consisting of two samples with 100 observations
each, for a total of 200 observations (rows).</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">42</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/">NAIR</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Welcome to NAIR: Network Analysis of Immune Repertoire.</span></span>
<span><span class="co">#&gt; Get started using `vignette("NAIR")`, or by visiting</span></span>
<span><span class="co">#&gt; https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/</span></span>
<span><span class="va">dir_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">toy_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulateToyData.html">simulateToyData</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">toy_data</span><span class="op">)</span></span>
<span><span class="co">#&gt;        CloneSeq CloneFrequency CloneCount SampleID</span></span>
<span><span class="co">#&gt; 1 TTGAGGAAATTCG    0.007873775       3095  Sample1</span></span>
<span><span class="co">#&gt; 2 GGAGATGAATCGG    0.007777102       3057  Sample1</span></span>
<span><span class="co">#&gt; 3 GTCGGGTAATTGG    0.009094910       3575  Sample1</span></span>
<span><span class="co">#&gt; 4 GCCGGGTAATTCG    0.010160859       3994  Sample1</span></span>
<span><span class="co">#&gt; 5 GAAAGAGAATTCG    0.009336593       3670  Sample1</span></span>
<span><span class="co">#&gt; 6 AGGTGGGAATTCG    0.010369470       4076  Sample1</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">toy_data</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 200</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="required-arguments">
<a name="mainargs"></a> Required Arguments<a class="anchor" aria-label="anchor" href="#required-arguments"></a>
</h2>
<ul>
<li>The first parameter <code>data</code> accepts a data frame
containing the AIRR-seq data, where each row corresponds to a single
TCR/BCR clone (bulk data) or cell (single-cell data).</li>
<li>The second parameter <code>seq_col</code> specifies the data column
containing the receptor sequences to be used as the basis of similarity
between two cells or clones. It accepts the column name as a character
string or the column position index.</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">net</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/buildRepSeqNetwork.html">buildNet</a></span><span class="op">(</span><span class="va">toy_data</span>, <span class="st">"CloneSeq"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 class="tabset" id="filtering-the-input-data">
<a name="input_options"></a> Filtering the Input
Data<a class="anchor" aria-label="anchor" href="#filtering-the-input-data"></a>
</h2>
<p>The following options are useful for removing noise or irrelevant
data from the analysis, potentially improving the quality of the network
graph and downstream analysis results.</p>
<div class="section level5">
<h5 id="sequence-length">Sequence Length<a class="anchor" aria-label="anchor" href="#sequence-length"></a>
</h5>
<p>The <code>min_seq_length</code> parameter specifies the minimum
TCR/BCR sequence length (in number of characters). Sequences with
lengths below this value are dropped prior to analysis. An argument
value of <code>NULL</code> bypasses this filter. This filter applies to
the sequence in the column specified by <a href="#mainargs"><code>seq_col</code></a>.</p>
<p>By default, <code>min_seq_length</code> is set to 3.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">net</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/buildRepSeqNetwork.html">buildNet</a></span><span class="op">(</span><span class="va">toy_data</span>, <span class="st">"CloneSeq"</span>, min_seq_length <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="sequence-content">Sequence Content<a class="anchor" aria-label="anchor" href="#sequence-content"></a>
</h5>
<p>The optional <code>drop_matches</code> parameter can be used to
remove TCR/BCR sequences matching a specified pattern. It accepts a
character string containing a regular expression specifying the pattern
of content to search for. Each sequence in the column specified by <a href="#mainargs"><code>seq_col</code></a> is checked for a pattern match
using <code><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep()</a></code>. If a match is found, the sequence is removed
prior to analysis.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">net</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/buildRepSeqNetwork.html">buildNet</a></span><span class="op">(</span><span class="va">toy_data</span>, <span class="st">"CloneSeq"</span>, drop_matches <span class="op">=</span> <span class="st">"\\W"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 class="tabset" id="network-settings">
<a name="network_settings"></a> Network Settings<a class="anchor" aria-label="anchor" href="#network-settings"></a>
</h2>
<p>The following settings pertain to the network construction.</p>
<div class="section level5">
<h5 id="distance-function">Distance Function<a class="anchor" aria-label="anchor" href="#distance-function"></a>
</h5>
<p>The default method for quantifying the similarity between TCR/BCR
sequences is the Hamming distance, which is computed using
<code><a href="../reference/hamDistBounded.html">hamDistBounded()</a></code>. It calculates the number of differences
between two sequences of the same length. For each position in one
sequence, the character in that position is checked to see whether it
differs from the character in the same position of the other sequence.
If the sequences have different lengths, the shorter sequence is
extended by adding non-matching characters to make it the same length as
the longer sequence.</p>
<p>The Levenshtein distance, which is computed using
<code><a href="../reference/levDistBounded.html">levDistBounded()</a></code>, can be used as an alternative measurement
to determine the similarity between sequences. It calculates the minimum
number of single-character edits (insertions, deletions and
transformations) needed to transform one sequence into the other. This
method is particularly useful for comparing sequences of different
lengths and can account for insertions and deletions. When constructing
a network based on the similarity of CDR-3 nucleotide sequences, it is
preferable to use the Levenshtein distance instead of the default
Hamming distance by specifying <code>dist_type = "lev"</code>. However,
the Levenshtein distance requires significantly more computation than
the Hamming distance, which may be challenging when working with large
data sets having long TCR/BCR sequences.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">net</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/buildRepSeqNetwork.html">buildNet</a></span><span class="op">(</span><span class="va">toy_data</span>, <span class="st">"CloneSeq"</span>, dist_type <span class="op">=</span> <span class="st">"lev"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="distance-cutoff">Distance Cutoff<a class="anchor" aria-label="anchor" href="#distance-cutoff"></a>
</h5>
<p>The function specified by <code>dist_type</code> quantifies the
similarity between TCR/BCR sequences as a nonnegative distance, with
values closer to 0 indicating greater similarity.</p>
<p>Each node in the network graph corresponds to a row of the AIRR-Seq
data. By default, two nodes are connected by an edge if the distance
between their TCR/BCR sequences does not exceed 1. This cutoff value is
specified by the <code>dist_cutoff</code> parameter, which controls the
stringency of the network construction and affects the number and
density of edges in the network. A lower cutoff requires greater
sequence similarity to form an edge connection. If
<code>dist_cutoff = 0</code>, two sequences must be identical in order
for their nodes to be joined by an edge.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">net</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/buildRepSeqNetwork.html">buildNet</a></span><span class="op">(</span><span class="va">toy_data</span>, <span class="st">"CloneSeq"</span>, dist_cutoff <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="keepremove-isolated-nodes">
<a name="isolated_nodes"></a> Keep/Remove Isolated Nodes<a class="anchor" aria-label="anchor" href="#keepremove-isolated-nodes"></a>
</h5>
<p>By default, only nodes that are joined by an edge connection to at
least one other node will be kept in the network.</p>
<p>If <code>drop_isolated_nodes = FALSE</code>, then all nodes are kept
in the network, including those that do not have any edge connections to
other nodes.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">net</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/buildRepSeqNetwork.html">buildNet</a></span><span class="op">(</span><span class="va">toy_data</span>, <span class="st">"CloneSeq"</span>, drop_isolated_nodes <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="network-properties-and-cluster-analysis">
<a name="network_properties"></a> Network Properties and Cluster
Analysis<a class="anchor" aria-label="anchor" href="#network-properties-and-cluster-analysis"></a>
</h2>
<p><code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code> can perform additional analysis
after constructing the network, including cluster analysis (partitioning
the network graph into densely-connected subgraphs) and computation of
network properties (which describe the structural organization of the
network).</p>
<div class="section level4">
<h4 class="tabset" id="node-level-network-properties">
<a name="node_properties"></a> Node-Level Network
Properties<a class="anchor" aria-label="anchor" href="#node-level-network-properties"></a>
</h4>
<p>Node-level network properties pertain to individual nodes in the
network graph.</p>
<p>Some are local properties, meaning that their value for a given node
depends only on a subset of the nodes in the network. One example is the
network degree of a given node, which represents the number of other
nodes that are directly joined to the given node by an edge
connection.</p>
<p>Other properties are global properties, meaning that their value for
a given node depends on all of the nodes in the network. An example is
the authority score of a node, which is computed using the entire graph
adjacency matrix (if we denote this matrix by <span class="math inline">\(A\)</span>, then the principal eigenvector of
<span class="math inline">\(A^T A\)</span> represents the authority
scores of the network nodes).</p>
<p>Node-level network properties can be computed when calling
<code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code> by setting
<code>node_stats = TRUE</code>, or as a separate step using
<code><a href="../reference/addNodeStats.html">addNodeStats()</a></code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">net</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/buildRepSeqNetwork.html">buildNet</a></span><span class="op">(</span><span class="va">toy_data</span>, <span class="st">"CloneSeq"</span>, node_stats <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>See <a href="node_properties.html">this vignette</a> for more details
on computing node-level network properties.</p>
</div>
<div class="section level4">
<h4 class="tabset" id="cluster-analysis">
<a name="cluster_analysis"></a> Cluster Analysis<a class="anchor" aria-label="anchor" href="#cluster-analysis"></a>
</h4>
<p>Cluster analysis uses a community-finding algorithm to partition the
network graph into clusters (densely-connected subgraphs). These
clusters represent groups of clones/cells with similar receptor
sequences.</p>
<p>Cluster analysis can be performed when calling
<code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code> by setting
<code>cluster_stats = TRUE</code> or as a separate step using
<code><a href="../reference/addClusterStats.html">addClusterStats()</a></code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">net</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/buildRepSeqNetwork.html">buildNet</a></span><span class="op">(</span><span class="va">toy_data</span>, <span class="st">"CloneSeq"</span>, cluster_stats <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>The cluster membership of each node is recorded as a variable in the
node metadata. Cluster properties, such as node count and mean sequence
length, are included in their own data frame within the output.</p>
<p>See <a href="cluster_analysis.html">this vignette</a> for more
details.</p>
</div>
</div>
<div class="section level2">
<h2 id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a>
</h2>
<p>Customized plots can be produced when calling
<code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code> or created afterward using
<code><a href="../reference/addPlots.html">addPlots()</a></code>.</p>
<p>Specify <code>print_plots = TRUE</code> to print the plots to the R
plotting window.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">net</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/buildRepSeqNetwork.html">buildNet</a></span><span class="op">(</span><span class="va">toy_data</span>, <span class="st">"CloneSeq"</span>,</span>
<span>                node_stats <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                color_nodes_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SampleID"</span>, <span class="st">"transitivity"</span>, <span class="st">"coreness"</span><span class="op">)</span>,</span>
<span>                color_scheme <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"default"</span>, <span class="st">"plasma-1"</span>, <span class="st">"mako-1"</span><span class="op">)</span>,</span>
<span>                color_title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">""</span>, <span class="st">"Transitivity"</span>, <span class="st">"Coreness"</span><span class="op">)</span>,</span>
<span>                size_nodes_by <span class="op">=</span> <span class="st">"degree"</span>,</span>
<span>                node_size_limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">1.5</span><span class="op">)</span>,</span>
<span>                plot_title <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                plot_subtitle <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                print_plots <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="buildRepSeqNetwork_files/figure-html/unnamed-chunk-11-1.png" width="100%" style="display: block; margin: auto;"><img src="buildRepSeqNetwork_files/figure-html/unnamed-chunk-11-2.png" width="100%" style="display: block; margin: auto;"><img src="buildRepSeqNetwork_files/figure-html/unnamed-chunk-11-3.png" width="100%" style="display: block; margin: auto;"></p>
<p>See <a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/articles/network_visualization.html">this
article</a> for more on how to customize the visualizations.</p>
</div>
<div class="section level2">
<h2 class="tabset" id="output">
<a name="base_output"></a> Output<a class="anchor" aria-label="anchor" href="#output"></a>
</h2>
<p>The function returns a list with the following elements:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">net</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "details"          "igraph"           "adjacency_matrix" "node_data"       </span></span>
<span><span class="co">#&gt; [5] "plots"</span></span></code></pre></div>
<p>We describe each element below.</p>
<div class="section level5">
<h5 id="network-metadata">Network Metadata<a class="anchor" aria-label="anchor" href="#network-metadata"></a>
</h5>
<p><code>details</code> records argument values supplied in the call to
<code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code>:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">net</span><span class="op">$</span><span class="va">details</span></span>
<span><span class="co">#&gt; $seq_col</span></span>
<span><span class="co">#&gt; [1] "CloneSeq"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $dist_type</span></span>
<span><span class="co">#&gt; [1] "hamming"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $dist_cutoff</span></span>
<span><span class="co">#&gt; [1] 1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $drop_isolated_nodes</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $nodes_in_network</span></span>
<span><span class="co">#&gt; [1] 122</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $min_seq_length</span></span>
<span><span class="co">#&gt; [1] 3</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $drop_matches</span></span>
<span><span class="co">#&gt; [1] "NULL"</span></span></code></pre></div>
<p><code>nodes_in_network</code> records the number of nodes in the
network. Additional network information is included when <a href="#cluster_analysis">cluster analysis</a> is performed.</p>
</div>
<div class="section level5">
<h5 id="node-metadata">Node Metadata<a class="anchor" aria-label="anchor" href="#node-metadata"></a>
</h5>
<p><code>node_data</code> is a data frame containing metadata for the
network nodes, where each row corresponds to a node in the network
graph:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">net</span><span class="op">$</span><span class="va">node_data</span><span class="op">)</span></span>
<span><span class="co">#&gt;         CloneSeq CloneFrequency CloneCount SampleID degree transitivity</span></span>
<span><span class="co">#&gt; 2  GGAGATGAATCGG    0.007777102       3057  Sample1      1          NaN</span></span>
<span><span class="co">#&gt; 5  GAAAGAGAATTCG    0.009336593       3670  Sample1      3    0.3333333</span></span>
<span><span class="co">#&gt; 8  GGGGAGAAATTGG    0.006220155       2445  Sample1      2    1.0000000</span></span>
<span><span class="co">#&gt; 11 GGGGGAGAATTGC    0.012969469       5098  Sample1      4    0.6666667</span></span>
<span><span class="co">#&gt; 12 GGGGGGGAATTGC    0.009079646       3569  Sample1     10    0.3555556</span></span>
<span><span class="co">#&gt; 13 AGGGGGAAATTGG    0.014941093       5873  Sample1      5    0.1000000</span></span>
<span><span class="co">#&gt;    eigen_centrality centrality_by_eigen betweenness centrality_by_betweenness</span></span>
<span><span class="co">#&gt; 2        0.00000000        0.000000e+00    0.000000                  0.000000</span></span>
<span><span class="co">#&gt; 5        0.00000000        1.702099e-17  108.000000                108.000000</span></span>
<span><span class="co">#&gt; 8        0.04558649        4.558649e-02    0.000000                  0.000000</span></span>
<span><span class="co">#&gt; 11       0.15055366        1.505537e-01    1.549451                  1.549451</span></span>
<span><span class="co">#&gt; 12       0.52691798        5.269180e-01   73.970918                 73.970918</span></span>
<span><span class="co">#&gt; 13       0.14682343        1.468234e-01   75.439560                 75.439560</span></span>
<span><span class="co">#&gt;    authority_score coreness   page_rank</span></span>
<span><span class="co">#&gt; 2       0.00000000        1 0.008196721</span></span>
<span><span class="co">#&gt; 5       0.00000000        2 0.010578507</span></span>
<span><span class="co">#&gt; 8       0.04558649        2 0.003936684</span></span>
<span><span class="co">#&gt; 11      0.15055366        4 0.005034736</span></span>
<span><span class="co">#&gt; 12      0.52691798        6 0.011491588</span></span>
<span><span class="co">#&gt; 13      0.14682343        3 0.008703523</span></span></code></pre></div>
<p><a name="subset_cols"></a></p>
<p>Since <code><a href="../reference/buildRepSeqNetwork.html">buildNet()</a></code> was called with <a href="#isolated_nodes"><code>drop_isolated_nodes = TRUE</code></a>, some
of the original rows are missing. The original row names are preserved,
facilitating cross-referencing with the original data.</p>
<p>By default, all variables from the original input data are present.
To include only a subset of the original variables, specify those to
keep using the <code>subset_cols</code> parameter, which accepts a
character vector of column names or a vector of column indices. The
sequence column is always included.</p>
<p>Variables for <a href="#node_properties">node-level network
properties</a> will also be present if computed, as seen <a href="node_properties.html#results">here</a>.</p>
<p>Variables present in the node metadata can be used to <a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/articles/network_visualization.html#node-colors">encode
node colors</a> in visualizations.</p>
</div>
<div class="section level5">
<h5 id="cluster-metadata">Cluster Metadata<a class="anchor" aria-label="anchor" href="#cluster-metadata"></a>
</h5>
<p>If <code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code> is called with <a href="#cluster_analysis"><code>cluster_stats = TRUE</code></a>, the
returned list will contain a data frame <code>cluster_data</code> with
cluster properties such as node count and mean sequence length, as seen
<a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/articles/cluster_analysis.html#cluster-properties">here</a>.</p>
</div>
<div class="section level5">
<h5 id="visual-plots">Visual Plots<a class="anchor" aria-label="anchor" href="#visual-plots"></a>
</h5>
<p><code>plots</code> is a list containing any plots created as well as
a matrix <code>graph_layout</code> storing the node layout used in the
plots. Each plot is named according to the variable used to color the
nodes.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">net</span><span class="op">$</span><span class="va">plots</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "SampleID"     "transitivity" "coreness"     "graph_layout"</span></span></code></pre></div>
<p>Each plot is a <a href="https://ggraph.data-imaginist.com/" class="external-link"><code>ggraph</code></a> (a
special kind of <a href="https://ggplot2.tidyverse.org/" class="external-link"><code>ggplot</code></a>).</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">net</span><span class="op">$</span><span class="va">plots</span><span class="op">$</span><span class="va">uniform_color</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "NULL"</span></span></code></pre></div>
<p>Custom plots can be created when calling <code><a href="../reference/buildRepSeqNetwork.html">buildNet()</a></code> or
afterward using <code><a href="../reference/addPlots.html">addPlots()</a></code> as described <a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/articles/network_visualization.html">here</a>.</p>
</div>
<div class="section level5">
<h5 id="network-graph-and-adjacency-matrix">Network Graph and Adjacency Matrix<a class="anchor" aria-label="anchor" href="#network-graph-and-adjacency-matrix"></a>
</h5>
<p><code>igraph</code> is an object of class <a href="https://r.igraph.org" class="external-link"><code>igraph</code></a> and
<code>adjacency_matrix</code> is a <a href="https://www.rdocumentation.org/packages/Matrix/versions/1.5-4/topics/dgCMatrix-class" class="external-link"><code>dgCMatrix</code></a>.
Both objects encode the nodes and edges of the network graph, each using
a different format. These objects are used by other <code>NAIR</code>
functions, and are typically not of direct importance to the user.</p>
</div>
</div>
<div class="section level2">
<h2 class="tabset" id="saving-output">
<a name="output_settings"></a> Saving Output<a class="anchor" aria-label="anchor" href="#saving-output"></a>
</h2>
<div class="section level5">
<h5 id="output-directory">Output Directory<a class="anchor" aria-label="anchor" href="#output-directory"></a>
</h5>
<p><code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code> writes its output to files if a
directory path is provided to the <code>output_dir</code> parameter. The
specified output directory will be created if it does not already
exist.</p>
</div>
<div class="section level5">
<h5 id="output-file-format">
<a name="output_format"></a> Output File Format<a class="anchor" aria-label="anchor" href="#output-file-format"></a>
</h5>
<ul>
<li>By default, the list returned by <code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code>
is saved as a compressed RDS file.</li>
<li>If the file will be transferred across machines, the RData format is
preferred. It can be used by specifying
<code>output_type = "rda"</code>. The list will be named
<code>net</code> in the R environment.</li>
<li>To use the results outside of R, specify
<code>output_type = "individual"</code>. Each list element will be saved
as a separate file using the following formats:
<ul>
<li>
<code>.csv</code> for <code>node_data</code> and
<code>cluster_data</code> (using <code><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.csv()</a></code>, with
<code>row.names = FALSE</code> for <code>cluster_data</code>)</li>
<li>
<code>.mtx</code> for <code>adjacency_matrix</code>
</li>
<li>
<code>.txt</code> for <code>details</code>, <code>igraph</code> and
<code>plots$graph_layout</code>
</li>
<li>
<code>.rda</code> (RData) for <code>plots</code>
</li>
</ul>
</li>
<li>Regardless of <code>output_type</code>, plots are printed to a pdf
file containing one plot per page. The dimensions (in inches) for each
page can be adjusted using the <code>plot_width</code> and
<code>plot_height</code> parameters, with the defaults being
<code>12</code> and <code>10</code>, respectively.</li>
</ul>
</div>
<div class="section level5">
<h5 id="output-file-names">
<a name="output_filenames"></a> Output File Name(s)<a class="anchor" aria-label="anchor" href="#output-file-names"></a>
</h5>
<p>By default, the name of each saved file begins with
<code>MyRepSeqNetwork</code>. This prefix can be changed using the
<code>output_name</code> parameter, which accepts a character
string.</p>
<p>When <code>output_type</code> is <code>"rds"</code> or
<code>"rda"</code>, only two files are saved (the data file and the
pdf). For each file, <code>output_name</code> is followed by the
appropriate file extension. For example, if
<code>output_name = "NetworkABC"</code> and
<code>output_type = "rds"</code>, then the files will be named
<code>NetworkABC.rds</code> and <code>NetworkABC.pdf</code>.</p>
<ul>
<li>If <code>output_type = "individual"</code>, the prefix specified by
<code>output_name</code> is followed by:
<ul>
<li>
<code>_Details.txt</code> for <code>details</code>
</li>
<li>
<code>_EdgeList.txt</code> for <code>igraph</code>
</li>
<li>
<code>_AdjacencyMatrix.mtx</code> for
<code>adjacency_matrix</code>
</li>
<li>
<code>_NodeMetadata.csv</code> for <code>node_data</code>
</li>
<li>
<code>_ClusterMetadata.csv</code> for <code>cluster_data</code> (if
present)</li>
<li>
<code>_Plots.rda</code> for the list <code>plots</code>
</li>
<li>
<code>.pdf</code> for the pdf of the plots</li>
<li>
<code>_GraphLayout.txt</code> for
<code>plots$graph_layout</code>
</li>
</ul>
</li>
</ul>
<p>For example, if <code>output_name = "NetworkABC"</code> and
<code>output_type = "individual"</code>, then the node metadata is saved
as <code>NetworkABC_NodeMetadata.csv</code>, the pdf is saved as
<code>NetworkABC.pdf</code>, the list <code>plots</code> is saved as
<code>NetworkABC_Plots.rds</code>, and so on.</p>
</div>
</div>
<div class="section level2">
<h2 class="tabset" id="supplementary-functions">
<a name="downstream"></a> Supplementary
Functions<a class="anchor" aria-label="anchor" href="#supplementary-functions"></a>
</h2>
<p>The <code>NAIR</code> package contains a number of functions
supplementary to <code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code> that can be used to
perform additional downstream tasks. See <a href="supplementary.html">this vignette</a>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span> <span class="co"># For pipe operator (%&gt;%)</span></span>
<span><span class="va">toy_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/filterInputData.html">filterInputData</a></span><span class="op">(</span><span class="st">"CloneSeq"</span>, drop_matches <span class="op">=</span> <span class="st">"\\W"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/buildRepSeqNetwork.html">buildNet</a></span><span class="op">(</span><span class="st">"CloneSeq"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/addNodeStats.html">addNodeStats</a></span><span class="op">(</span><span class="st">"all"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/addClusterMembership.html">addClusterMembership</a></span><span class="op">(</span><span class="st">"greedy"</span>, cluster_id_name <span class="op">=</span> <span class="st">"cluster_greedy"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/addClusterMembership.html">addClusterMembership</a></span><span class="op">(</span><span class="st">"leiden"</span>, cluster_id_name <span class="op">=</span> <span class="st">"cluster_leiden"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/addClusterStats.html">addClusterStats</a></span><span class="op">(</span><span class="st">"cluster_leiden"</span>, <span class="st">"CloneSeq"</span>, <span class="st">"CloneCount"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/addPlots.html">addPlots</a></span><span class="op">(</span>color_nodes_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cluster_leiden"</span>, <span class="st">"cluster_greedy"</span><span class="op">)</span>, </span>
<span>           color_scheme <span class="op">=</span> <span class="st">"Viridis"</span></span>
<span>  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/addClusterLabels.html">labelClusters</a></span><span class="op">(</span><span class="st">"cluster_leiden"</span>, cluster_id_col <span class="op">=</span> <span class="st">"cluster_leiden"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/addClusterLabels.html">labelClusters</a></span><span class="op">(</span><span class="st">"cluster_greedy"</span>, cluster_id_col <span class="op">=</span> <span class="st">"cluster_greedy"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/saveNetwork.html">saveNetwork</a></span><span class="op">(</span>output_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, output_name <span class="op">=</span> <span class="st">"my_network"</span><span class="op">)</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Brian Neal, Hai Yang, Phi Long Le, Li Zhang.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
