<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="NAIR">
<title>buildRepSeqNetwork • NAIR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="buildRepSeqNetwork">
<meta property="og:description" content="NAIR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">NAIR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.9043</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/NAIR.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/associated_clusters.html">Searching for Associated TCR/BCR Clusters</a>
    <a class="dropdown-item" href="../articles/buildRepSeqNetwork.html">buildRepSeqNetwork</a>
    <a class="dropdown-item" href="../articles/dual_chain.html">Dual-Chain Network Analysis</a>
    <a class="dropdown-item" href="../articles/network_visualization.html">Network Visualization</a>
    <a class="dropdown-item" href="../articles/public_clusters.html">Searching for Public TCR/BCR Clusters</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/mlizhangx/Network-Analysis-for-Repertoire-Sequencing-/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>buildRepSeqNetwork</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/mlizhangx/Network-Analysis-for-Repertoire-Sequencing-/blob/HEAD/vignettes/buildRepSeqNetwork.Rmd" class="external-link"><code>vignettes/buildRepSeqNetwork.Rmd</code></a></small>
      <div class="d-none name"><code>buildRepSeqNetwork.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>General network analysis on Adaptive Immune Receptor Repertoire
Sequencing (AIRR-Seq) data is performed using the
<code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code> function. This function does the
following:</p>
<ul>
<li>Filters the input AIRR-Seq data according to user
specifications</li>
<li>Builds the network graph for the immune repertoire</li>
<li>Performs additional network analysis, which can include:
<ul>
<li>Cluster analysis</li>
<li>Computation of network properties</li>
</ul>
</li>
<li>Generates customizable visual plots of the network graph</li>
<li>Saves and returns the following output:
<ul>
<li>Metadata for the nodes in the network</li>
<li>Metadata for the clusters in the network</li>
<li>The plots of the network graph</li>
<li>The network graph itself, both as an adjacency matrix and as an
<code>igraph</code> object</li>
</ul>
</li>
</ul>
<p>In this vignette we demonstrate the function’s usage and output.</p>
<div class="section level3">
<h3 id="simulate-data-for-demonstration">Simulate Data for Demonstration<a class="anchor" aria-label="anchor" href="#simulate-data-for-demonstration"></a>
</h3>
<p>We simulate some toy data for demonstration.</p>
<p>We simulate data consisting of two samples with 100 observations
each, for a total of 200 observations (rows).</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/">NAIR</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Welcome to the NAIR package. Get started using `vignette("NAIR", package = "NAIR")` or by visiting https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/</span></span>
<span><span class="va">dir_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">toy_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulateToyData.html">simulateToyData</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">toy_data</span><span class="op">)</span></span>
<span><span class="co">#&gt;        CloneSeq CloneFrequency CloneCount SampleID</span></span>
<span><span class="co">#&gt; 1 TTGAGGAAATTCG    0.007873775       3095  Sample1</span></span>
<span><span class="co">#&gt; 2 GGAGATGAATCGG    0.007777102       3057  Sample1</span></span>
<span><span class="co">#&gt; 3 GTCGGGTAATTGG    0.009094910       3575  Sample1</span></span>
<span><span class="co">#&gt; 4 GCCGGGTAATTCG    0.010160859       3994  Sample1</span></span>
<span><span class="co">#&gt; 5 GAAAGAGAATTCG    0.009336593       3670  Sample1</span></span>
<span><span class="co">#&gt; 6 AGGTGGGAATTCG    0.010369470       4076  Sample1</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">toy_data</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 200</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="primary-inputs">
<a name="mainargs"></a> Primary Inputs<a class="anchor" aria-label="anchor" href="#primary-inputs"></a>
</h2>
<p>The <code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code> function has two required
arguments for specifying the input data.</p>
<ul>
<li>The function’s first argument <code>data</code> accepts a data frame
containing the AIRR-seq data, where each row corresponds to a single
TCR/BCR clone (bulk data) or cell (single-cell data).</li>
<li>The second argument <code>seq_col</code> specifies the data column
containing the receptor sequences to be used as the basis of similarity
between two cells or clones. This argument accepts either the column
name (e.g., <code>seq_col = "CloneSeq"</code>) or the column index
(e.g., <code>seq_col = 1</code>, denoting the first column).</li>
</ul>
<p><em>For single-cell data, see <a href="dual_chain.html">this
vignette</a> on building networks based on joint similarity in both
alpha chain sequences and beta chain sequences.</em></p>
<p>Below, we execute the <code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code> function. A
visual plot of the network graph is automatically printed to the
plotting window in R.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">output</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork</a></span><span class="op">(</span><span class="va">toy_data</span>, seq_col <span class="op">=</span> <span class="st">"CloneSeq"</span>, </span>
<span>                             output_dir <span class="op">=</span> <span class="va">dir_out</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Input data contains 200 rows.</span></span>
<span><span class="co">#&gt; Removing sequences with length fewer than 3 characters... Done. 200 rows remaining.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 122 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Generating graph plot...</span></span>
<span><span class="co">#&gt;  Done.</span></span>
<span><span class="co">#&gt; Node-level meta-data saved to file:</span></span>
<span><span class="co">#&gt;   /tmp/Rtmpu6X8US/MyRepSeqNetwork_NodeMetadata.csv</span></span></code></pre></div>
<p><img src="buildRepSeqNetwork_files/figure-html/unnamed-chunk-3-1.png" width="100%" style="display: block; margin: auto;"></p>
<pre><code><span><span class="co">#&gt; Network graph plots saved to file:</span></span>
<span><span class="co">#&gt;   /tmp/Rtmpu6X8US/MyRepSeqNetwork.pdf</span></span>
<span><span class="co">#&gt; Network igraph saved in edgelist format to file:</span></span>
<span><span class="co">#&gt;   /tmp/Rtmpu6X8US/MyRepSeqNetwork_EdgeList.txt</span></span>
<span><span class="co">#&gt; Adjacency matrix saved to file:</span></span>
<span><span class="co">#&gt;   /tmp/Rtmpu6X8US/MyRepSeqNetwork_AdjacencyMatrix.mtx</span></span></code></pre>
</div>
<div class="section level2">
<h2 class="tabset" id="base-output">
<a name="base_output"></a> Base Output<a class="anchor" aria-label="anchor" href="#base-output"></a>
</h2>
<p>When no optional arguments are used, the
<code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code> function returns a list containing the
following elements:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">output</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "igraph"           "adjacency_matrix" "node_data"        "plots"</span></span></code></pre></div>
<p>We describe each of these list elements below.</p>
<div class="section level3">
<h3 id="node_data">
<code>node_data</code><a class="anchor" aria-label="anchor" href="#node_data"></a>
</h3>
<p>The list element <code>node_data</code> is a data frame containing
metadata for the network nodes, where each row corresponds to a node in
the network graph:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">output</span><span class="op">$</span><span class="va">node_data</span><span class="op">)</span></span>
<span><span class="co">#&gt;         CloneSeq CloneFrequency CloneCount SampleID</span></span>
<span><span class="co">#&gt; 2  GGAGATGAATCGG    0.007777102       3057  Sample1</span></span>
<span><span class="co">#&gt; 5  GAAAGAGAATTCG    0.009336593       3670  Sample1</span></span>
<span><span class="co">#&gt; 8  GGGGAGAAATTGG    0.006220155       2445  Sample1</span></span>
<span><span class="co">#&gt; 11 GGGGGAGAATTGC    0.012969469       5098  Sample1</span></span>
<span><span class="co">#&gt; 12 GGGGGGGAATTGC    0.009079646       3569  Sample1</span></span>
<span><span class="co">#&gt; 13 AGGGGGAAATTGG    0.014941093       5873  Sample1</span></span></code></pre></div>
<p>By default, all variables (data columns) from the original input
AIRR-seq data are present in the returned node data. <a href="#input_options">This can be changed using the
<code>subset_cols</code> argument</a>.</p>
<p>The row names from the original input data are preserved in the
returned node data. This allows the user to match rows of the input data
with their corresponding rows in the node data.</p>
<p>Notice from the row names that some rows from the original input data
are missing in the returned node data. This is because by default, only
those nodes that are connected to at least one other node by an edge
will be kept in the network. <a href="#network_settings">This can be
changed using the <code>drop_isolated_nodes</code> argument</a>.</p>
</div>
<div class="section level3">
<h3 id="plots">
<code>plots</code><a class="anchor" aria-label="anchor" href="#plots"></a>
</h3>
<p>The list element <code>plots</code> is a list containing each of the
plot(s) created by <code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code>. Each plot is an
object of class <code>ggraph</code> (which inherits from the
<code>ggplot</code> class), and is named according to the variable used
to color the nodes. Here the nodes are not colored according to any
variable, so the name of the plot is <code>uniform_color</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">output</span><span class="op">$</span><span class="va">plots</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "uniform_color"</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">output</span><span class="op">$</span><span class="va">plots</span><span class="op">$</span><span class="va">uniform_color</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "ggraph" "gg"     "ggplot"</span></span></code></pre></div>
<p>Plots can be customized through <a href="network_visualization.html">optional arguments</a> to
<code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code> as well as by <a href="network_visualization.html#modifying-plots-after-creation">modifying
them</a> after they are created.</p>
</div>
<div class="section level3">
<h3 id="igraph">
<code>igraph</code><a class="anchor" aria-label="anchor" href="#igraph"></a>
</h3>
<p>The list element <code>igraph</code> is an object of class
<code>igraph</code> from the <a href="https://r.igraph.org" class="external-link">igraph
package</a>. This object contains the list of nodes and edges that
comprise the network graph.</p>
<p>The <code>igraph</code> object can be used for <a href="#downstream">downstream computation of network properties</a>, as
an alternative to <a href="#network_properties">computing network
properties when calling <code>buildRepSeqNetwork()</code></a>. It can
also be used to <a href="network_visualization.html#new_plots">generate
additional plots of the network graph</a>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">output</span><span class="op">$</span><span class="va">igraph</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "igraph"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="adjacency_matrix">
<code>adjacency_matrix</code><a class="anchor" aria-label="anchor" href="#adjacency_matrix"></a>
</h3>
<p>The list element <code>adjacency_matrix</code> is an object of class
<code>dgCMatrix</code>, which is a sparse matrix format from the
<code>Matrix</code> package.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">output</span><span class="op">$</span><span class="va">adjacency_matrix</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "dgCMatrix"</span></span>
<span><span class="co">#&gt; attr(,"package")</span></span>
<span><span class="co">#&gt; [1] "Matrix"</span></span></code></pre></div>
<p>The network adjacency matrix encodes the graph (i.e., the set of
nodes and edges) as a matrix. It is symmetric, with each row
(equivalently, each column) corresponding to a distinct network node.
The network nodes follow the same order in the rows as they do in the
columns. Each column name is the receptor sequence corresponding to the
node for that column. Each row name is the row index of the original
input data that corresponds to the node for that row.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># First 5 rows and columns of the network adjacency matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">output</span><span class="op">$</span><span class="va">adjacency_matrix</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; 5 x 5 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">#&gt;    GGAGATGAATCGG GAAAGAGAATTCG GGGGAGAAATTGG GGGGGAGAATTGC GGGGGGGAATTGC</span></span>
<span><span class="co">#&gt; 2              1             .             .             .             .</span></span>
<span><span class="co">#&gt; 5              .             1             .             .             .</span></span>
<span><span class="co">#&gt; 8              .             .             1             .             .</span></span>
<span><span class="co">#&gt; 11             .             .             .             1             1</span></span>
<span><span class="co">#&gt; 12             .             .             .             1             1</span></span></code></pre></div>
<p>Each matrix entry has a value of 0 or 1, with 1 indicating an edge
connecting the node for that row to the node for that column, and 0
indicating no edge connection. Although the diagonal entries are 1,
their values are not meaningful, since an edge is not drawn from a given
node to itself.</p>
<p>Note that the rows (nodes) in the adjacency matrix correspond to the
rows (nodes) in the <code>node_data</code> data frame. Both have the
same number of rows and the same row names.</p>
<p>The adjacency matrix can be used for <a href="#downstream">downstream
computation of network properties</a>, as an alternative to <a href="#network_properties">computing network properties when calling
<code>buildRepSeqNetwork()</code></a>.</p>
</div>
</div>
<div class="section level2">
<h2 class="tabset" id="filtering-the-input-data">
<a name="input_options"></a> Filtering the Input
Data<a class="anchor" aria-label="anchor" href="#filtering-the-input-data"></a>
</h2>
<p>Any rows of the input data having a value of <code>NA</code> for
their receptor sequence (the value in the column specified by the
<code>seq_col</code> argument) will be automatically removed prior to
building the network.</p>
<p>The following options are useful for removing noise or irrelevant
data from the analysis, potentially improving the quality of the network
graph and downstream analysis results.</p>
<div class="section level3">
<h3 id="min_seq_length">
<code>min_seq_length</code><a class="anchor" aria-label="anchor" href="#min_seq_length"></a>
</h3>
<p>This parameter is used to filter out TCR/BCR sequences below a
certain length. The argument accepts a numeric value that specifies the
minimum sequence length. Every data row whose sequence length is below
this value will be dropped before constructing the network graph. The
receptor sequence is the value in the column specified by the <a href="#mainargs"><code>seq_col</code></a> argument.</p>
<p>By default, the minimum sequence length is set to 3.</p>
<p>This check can be bypassed by setting
<code>min_seq_length = NULL</code>.</p>
</div>
<div class="section level3">
<h3 id="drop_matches">
<code>drop_matches</code><a class="anchor" aria-label="anchor" href="#drop_matches"></a>
</h3>
<p>This parameter can be optionally used to filter out TCR/BCR sequences
with content matching a certain pattern. The argument accepts a
character string containing a regular expression that specifies the
pattern of content to search for (see <code><a href="https://rdrr.io/r/base/regex.html" class="external-link">?regex</a></code>). Each TCR/BCR
sequence is checked for a match with the specified content pattern.
Every data row whose sequence matches the pattern is dropped before
constructing the network graph.</p>
<p>The receptor sequence is the value in the column specified by the <a href="#mainargs"><code>seq_col</code></a> argument. It is checked for a
match with the specified content pattern using <code><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep()</a></code>.</p>
</div>
</div>
<div class="section level2">
<h2 class="tabset" id="network-settings">
<a name="network_settings"></a> Network Settings<a class="anchor" aria-label="anchor" href="#network-settings"></a>
</h2>
<p>The settings used to construct the network can be customized using
the arguments below.</p>
<div class="section level3">
<h3 id="distance-function">Distance Function<a class="anchor" aria-label="anchor" href="#distance-function"></a>
</h3>
<p>The default method for measuring the similarity between TCR/BCR
sequences is the Hamming distance, which is computed using
<code><a href="../reference/hamDistBounded.html">hamDistBounded()</a></code>. It calculates the number of differences
between two sequences of the same length. For each position in one
sequence, the character in that position is checked to see whether it
differs from the character in the same position of the other sequence.
If the sequences have different lengths, the shorter sequence is
extended by adding non-matching characters to make it the same length as
the longer sequence.</p>
<p>The Levenshtein distance, which is computed using
<code><a href="../reference/levDistBounded.html">levDistBounded()</a></code>, can be used as an alternative measurement
to determine the similarity between sequences. It calculates the minimum
number of single-character edits (insertions, deletions and
transformations) needed to transform one sequence into the other. This
method is particularly useful for comparing sequences of different
lengths and can account for insertions and deletions. When constructing
a network based on the similarity of CDR-3 nucleotide sequences, it is
preferable to use the Levenshtein distance instead of the default
Hamming distance by setting the argument
<code>dist_type = "levenshtein"</code>. However, the Levenshtein
distance requires significantly more computational time than the Hamming
distance, which may be challenging when working with large data sets
having long TCR/BCR sequences.</p>
</div>
<div class="section level3">
<h3 id="distance-cutoff">Distance Cutoff<a class="anchor" aria-label="anchor" href="#distance-cutoff"></a>
</h3>
<p>The distance function specified in the <code>dist_type</code>
argument (Hamming distance by default) is used to quantify the
similarity between TCR/BCR sequences. The chosen distance measurement
determines the distance between two nodes in the network graph.</p>
<p>By default, two nodes in the graph are connected by an edge if their
distance is at most 1. However, if users want to adjust this cutoff, the
<code>dist_cutoff</code> argument can be set to a different value. For
example, if <code>dist_cutoff = 2</code>, then two nodes will be
connected by an edge if their distance is at most 2.</p>
<p>The cutoff value controls the stringency of the network construction
and affects the number and density of edges in the network. A lower
cutoff value requires greater similarity between sequences in order for
their respective nodes to be joined by an edge connection. If
<code>dist_cutoff = 0</code>, two sequences must be identical in order
for their nodes to be joined by an edge.</p>
</div>
<div class="section level3">
<h3 id="keepremove-isolated-nodes">
<a name="isolated_nodes"></a> Keep/Remove Isolated Nodes<a class="anchor" aria-label="anchor" href="#keepremove-isolated-nodes"></a>
</h3>
<p>By default, if a given node has no edges connecting it to any other
nodes in the network, it will be removed from the network graph and will
not be included in the output.</p>
<p>Notice that while our input data contained 200 rows, the
<code>node_data</code> data frame in the returned output contains fewer
rows than this:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">output</span><span class="op">$</span><span class="va">node_data</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 122</span></span></code></pre></div>
<p>These rows correspond to the nodes that are joined by an edge to at
least one other node in the network.</p>
<p>The row names from the original input data are preserved, allowing
the user to conveniently cross-reference each node with its
corresponding row in the original input data.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">output</span><span class="op">$</span><span class="va">node_data</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "2"  "5"  "8"  "11" "12" "13"</span></span></code></pre></div>
<p>If desired, all nodes can be kept in the network, including those
that do not have any edge connections to other nodes. This is
accomplished by setting the <code>drop_isolated_nodes</code> argument to
<code>FALSE</code>.</p>
</div>
</div>
<div class="section level2">
<h2 id="network-properties-and-cluster-analysis">
<a name="network_properties"></a> Network Properties and Cluster
Analysis<a class="anchor" aria-label="anchor" href="#network-properties-and-cluster-analysis"></a>
</h2>
<p>The <code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code> function can perform additional
network analysis after constructing the network. This includes cluster
analysis and computation of network properties. Cluster analysis
partitions the network graph into densely-connected subgraphs, while
network properties describe the structural organization of the
network.</p>
<div class="section level3">
<h3 class="tabset" id="node-level-network-properties">
<a name="node_properties"></a> Node-Level Network
Properties<a class="anchor" aria-label="anchor" href="#node-level-network-properties"></a>
</h3>
<div class="section level4">
<h4 id="description">Description<a class="anchor" aria-label="anchor" href="#description"></a>
</h4>
<p>Node-level network properties are properties that pertain to each
individual node in the network graph.</p>
<p>Some are local properties, meaning that their value for a given node
depends only on a subset of the nodes in the network. One example is the
network degree of a given node, which represents the number of other
nodes that are directly joined to the given node by an edge
connection.</p>
<p>Other properties are global properties, meaning that their value for
a given node depends on all of the nodes in the network. An example is
the authority score of a node, which is computed using the entire graph
adjacency matrix (if we denote this matrix by <span class="math inline">\(A\)</span>, then the principal eigenvector of
<span class="math inline">\(A^T A\)</span> represents the authority
scores of the network nodes).</p>
<p>The names of the node-level network properties that can be computed
are listed below:</p>
<ul>
<li><code>degree</code></li>
<li><code>cluster_id</code></li>
<li><code>transitivity</code></li>
<li><code>closeness</code></li>
<li><code>centrality_by_closeness</code></li>
<li><code>eigen_centrality</code></li>
<li><code>centrality_by_eigen</code></li>
<li><code>betweenness</code></li>
<li><code>centrality_by_betweenness</code></li>
<li><code>authority_score</code></li>
<li><code>coreness</code></li>
<li><code>page_rank</code></li>
</ul>
<p>For details on the individual properties, refer to the documentation
for <code><a href="../reference/chooseNodeStats.html">chooseNodeStats()</a></code>.</p>
<p>The <code>cluster_id</code> property warrants further explanation and
is discussed in the <a href="#cluster_analysis">subsection on cluster
analysis</a>.</p>
</div>
<div class="section level4">
<h4 id="computing-node-level-properties">Computing Node-Level Properties<a class="anchor" aria-label="anchor" href="#computing-node-level-properties"></a>
</h4>
<p>Use the setting <code>node_stats = TRUE</code> to enable computation
of node-level network properties.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># build network with computation of node-level network properties</span></span>
<span><span class="va">output2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork</a></span><span class="op">(</span><span class="va">toy_data</span>, <span class="st">"CloneSeq"</span>, </span>
<span>                              node_stats <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                              output_dir <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Input data contains 200 rows.</span></span>
<span><span class="co">#&gt; Removing sequences with length fewer than 3 characters... Done. 200 rows remaining.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 122 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Generating graph plot with nodes colored by degree...</span></span></code></pre></div>
<p><img src="buildRepSeqNetwork_files/figure-html/unnamed-chunk-13-1.png" width="100%" style="display: block; margin: auto;"></p>
<pre><code><span><span class="co">#&gt;  Done.</span></span></code></pre>
<p>The node data now contains node-level network properties in addition
to the original data columns:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">output2</span><span class="op">$</span><span class="va">node_data</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "CloneSeq"                  "CloneFrequency"           </span></span>
<span><span class="co">#&gt;  [3] "CloneCount"                "SampleID"                 </span></span>
<span><span class="co">#&gt;  [5] "degree"                    "transitivity"             </span></span>
<span><span class="co">#&gt;  [7] "eigen_centrality"          "centrality_by_eigen"      </span></span>
<span><span class="co">#&gt;  [9] "betweenness"               "centrality_by_betweenness"</span></span>
<span><span class="co">#&gt; [11] "authority_score"           "coreness"                 </span></span>
<span><span class="co">#&gt; [13] "page_rank"</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">output2</span><span class="op">$</span><span class="va">node_data</span><span class="op">[</span> , <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CloneSeq"</span>, <span class="st">"degree"</span>, <span class="st">"authority_score"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;         CloneSeq degree authority_score</span></span>
<span><span class="co">#&gt; 2  GGAGATGAATCGG      1      0.00000000</span></span>
<span><span class="co">#&gt; 5  GAAAGAGAATTCG      3      0.00000000</span></span>
<span><span class="co">#&gt; 8  GGGGAGAAATTGG      2      0.04558649</span></span>
<span><span class="co">#&gt; 11 GGGGGAGAATTGC      4      0.15055366</span></span>
<span><span class="co">#&gt; 12 GGGGGGGAATTGC     10      0.52691798</span></span>
<span><span class="co">#&gt; 13 AGGGGGAAATTGG      5      0.14682343</span></span></code></pre></div>
<p>By default, all of the available node-level properties are computed
except for <code>closeness</code>, <code>centrality_by_closeness</code>
and <code>cluster_id</code>.</p>
</div>
<div class="section level4">
<h4 id="specifying-the-node-level-properties">
<a name="choose_node_stats"></a> Specifying the Node-Level
Properties<a class="anchor" aria-label="anchor" href="#specifying-the-node-level-properties"></a>
</h4>
<p>The set of node-level properties that are computed can be specified
using the <code>stats_to_include</code> argument.</p>
<p>To compute all node-level properties, the user can simply specify
<code>stats_to_include = "all"</code>.</p>
<p>To specify a particular subset of the available node-level
properties, the user must provide a named logical vector following a
particular format. This vector can be created using the
<code><a href="../reference/chooseNodeStats.html">chooseNodeStats()</a></code> function. Each node-level property name
is an argument of <code><a href="../reference/chooseNodeStats.html">chooseNodeStats()</a></code>, with the argument
accepting either <code>TRUE</code> or <code>FALSE</code> to specify
whether the property is computed. The default value for each argument
agrees with the default set of node properties seen in the previous
example. In other words, setting
<code>stats_to_include = chooseNodeStats()</code> is the same as leaving
the <code>stats_to_include</code> argument unspecified.</p>
<p>Below is an example where the user wishes to compute the
<code>closeness</code> property in addition to the default properties,
with the <code>page_rank</code> property excluded:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Modifying the default set of node-level properties</span></span>
<span><span class="va">output2</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork</a></span><span class="op">(</span><span class="va">toy_data</span>, <span class="st">"CloneSeq"</span>, </span>
<span>                     node_stats <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     stats_to_include <span class="op">=</span> </span>
<span>                       <span class="fu"><a href="../reference/chooseNodeStats.html">chooseNodeStats</a></span><span class="op">(</span>closeness <span class="op">=</span> <span class="cn">TRUE</span>, page_rank <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>If the user wishes to include only a small number of properties and
exclude the rest, this requires setting many argument values of
<code>chooseNodeStats</code> to <code>FALSE</code>, which can be
inconvenient. In this case, it may instead be easier to use the
<code>exclusiveNodeStats</code> function, which behaves in the same
manner as <code>chooseNodeStats</code>, but all of its arguments are set
to <code>FALSE</code> by default.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Include only the node-level properties specified below</span></span>
<span><span class="va">output2</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork</a></span><span class="op">(</span><span class="va">toy_data</span>, <span class="st">"CloneSeq"</span>, </span>
<span>                     node_stats <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                     stats_to_include <span class="op">=</span> </span>
<span>                       <span class="fu"><a href="../reference/chooseNodeStats.html">exclusiveNodeStats</a></span><span class="op">(</span>degree <span class="op">=</span> <span class="cn">TRUE</span>, transitivity <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 class="tabset" id="cluster-analysis">
<a name="cluster_analysis"></a> Cluster Analysis<a class="anchor" aria-label="anchor" href="#cluster-analysis"></a>
</h3>
<p>Cluster analysis involves using a community-finding algorithm to
partition the network graph into clusters (densely-connected subgraphs).
These clusters represent groups of clones/cells with similar receptor
sequences.</p>
<p>Each cluster is assigned a numeric cluster ID, and the cluster
membership ID of each node is recorded. The cluster membership ID is the
cluster ID of the cluster to which the node belongs. The cluster
membership ID is recorded as the <code>cluster_id</code> node-level
network property, included as a variable in the node-level metadata.</p>
<p>Cluster-level network properties of each cluster are also computed.
An additional data frame containing these cluster-level network
properties will be included in the output list returned by
<code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code>.</p>
<div class="section level4">
<h4 id="enabling-cluster-analysis">Enabling Cluster Analysis<a class="anchor" aria-label="anchor" href="#enabling-cluster-analysis"></a>
</h4>
<p>To perform cluster analysis, use the setting
<code>cluster_stats = TRUE</code>. This will also compute the
cluster-level network properties of each cluster.</p>
<p>If the user has set <code>node_stats = TRUE</code> and has specified
inclusion of the <code>cluster_id</code> node-level property through the
<code>stats_to_include</code> argument, then cluster analysis will
automatically be performed and the <code>cluster_id</code> values will
be added to the node-level data. However, cluster-level network
properties will not be computed unless the user has set
<code>cluster_stats = TRUE</code>.</p>
</div>
<div class="section level4">
<h4 id="clustering-algorithm">
<a name="cluster_alg"></a> Clustering Algorithm<a class="anchor" aria-label="anchor" href="#clustering-algorithm"></a>
</h4>
<p>By default, clustering is performed using
<code><a href="../reference/clustering_algorithms.html">cluster_fast_greedy()</a></code>. Other clustering algorithms can be
used instead of the default algorithm. The algorithm is specified using
the <code>cluster_fun</code> argument, which accepts one of the
following functions:</p>
<ul>
<li><code>cluster_fast_greedy</code></li>
<li><code>cluster_edge_betweenness</code></li>
<li><code>cluster_fluid_communities</code></li>
<li><code>cluster_infomap</code></li>
<li><code>cluster_label_prop</code></li>
<li><code>cluster_leading_eigen</code></li>
<li><code>cluster_leiden</code></li>
<li><code>cluster_louvain</code></li>
<li><code>cluster_optimal</code></li>
<li><code>cluster_spinglass</code></li>
<li><code>cluster_walktrap</code></li>
</ul>
<p>For example, setting <code>cluster_fun = cluster_leiden</code>
performs clustering using the <code>cluster_leiden</code> algorithm.</p>
<p>For details on the clustering algorithms, refer to <a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/reference/clustering_algorithms.html">this
documentation file</a>, which can be accessed offline using the command
<code><a href="../reference/clustering_algorithms.html">?NAIR::clustering_algorithms</a></code> in R.</p>
</div>
<div class="section level4">
<h4 id="cluster-level-network-properties">
<a name="cluster_properties"></a> Cluster-Level Network
Properties<a class="anchor" aria-label="anchor" href="#cluster-level-network-properties"></a>
</h4>
<p>If the user has set <code>cluster_stats = TRUE</code>, then after
partitioning the network into clusters, various network properties are
computed for each cluster.</p>
<p>These cluster-level network properties are recorded in their own data
frame, containing one row per cluster.</p>
<p>Below, we re-generate the output of
<code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code>, this time with cluster analysis
included.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">output3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork</a></span><span class="op">(</span><span class="va">toy_data</span>, <span class="st">"CloneSeq"</span>, </span>
<span>                              node_stats <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                              cluster_stats <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                              print_plots <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                              output_dir <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Input data contains 200 rows.</span></span>
<span><span class="co">#&gt; Removing sequences with length fewer than 3 characters... Done. 200 rows remaining.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 122 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 20 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; Generating graph plot with nodes colored by cluster_id... Done.</span></span></code></pre></div>
<p>The output list now includes an additional data frame containing the
cluster-level meta data:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">output3</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "igraph"           "adjacency_matrix" "node_data"        "cluster_data"    </span></span>
<span><span class="co">#&gt; [5] "plots"</span></span></code></pre></div>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">output3</span><span class="op">$</span><span class="va">cluster_data</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "cluster_id"                  "node_count"                 </span></span>
<span><span class="co">#&gt;  [3] "mean_seq_length"             "mean_degree"                </span></span>
<span><span class="co">#&gt;  [5] "max_degree"                  "seq_w_max_degree"           </span></span>
<span><span class="co">#&gt;  [7] "agg_count"                   "max_count"                  </span></span>
<span><span class="co">#&gt;  [9] "seq_w_max_count"             "diameter_length"            </span></span>
<span><span class="co">#&gt; [11] "global_transitivity"         "assortativity"              </span></span>
<span><span class="co">#&gt; [13] "edge_density"                "degree_centrality_index"    </span></span>
<span><span class="co">#&gt; [15] "closeness_centrality_index"  "eigen_centrality_index"     </span></span>
<span><span class="co">#&gt; [17] "eigen_centrality_eigenvalue"</span></span></code></pre></div>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">output3</span><span class="op">$</span><span class="va">cluster_data</span><span class="op">[</span> , <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;   cluster_id node_count mean_seq_length mean_degree max_degree seq_w_max_degree</span></span>
<span><span class="co">#&gt; 1          1         14           13.00        3.36          9    AAAAAAAAATTGC</span></span>
<span><span class="co">#&gt; 2          2         28           12.96        8.43         18    GGGGGGGAATTGG</span></span>
<span><span class="co">#&gt; 3          3          9           12.67        2.22          4     AGAAGAAAATTC</span></span>
<span><span class="co">#&gt; 4          4          6           13.00        3.33          9    GGGGGGAAATTGG</span></span>
<span><span class="co">#&gt; 5          5          6           12.00        2.17          3     AGGGAGGAATTC</span></span>
<span><span class="co">#&gt; 6          6         25           12.00        4.60         10     AAAAAAAAATTG</span></span></code></pre></div>
<p>A brief description of each cluster-level property is given
below:</p>
<ul>
<li>
<code>node_count</code>: The number of nodes in the cluster.</li>
<li>
<code>mean_seq_length</code>: The mean sequence length in the
cluster.</li>
<li>
<code>mean_degree</code>: The mean network degree in the
cluster.</li>
<li>
<code>max_degree</code>: The maximum network degree in the
cluster.</li>
<li>
<code>seq_w_max_degree</code>: The receptor sequence possessing the
maximum degree within the cluster.</li>
<li>
<code>agg_count</code>: The aggregate count among all nodes in the
cluster (based on the counts in <code>count_col</code>, if
provided).</li>
<li>
<code>max_count</code>: The maximum count among all nodes in the
cluster (based on the counts in <code>count_col</code>, if
provided).</li>
<li>
<code>seq_w_max_count</code>: The receptor sequence possessing the
maximum count within the cluster.</li>
<li>
<code>diameter_length</code>: The longest geodesic distance in the
cluster.</li>
<li>
<code>assortativity</code>: The assortativity coefficient of the
cluster’s graph, based on the degree (minus one) of each node in the
cluster (with the degree computed based only upon the nodes within the
cluster).</li>
<li>
<code>global_transitivity</code>: The transitivity (i.e., clustering
coefficient) for the cluster’s graph, which estimates the probability
that adjacent vertices are connected.</li>
<li>
<code>edge_density</code>: The number of edges in the cluster as a
fraction of the maximum possible number of edges.</li>
<li>
<code>degree_centrality_index</code>: The cluster-level centrality
index based on degree within the cluster graph.</li>
<li>
<code>closeness_centrality_index</code>: The cluster-level
centrality index based on closeness, i.e., distance to other nodes in
the cluster.</li>
<li>
<code>eigen_centrality_index</code>: The cluster-level centrality
index based on the eigenvector centrality scores, i.e., values of the
principal eigenvector of the adjacency matrix for the cluster.</li>
<li>
<code>eigen_centrality_eigenvalue</code>: The eigenvalue
corresponding to the principal eigenvector of the adjacency matrix for
the cluster.</li>
</ul>
</div>
<div class="section level4">
<h4 id="specifying-the-count-column">
<a name="count_col"></a> Specifying the Count Column<a class="anchor" aria-label="anchor" href="#specifying-the-count-column"></a>
</h4>
<p>Some cluster-level network properties, such as <code>agg_count</code>
and <code>max_count</code>, are only computed if the user specifies a
column of the input data containing counts for each row (i.e., clone
count for bulk data or Unique Molecular Identifier count for single-cell
data). This column is specified using the <code>count_col</code>
function, which accepts a column name or column index.</p>
<p>If a count column is specified, then its values will be coerced to
numeric values. Any rows of the input data for which the count value is
<code>NA</code> or <code>NaN</code> will be dropped before constructing
the network.</p>
</div>
<div class="section level4">
<h4 id="labeling-clusters">Labeling Clusters<a class="anchor" aria-label="anchor" href="#labeling-clusters"></a>
</h4>
<p>In order to more easily reference the clusters within the visual plot
of the network graph, it is possible to label the clusters in the plot
with their cluster IDs. This must be done after calling
<code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code>, and is accomplished using the
<code>addClusterLabels</code> function <a href="network_visualization.html#labeling_clusters">as described
here</a>.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a>
</h2>
<p>The <code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code> function includes various
arguments that facilitate customization of the network visualization.
See <a href="network_visualization.html">this vignette</a> for an
overview.</p>
<p>Additionally, the following arguments can be used to control whether
plots are generated and/or displayed in R:</p>
<ul>
<li>Use <code>print_plots = FALSE</code> to prevent plots from being
printed to the R plotting window. Plots will still be generated,
included in the output and (assuming a <a href="#output_settings">valid
output directory</a>) saved to a pdf as usual.</li>
<li>Use <code>plots = FALSE</code> to prevent plots from being generated
entirely.</li>
</ul>
</div>
<div class="section level2">
<h2 class="tabset" id="output-settings">
<a name="output_settings"></a> Output Settings<a class="anchor" aria-label="anchor" href="#output-settings"></a>
</h2>
<div class="section level3">
<h3 id="subsetting-columns-from-the-input-data">
<a name="subset_cols"></a> Subsetting Columns From the Input
Data<a class="anchor" aria-label="anchor" href="#subsetting-columns-from-the-input-data"></a>
</h3>
<p>The <code>node_data</code> data frame contained in the output of
<code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code> includes all of the columns that were
present in the original input data frame.</p>
<p>In some cases, the user may not wish for all columns to be included.
In this case, the user can specify the columns to be included using the
<code>subset_cols</code> argument, which accepts a vector containing
either the column names or the column indices of the columns to be
kept.</p>
<p>The column specified by the <code>seq_col</code> argument will
automatically be included regardless of the value of
<code>subset_cols</code>.</p>
</div>
<div class="section level3">
<h3 id="output-directory">Output Directory<a class="anchor" aria-label="anchor" href="#output-directory"></a>
</h3>
<ul>
<li>By default, <code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code> saves its output to
the current working directory.</li>
<li>Users can specify a different directory using the
<code>output_dir</code> argument. The specified output directory will be
created if it does not already exist.</li>
<li>Setting <code>output_dir</code> to <code>NULL</code> will prevent
all output from being saved.</li>
</ul>
</div>
<div class="section level3">
<h3 id="output-file-format">
<a name="output_format"></a> Output File Format<a class="anchor" aria-label="anchor" href="#output-file-format"></a>
</h3>
<ul>
<li>By default, each element of the list returned by
<code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code> is written to its own file using the
following file formats:
<ul>
<li>csv for <code>node_data</code> and <code>cluster_data</code>
</li>
<li>mtx for <code>adjacency_matrix</code>
</li>
<li>txt for <code>igraph</code>
</li>
<li>pdf for <code>plots</code> (see next subsection)</li>
</ul>
</li>
<li>For better compression and fewer files, users can specify
<code>output_type = "rds"</code> or <code>output_type = "rda"</code>,
which will save the entire output list to a single rds file or a single
rda file, respectively.
<ul>
<li>If the output list is saved using the <code>"rda"</code> (Rdata)
format, the output list will be named <code>net</code> in the R
environment when the Rdata file is loaded into R using the
<code>load</code> function.</li>
</ul>
</li>
<li>Regardless of the output file format specified, if
<code>plots = TRUE</code>, a pdf file is written to the output
directory. The pdf contains all plots generated for the network, with
one plot per page. The dimensions (in inches) for each page can be
adjusted using the <code>plot_width</code> and <code>plot_height</code>
arguments, with the defaults being <code>12</code> and <code>10</code>,
respectively.</li>
<li>
<strong>Note:</strong> the <code>ggraph</code> object for each plot
is only saved if the user sets <code>output_type = "rds"</code> or
<code>output_type = "rda"</code>. Using one of these settings is
recommended if the user wishes to modify any plots in the future. Note,
however, that plots can always be re-generated from the node-level meta
data using the <code>generateGraphPlot()</code> function as described <a href="network_visualization.html#new_plots">here</a>.</li>
</ul>
</div>
<div class="section level3">
<h3 id="output-file-names">
<a name="output_filenames"></a> Output File Name(s)<a class="anchor" aria-label="anchor" href="#output-file-names"></a>
</h3>
<ul>
<li>By default, the name of each saved file begins with
<code>MyRepSeqNetwork</code>. This prefix can be changed using the
<code>output_name</code> argument, which accepts a character
vector.</li>
<li>If <code>output_type = "individual"</code> (the default), the prefix
specified by <code>output_name</code> is followed by:
<ul>
<li>
<code>_NodeMetadata.csv</code> for <code>node_data</code>
</li>
<li>
<code>_ClusterMetadata.csv</code> for <code>cluster_data</code> (if
present)</li>
<li>
<code>_EdgeList.txt</code> for <code>igraph</code>
</li>
<li>
<code>_AdjacencyMatrix.mtx</code> for
<code>adjacency_matrix</code>
</li>
</ul>
</li>
<li>If <code>output_type</code> is set to <code>"rds"</code> or
<code>"rda"</code>, then the name of the file will be the value of
<code>output_name</code> followed by the appropriate file extension
(either <code>.rds</code> or <code>.rda</code>).</li>
<li>The name of the pdf file will be the value of
<code>output_name</code> followed by <code>.pdf</code>.</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 class="tabset" id="downstream-tasks-and-complementary-functions">
<a name="downstream"></a> Downstream Tasks and
Complementary Functions<a class="anchor" aria-label="anchor" href="#downstream-tasks-and-complementary-functions"></a>
</h2>
<p>This section covers some basic tasks the user may wish to perform
following the use of <code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code>. We describe how
these tasks can be conveniently performed using additional functions
within the <code>NAIR</code> package.</p>
<div class="section level3">
<h3 id="modifying-plots">Modifying Plots<a class="anchor" aria-label="anchor" href="#modifying-plots"></a>
</h3>
<p>In some cases, the user may wish to make changes to a plot that was
previously created by the <code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code>
function.</p>
<p>Refer to <a href="network_visualization#modifying_plots">this
vignette</a> for further details.</p>
</div>
<div class="section level3">
<h3 class="tabset" id="downstream-computation-of-network-properties">Downstream Computation of Network Properties<a class="anchor" aria-label="anchor" href="#downstream-computation-of-network-properties"></a>
</h3>
<p>If the user has already constructed the network using
<code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code> and wishes to compute network
properties without constructing the network again (which could be
time-consuming for large networks), this can be done using the output of
<code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code> along with built-in package
functions.</p>
<div class="section level4">
<h4 id="compute-node-level-properties-downstream">Compute Node-level Properties Downstream<a class="anchor" aria-label="anchor" href="#compute-node-level-properties-downstream"></a>
</h4>
<p>Node-level network properties can be computed and added to the
node-level meta data by passing the data frame to
<code><a href="../reference/addNodeNetworkStats.html">addNodeNetworkStats()</a></code>. The <code>igraph</code> object
contained in the output list of <code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code> is
provided to the <code>net</code> argument of
<code><a href="../reference/addNodeNetworkStats.html">addNodeNetworkStats()</a></code>. The <code>stats_to_include</code>
argument can be used to <a href="#node_properties">specify which network
properties to compute</a> just as in the
<code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code> function.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Add network stats to node-level meta data </span></span>
<span><span class="va">output</span><span class="op">$</span><span class="va">node_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addNodeNetworkStats.html">addNodeNetworkStats</a></span><span class="op">(</span><span class="va">output</span><span class="op">$</span><span class="va">node_data</span>,</span>
<span>                                        net <span class="op">=</span> <span class="va">output</span><span class="op">$</span><span class="va">igraph</span>,</span>
<span>                                        stats_to_include <span class="op">=</span> <span class="st">"all"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="perform-cluster-analysis-downstream">Perform Cluster Analysis Downstream<a class="anchor" aria-label="anchor" href="#perform-cluster-analysis-downstream"></a>
</h4>
<p>Cluster-level properties can be computed by passing the node-level
meta data to <code><a href="../reference/getClusterStats.html">getClusterStats()</a></code> The adjacency matrix
contained in the output list of <code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code> is
provided to the <code>adjacency_matrix</code> argument of
<code><a href="../reference/getClusterStats.html">getClusterStats()</a></code>. The <code>seq_col</code> argument
accepts a column name or column index specifying the column of the node
data that contains the receptor sequences. If not provided, cluster
properties involving receptor sequences, such as
<code>mean_seq_length</code> and <code>seq_w_max_degree</code>, will not
be computed. Similarly, the <code>count_col</code> argument is used to
<a href="#cluster_analysis">specify the column containing the
counts</a>.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Add cluster-level properties to output list from buildRepSeqNetwork</span></span>
<span><span class="va">output</span><span class="op">$</span><span class="va">cluster_data</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/getClusterStats.html">getClusterStats</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">output</span><span class="op">$</span><span class="va">node_data</span>,</span>
<span>                  adjacency_matrix <span class="op">=</span> <span class="va">output</span><span class="op">$</span><span class="va">adjacency_matrix</span>,</span>
<span>                  seq_col <span class="op">=</span> <span class="st">"CloneSeq"</span>, </span>
<span>                  count_col <span class="op">=</span> <span class="st">"CloneCount"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 20 clusters in the network... Done.</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="saving-output-after-creation">
<a name="downstream_save_output"></a> Saving Output After
Creation<a class="anchor" aria-label="anchor" href="#saving-output-after-creation"></a>
</h3>
<p>In some cases, the user may wish to save the output list returned by
<code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code> after it is created. This situation
may arise if the user has modified the output list, or if the user
wishes to use different output settings but does not wish to reconstruct
the entire network by calling <code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code>
again.</p>
<p>The <code>saveNetwork</code> function can be used to save the output
list in a manner similar to the way in which
<code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code> saves its output. The output list is
passed to the <code>net</code> argument, while the
<code>output_dir</code>, <code>output_type</code>,
<code>pdf_width</code> and <code>pdf_height</code> arguments have the <a href="#output_settings">same behavior</a> as they do in the
<code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code> function. The
<code>output_filename</code> argument has the same behavior as the
<code>output_name</code> argument of
<code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code>.</p>
<p>As with <code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code>, if the user specifies
<code>output_type = "rda"</code>, then the output list will be named
<code>net</code> in the R environment when the saved file is loaded into
R using the <code>load</code> function.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># save output list returned by buildRepSeqNetwork</span></span>
<span><span class="fu"><a href="../reference/saveNetwork.html">saveNetwork</a></span><span class="op">(</span>net <span class="op">=</span> <span class="va">output</span>, output_dir <span class="op">=</span> <span class="va">dir_out</span><span class="op">)</span></span>
<span><span class="co">#&gt; Node-level meta-data saved to file:</span></span>
<span><span class="co">#&gt;   /tmp/Rtmpu6X8US/MyRepSeqNetwork_NodeMetadata.csv</span></span>
<span><span class="co">#&gt; Cluster-level meta-data saved to file:</span></span>
<span><span class="co">#&gt;   /tmp/Rtmpu6X8US/MyRepSeqNetwork_ClusterMetadata.csv</span></span>
<span><span class="co">#&gt; Network graph plots saved to file:</span></span>
<span><span class="co">#&gt;   /tmp/Rtmpu6X8US/MyRepSeqNetwork.pdf</span></span>
<span><span class="co">#&gt; Network igraph saved in edgelist format to file:</span></span>
<span><span class="co">#&gt;   /tmp/Rtmpu6X8US/MyRepSeqNetwork_EdgeList.txt</span></span>
<span><span class="co">#&gt; Adjacency matrix saved to file:</span></span>
<span><span class="co">#&gt;   /tmp/Rtmpu6X8US/MyRepSeqNetwork_AdjacencyMatrix.mtx</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="regenerating-output-components">Regenerating Output Components<a class="anchor" aria-label="anchor" href="#regenerating-output-components"></a>
</h3>
<p>In case the <code>igraph</code> object contained in the output of
<code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code> is lost, the adjacency matrix can be
used to quickly generate the <code>igraph</code> object without
recomputing the network edges. This is done by passing the adjacency
matrix to <code><a href="../reference/generateNetworkFromAdjacencyMat.html">generateNetworkFromAdjacencyMat()</a></code>.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create igraph from adjacency matrix</span></span>
<span><span class="va">output</span><span class="op">$</span><span class="va">igraph</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generateNetworkFromAdjacencyMat.html">generateNetworkFromAdjacencyMat</a></span><span class="op">(</span><span class="va">output</span><span class="op">$</span><span class="va">adjacency_matrix</span><span class="op">)</span></span></code></pre></div>
<p>If the adjacency matrix itself is lost, it can be recomputed from the
<code>node_data</code> data frame. This is achieved by passing the
column vector of receptor sequences to
<code><a href="../reference/sparseAdjacencyMatFromSeqs.html">sparseAdjacencyMatFromSeqs()</a></code>. Doing so is typically faster
than calling <code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code> again, particularly if
<code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code> was previously called with the default
setting <a href="#network_settings"><code>drop_isolated_nodes = TRUE</code></a>.</p>
<p><code><a href="../reference/sparseAdjacencyMatFromSeqs.html">sparseAdjacencyMatFromSeqs()</a></code> has three additional
arguments. The argument <code>max_dist</code> plays the role of the <a href="#network_settings"><code>dist_cutoff</code></a> argument from
<code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code>, with the same default value of 1. The
arguments <code>dist_type</code> and <code>drop_isolated_nodes</code>
have the <a href="#network_settings">same behavior and default
values</a> as in <code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code>. These arguments
should be set to match the values used in
<code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code> when the <code>node_data</code> data
frame was created.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create adjacency matrix from receptor sequence data column</span></span>
<span><span class="va">output</span><span class="op">$</span><span class="va">adjacency_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sparseAdjacencyMatFromSeqs.html">sparseAdjacencyMatFromSeqs</a></span><span class="op">(</span><span class="va">output</span><span class="op">$</span><span class="va">node_data</span><span class="op">$</span><span class="va">CloneSeq</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 class="tabset" id="compartmentalized-workflow">Compartmentalized Workflow<a class="anchor" aria-label="anchor" href="#compartmentalized-workflow"></a>
</h2>
<p>It is possible to split the tasks performed by
<code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code> into separate, individual steps. This
allows the user greater freedom and flexibility in customizing their
workflow and analysis.</p>
<div class="section level3">
<h3 id="generatenetworkobjects">
<code>generateNetworkObjects</code><a class="anchor" aria-label="anchor" href="#generatenetworkobjects"></a>
</h3>
<p><code><a href="../reference/generateNetworkObjects.html">generateNetworkObjects()</a></code> is a lightweight function that
performs the core tasks of <code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code> and leaves
the remaining tasks to be performed individually.</p>
<p>The <code><a href="../reference/generateNetworkObjects.html">generateNetworkObjects()</a></code> function performs
<strong>only</strong> the following tasks:</p>
<ul>
<li>Builds the network graph for the immune repertoire</li>
<li>Returns the following output:
<ul>
<li>Metadata for the nodes in the network</li>
<li>Network adjacency matrix</li>
<li>
<code>igraph</code> object</li>
</ul>
</li>
</ul>
<p>The function has arguments <a href="#mainargs"><code>data</code>,
<code>seq_col</code></a>, <a href="#network_settings"><code>dist_type</code>,
<code>dist_cutoff</code> and <code>drop_isolated_nodes</code></a>, all
of which have the same behavior and default values as seen in
<code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code>.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">output_minimal</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generateNetworkObjects.html">generateNetworkObjects</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">toy_data</span>, seq_col <span class="op">=</span> <span class="st">"CloneSeq"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 122 nodes (after removing isolated nodes).</span></span></code></pre></div>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">output_minimal</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "igraph"           "adjacency_matrix" "node_data"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="standalone-input-filtering">
<a name="standalone_filtering"></a> Standalone Input Filtering<a class="anchor" aria-label="anchor" href="#standalone-input-filtering"></a>
</h3>
<p>Filtering the input data can be performed as a separate step prior to
calling <code><a href="../reference/generateNetworkObjects.html">generateNetworkObjects()</a></code>. This is done using
<code><a href="../reference/filterInputData.html">filterInputData()</a></code>. The function has arguments <a href="#mainargs"><code>data</code>, <code>seq_col</code></a>, <a href="#input_options"><code>min_seq_length</code>,
<code>drop_matches</code></a> and <a href="#subset_cols"><code>subset_cols</code></a>, all of which behave in
the same manner as seen in <code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code>. In
addition, the <code>count_col</code> argument can be used to specify a
column containing the clone count or UMI count. If specified,
observations with <code>NA</code> values in this column will be removed
from the data.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Filter input data as a preprocessing step</span></span>
<span><span class="va">filtered_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filterInputData.html">filterInputData</a></span><span class="op">(</span><span class="va">toy_data</span>, </span>
<span>                                 seq_col <span class="op">=</span> <span class="st">"CloneSeq"</span>, </span>
<span>                                 min_seq_length <span class="op">=</span> <span class="fl">13</span>,</span>
<span>                                 drop_matches <span class="op">=</span> <span class="st">"GGGG"</span>, </span>
<span>                                 subset_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CloneFrequency"</span>, <span class="st">"SampleID"</span><span class="op">)</span>,</span>
<span>                                 count_col <span class="op">=</span> <span class="st">"CloneCount"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Input data contains 200 rows.</span></span>
<span><span class="co">#&gt; Removing sequences with length fewer than 13 characters... Done. 136 rows remaining.</span></span>
<span><span class="co">#&gt; Removing sequences containing matches to the expression 'GGGG'... Done. 105 rows remaining.</span></span></code></pre></div>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">filtered_data</span><span class="op">)</span></span>
<span><span class="co">#&gt;        CloneSeq CloneCount CloneFrequency SampleID</span></span>
<span><span class="co">#&gt; 1 TTGAGGAAATTCG       3095    0.007873775  Sample1</span></span>
<span><span class="co">#&gt; 2 GGAGATGAATCGG       3057    0.007777102  Sample1</span></span>
<span><span class="co">#&gt; 3 GTCGGGTAATTGG       3575    0.009094910  Sample1</span></span>
<span><span class="co">#&gt; 4 GCCGGGTAATTCG       3994    0.010160859  Sample1</span></span>
<span><span class="co">#&gt; 5 GAAAGAGAATTCG       3670    0.009336593  Sample1</span></span>
<span><span class="co">#&gt; 6 AGGTGGGAATTCG       4076    0.010369470  Sample1</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="other-steps">Other Steps<a class="anchor" aria-label="anchor" href="#other-steps"></a>
</h3>
<p>The following tasks performed by <code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code>
are not performed by <code><a href="../reference/generateNetworkObjects.html">generateNetworkObjects()</a></code>:</p>
<ul>
<li><a href="#downstream">Computation of network properties and cluster
analysis</a></li>
<li><a href="network_visualization.html#new_plots">Generation of visual
plot</a></li>
<li><a href="#downstream_save_output">Saving output</a></li>
</ul>
<p>Each of the omitted tasks can be performed as its own step using
functions in the <code>NAIR</code> package. Click on a particular task
for details on how it can be accomplished.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Brian Neal, Hai Yang, Phi Long Le, Li Zhang.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
