<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="NAIR">
<title>Searching for Public TCR/BCR Clusters • NAIR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Searching for Public TCR/BCR Clusters">
<meta property="og:description" content="NAIR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">NAIR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.9043</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/NAIR.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/associated_clusters.html">Searching for Associated TCR/BCR Clusters</a>
    <a class="dropdown-item" href="../articles/buildRepSeqNetwork.html">buildRepSeqNetwork</a>
    <a class="dropdown-item" href="../articles/dual_chain.html">Dual-Chain Network Analysis</a>
    <a class="dropdown-item" href="../articles/network_visualization.html">Network Visualization</a>
    <a class="dropdown-item" href="../articles/public_clusters.html">Searching for Public TCR/BCR Clusters</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/mlizhangx/Network-Analysis-for-Repertoire-Sequencing-/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Searching for Public TCR/BCR Clusters</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/mlizhangx/Network-Analysis-for-Repertoire-Sequencing-/blob/HEAD/vignettes/public_clusters.Rmd" class="external-link"><code>vignettes/public_clusters.Rmd</code></a></small>
      <div class="d-none name"><code>public_clusters.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The <code>NAIR</code> package includes a set of functions that
facilitate searching for public TCR/BCR clusters across multiple samples
of Adaptive Immune Receptor Repertoire Sequencing (AIRR-Seq) data.</p>
<p>In this context, a public cluster consists of similar TCR/BCR clones
(e.g., those whose CDR3 amino acid sequences differ by at most one amino
acid) that are shared across samples (e.g., across individuals or across
time points for a single individual).</p>
<p>We first provide a brief conceptual overview, followed by a
demonstration in which we explain the process in greater detail.</p>
<div class="section level3">
<h3 id="overview-of-process">Overview of Process<a class="anchor" aria-label="anchor" href="#overview-of-process"></a>
</h3>
<ol style="list-style-type: decimal">
<li>
<a href="#step1"><strong>Filter the clusters within each sample’s
network</strong></a>. For each sample, construct the repertoire network
and use cluster analysis to partition the network into clusters. Filter
the data, keeping only those clusters with sufficient node count or
clone count.</li>
<li>
<a href="#step2"><strong>Construct global network from the filtered
data and perform clustering.</strong></a> Combine the filtered data from
all samples into a single global network. Perform cluster analysis and
assign membership to the global clusters. These clusters are considered
as the public clusters.</li>
<li>
<a href="#step3">Perform additional tasks</a> such as labeling the
clusters in the global network plot and analyzing particular clusters of
interest.</li>
</ol>
</div>
<div class="section level3">
<h3 id="simulate-data-for-demonstration">Simulate Data for Demonstration<a class="anchor" aria-label="anchor" href="#simulate-data-for-demonstration"></a>
</h3>
<p>We simulate some toy data for demonstration.</p>
<p>We simulate a total of 30 samples, each containing 30
observations.</p>
<p>Some sequences are simulated with a tendency to appear in relatively
few samples, while others are simulated with a tendency to appear in
many samples.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/">NAIR</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Welcome to the NAIR package. Get started using `vignette("NAIR", package = "NAIR")` or by visiting https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/</span></span>
<span><span class="va">data_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">dir_input_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"input_samples"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">dir_input_samples</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fl">30</span></span>
<span><span class="va">sample_size</span> <span class="op">&lt;-</span> <span class="fl">30</span> <span class="co"># (seqs per sample)          </span></span>
<span><span class="va">base_seqs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"CASSIEGQLSTDTQYF"</span>, <span class="st">"CASSEEGQLSTDTQYF"</span>, <span class="st">"CASSSVETQYF"</span>,</span>
<span>  <span class="st">"CASSPEGQLSTDTQYF"</span>, <span class="st">"RASSLAGNTEAFF"</span>, <span class="st">"CASSHRGTDTQYF"</span>, <span class="st">"CASDAGVFQPQHF"</span>,</span>
<span>  <span class="st">"CASSLTSGYNEQFF"</span>, <span class="st">"CASSETGYNEQFF"</span>, <span class="st">"CASSLTGGNEQFF"</span>, <span class="st">"CASSYLTGYNEQFF"</span>,</span>
<span>  <span class="st">"CASSLTGNEQFF"</span>, <span class="st">"CASSLNGYNEQFF"</span>, <span class="st">"CASSFPWDGYGYTF"</span>, <span class="st">"CASTLARQGGELFF"</span>,</span>
<span>  <span class="st">"CASTLSRQGGELFF"</span>, <span class="st">"CSVELLPTGPLETSYNEQFF"</span>, <span class="st">"CSVELLPTGPSETSYNEQFF"</span>,</span>
<span>  <span class="st">"CVELLPTGPSETSYNEQFF"</span>, <span class="st">"CASLAGGRTQETQYF"</span>, <span class="st">"CASRLAGGRTQETQYF"</span>,</span>
<span>  <span class="st">"CASSLAGGRTETQYF"</span>, <span class="st">"CASSLAGGRTQETQYF"</span>, <span class="st">"CASSRLAGGRTQETQYF"</span>,</span>
<span>  <span class="st">"CASQYGGGNQPQHF"</span>, <span class="st">"CASSLGGGNQPQHF"</span>, <span class="st">"CASSNGGGNQPQHF"</span>, <span class="st">"CASSYGGGGNQPQHF"</span>,</span>
<span>  <span class="st">"CASSYGGGQPQHF"</span>, <span class="st">"CASSYKGGNQPQHF"</span>, <span class="st">"CASSYTGGGNQPQHF"</span>, </span>
<span>  <span class="st">"CAWSSQETQYF"</span>, <span class="st">"CASSSPETQYF"</span>, <span class="st">"CASSGAYEQYF"</span>, <span class="st">"CSVDLGKGNNEQFF"</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># relative generation probabilities</span></span>
<span><span class="va">pgen</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span></span>
<span>  <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/toeplitz.html" class="external-link">toeplitz</a></span><span class="op">(</span><span class="fl">0.6</span><span class="op">^</span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="op">(</span><span class="va">sample_size</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">1</span>, nrow <span class="op">=</span> <span class="va">samples</span>, ncol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">base_seqs</span><span class="op">)</span> <span class="op">-</span> <span class="va">samples</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/simulateToyData.html">simulateToyData</a></span><span class="op">(</span>    </span>
<span>  samples <span class="op">=</span> <span class="va">samples</span>, </span>
<span>  sample_size <span class="op">=</span> <span class="va">sample_size</span>,</span>
<span>  prefix_length <span class="op">=</span> <span class="fl">1</span>, </span>
<span>  prefix_chars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">""</span>, <span class="st">""</span><span class="op">)</span>,</span>
<span>  prefix_probs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>, <span class="va">samples</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">samples</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  affixes <span class="op">=</span> <span class="va">base_seqs</span>, </span>
<span>  affix_probs <span class="op">=</span> <span class="va">pgen</span>, </span>
<span>  num_edits <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  output_dir <span class="op">=</span> <span class="va">dir_input_samples</span>, </span>
<span>  no_return <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
<p>Each sample is saved in a separate file using the .rds file format.
The files are named “<code>Sample1.rds</code>”,
“<code>Sample2.rds</code>”, etc. The file path of their directory is
saved to the R environment variable <code>dir_input_samples</code> for
later reference.</p>
<p>The first few rows of the data for the first sample appear as
follows:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># View first few rows of data for sample 1</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir_input_samples</span>, <span class="st">"Sample1.rds"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;           CloneSeq CloneFrequency CloneCount SampleID</span></span>
<span><span class="co">#&gt; 1 CASSIEGQLSTDTQYF     0.02606559       2832  Sample1</span></span>
<span><span class="co">#&gt; 2 CASSEEGQLSTDTQYF     0.03718396       4040  Sample1</span></span>
<span><span class="co">#&gt; 3      CASSSPETQYF     0.03182726       3458  Sample1</span></span>
<span><span class="co">#&gt; 4 CASSIEGQLSTDTQYF     0.04615781       5015  Sample1</span></span>
<span><span class="co">#&gt; 5      CAWSSQETQYF     0.06006498       6526  Sample1</span></span>
<span><span class="co">#&gt; 6 CASSEEGQLSTDTQYF     0.03363123       3654  Sample1</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 class="tabset" id="step-1-filter-clusters-within-each-sample">
<a name="step1"></a> Step 1: Filter Clusters Within
Each Sample<a class="anchor" aria-label="anchor" href="#step-1-filter-clusters-within-each-sample"></a>
</h2>
<p>First, we use <code><a href="../reference/findPublicClusters.html">findPublicClusters()</a></code> to perform network
analysis on each sample individually and select clusters based on node
count and clone count.</p>
<p>For each sample, the repertoire network is constructed and cluster
analysis is used to partition the network into clusters. The clusters
are then filtered according to node count and clone count based on
user-specified criteria. The data for the clusters that remain after
filtering are saved as files to be used as inputs for <a href="#step2">step 2</a>.</p>
<p>Below, we explain the function’s usage.</p>
<div class="section level3">
<h3 class="tabset" id="filter-settings">Filter Settings<a class="anchor" aria-label="anchor" href="#filter-settings"></a>
</h3>
<p><code><a href="../reference/findPublicClusters.html">findPublicClusters()</a></code> has several parameters that control
the criteria used to filter the nodes and clusters in each sample. These
arguments are presented below.</p>
<div class="section level4">
<h4 id="top-n-clusters">Top <span class="math inline">\(n\)</span> Clusters<a class="anchor" aria-label="anchor" href="#top-n-clusters"></a>
</h4>
<p>Within each sample, the clusters are ranked by node count. The top
<span class="math inline">\(n\)</span> highest-ranking clusters (those
with the greatest node count) within each sample are automatically
retained. The default value of <span class="math inline">\(n\)</span> is
20. A different value of <span class="math inline">\(n\)</span> can be
specified using the <code>top_n_clusters</code> argument.</p>
<p>If more than one cluster is tied for <span class="math inline">\(n\)</span>-th place in the ranking by node count,
only one such cluster will be included in the top <span class="math inline">\(n\)</span> clusters. This ensures that no more
than <span class="math inline">\(n\)</span> clusters are selected from
each sample based on this criterion.</p>
<p>If fewer than <span class="math inline">\(n\)</span> clusters are
present in the network for a sample, then all of the clusters will be
retained.</p>
</div>
<div class="section level4">
<h4 id="minimum-node-count">Minimum Node Count<a class="anchor" aria-label="anchor" href="#minimum-node-count"></a>
</h4>
<p>In addition to retaining the top <span class="math inline">\(n\)</span> clusters from each sample, clusters
that contain a sufficient number of nodes will also be retained. By
default, any cluster containing at least ten nodes will be retained.
This value can be adjusted using the <code>min_node_count</code>
argument. For example, setting <code>min_clone_count = 30</code> will
retain all clusters containing at least 30 nodes.</p>
</div>
<div class="section level4">
<h4 id="minimum-clone-count">Minimum Clone Count<a class="anchor" aria-label="anchor" href="#minimum-clone-count"></a>
</h4>
<p>In addition to the clusters retained based on node count, clusters
with a sufficient aggregate clone count will also be retained. The
aggregate clone count of a cluster is the sum of the clone counts across
all nodes (clones) in the cluster.</p>
<p>By default, any cluster with an aggregate clone count of at least 100
will be retained. This value can be adjusted using the
<code>min_clone_count</code> argument. For example, setting
<code>min_clone_count = 500</code> will retain all clusters with an
aggregate clone count of at least 500.</p>
<p>Note that clusters are only selected according to this criterion if
clone counts are provided. This is done using the <code>count_col</code>
argument as explained in the <strong>Specifying the Count
Column</strong> subsection.</p>
</div>
<div class="section level4">
<h4 id="sequence-length">Sequence Length<a class="anchor" aria-label="anchor" href="#sequence-length"></a>
</h4>
<p>When building the network for each sample, only clones whose receptor
sequences are at least three characters in length will be included in
the network. This minimum value for sequence length can be adjusted by
setting the <code>min_seq_length</code> argument to a different value.
Setting the value to <code>NULL</code> bypasses this check.</p>
</div>
<div class="section level4">
<h4 id="sequence-content">Sequence Content<a class="anchor" aria-label="anchor" href="#sequence-content"></a>
</h4>
<p>When building the network for each sample, all clones whose receptor
sequences contain any of the characters <code>*</code>, <code>|</code>
or <code>_</code> will be omitted from the network. This can be changed
using the <code>drop_matches</code> argument, which accepts a character
string containing a regular expression that specifies the pattern of
content to search for (see <code><a href="https://rdrr.io/r/base/regex.html" class="external-link">?regex</a></code>). The content of each
clone’s sequence is checked for a match to this pattern using
<code><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep()</a></code>. If a match is found, the clone is omitted from the
network. Setting the value to <code>NULL</code> bypasses this check.</p>
</div>
</div>
<div class="section level3">
<h3 class="tabset" id="specify-input-data-for-step-1">
<a name="input_settings"></a> Specify Input Data for
Step 1<a class="anchor" aria-label="anchor" href="#specify-input-data-for-step-1"></a>
</h3>
<div class="section level4">
<h4 id="file-paths-of-sample-data">
<a name="files1"></a> File Paths of Sample Data<a class="anchor" aria-label="anchor" href="#file-paths-of-sample-data"></a>
</h4>
<p>The main argument of <code><a href="../reference/findPublicClusters.html">findPublicClusters()</a></code> is the
<code>file_list</code> argument, which accepts a vector containing file
paths. Each file is assumed to contain the AIRR-Seq data for a single
sample, with observations indexed by row.</p>
<p>Below, we prepare the vector <code>input_files</code> to be provided
to the <code>file_list</code> argument of
<code><a href="../reference/findPublicClusters.html">findPublicClusters()</a></code>:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># input files for step 1 (one per sample)</span></span>
<span><span class="va">input_files</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir_input_samples</span>, </span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Sample"</span>, <span class="fl">1</span><span class="op">:</span><span class="va">samples</span>, <span class="st">".rds"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">input_files</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "/tmp/Rtmp9ZTzYj/input_samples/Sample1.rds"</span></span>
<span><span class="co">#&gt; [2] "/tmp/Rtmp9ZTzYj/input_samples/Sample2.rds"</span></span>
<span><span class="co">#&gt; [3] "/tmp/Rtmp9ZTzYj/input_samples/Sample3.rds"</span></span>
<span><span class="co">#&gt; [4] "/tmp/Rtmp9ZTzYj/input_samples/Sample4.rds"</span></span>
<span><span class="co">#&gt; [5] "/tmp/Rtmp9ZTzYj/input_samples/Sample5.rds"</span></span>
<span><span class="co">#&gt; [6] "/tmp/Rtmp9ZTzYj/input_samples/Sample6.rds"</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="file-format-of-sample-data">
<a name="input_type"></a> File Format of Sample Data<a class="anchor" aria-label="anchor" href="#file-format-of-sample-data"></a>
</h4>
<p>The file format of the input files for
<code><a href="../reference/findPublicClusters.html">findPublicClusters()</a></code> is specified using the
<code>input_type</code> parameter. The supported formats include
<code>"rds"</code>, <code>"rda"</code>, <code>"csv"</code>, as well as
files that can be read using <code><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table()</a></code>, such as
<code>"tsv"</code> and <code>"txt"</code>.</p>
<p>Our samples are stored in .rds files, so we use
<code>input_type = "rds"</code>.</p>
<p>For text formats such as <code>"csv"</code>, <code>"tsv"</code> and
<code>"txt"</code>, users can specify the separation option by utilizing
the <code>sep</code> argument. The default setting <code>sep = ""</code>
accommodates all forms of white space, i.e., one or more spaces, tabs,
newlines or carriage returns. In addition, it is important to note that
the first line of the data is assumed to be the header by default. To
disable this behavior and treat the first line as data, users must set
the <code>header</code> parameter to <code>FALSE</code>.</p>
<p>For the <code>"rda"</code> (Rdata) format, the
<code>data_symbols</code> argument is used to specify the name of each
sample’s AIRR-Seq data frame within the Rdata file (i.e., the name of
the data frame as it will appear in the R environment when the Rdata
file is loaded). The argument accepts a character string. This requires
that each sample’s data frame has the same name within its respective
Rdata file. For example, if each sample’s data frame is named
<code>mydataframe</code> and saved using the command
<code>save(mydataframe, file = myfilepath)</code>, where the file path
<code>myfilepath</code> is different for each sample, then we would set
<code>data_symbols = "mydataframe"</code>.</p>
</div>
<div class="section level4">
<h4 id="sequence-column-in-sample-data">Sequence Column in Sample Data<a class="anchor" aria-label="anchor" href="#sequence-column-in-sample-data"></a>
</h4>
<p>Each clone’s TCR/BCR sequence is used as the basis of similarity
between clones in the network.</p>
<p>The <code>seq_col</code> argument is used to specify the column
containing the clone sequences in the data frame for each sample. The
argument accepts a character string containing the column name.</p>
<p>In our simulated data, the column containing the clone sequences is
named <code>CloneSeq</code>. Hence we specify
<code>seq_col = "CloneSeq"</code> when calling
<code><a href="../reference/findPublicClusters.html">findPublicClusters()</a></code>.</p>
</div>
<div class="section level4">
<h4 id="count-column-in-sample-data">Count Column in Sample Data<a class="anchor" aria-label="anchor" href="#count-column-in-sample-data"></a>
</h4>
<p>The clone count for each clone is used to filter the clusters within
each sample.</p>
<p>The <code>count_col</code> argument is used to specify the column
containing the clone counts in the data frame for each sample. The
argument accepts a character string containing the column name.</p>
<p>In our simulated data, the column containing the clone sequences is
named <code>CloneCount</code>. Hence we specify
<code>count_col = "CloneCount"</code> when calling
<code><a href="../reference/findPublicClusters.html">findPublicClusters()</a></code>.</p>
</div>
<div class="section level4">
<h4 id="custom-sample-ids-optional">
<a name="sample_ids"></a> Custom Sample IDs (Optional)<a class="anchor" aria-label="anchor" href="#custom-sample-ids-optional"></a>
</h4>
<p>The sample ID for each clone will be included in the filtered data
saved by <code><a href="../reference/findPublicClusters.html">findPublicClusters()</a></code>. This will allow us to
distinguish the sample of origin for each clone after we combine the
filtered data from all samples in <a href="#step2">step 2</a>.</p>
<p>By default, the samples are labeled as <code>"Sample1"</code>,
<code>"Sample2"</code>, etc., according to their order of appearance in
<code>file_list</code>. The <code>sample_ids</code> argument allows for
custom sample IDs to be assigned, if desired. The argument accepts a
vector of the same length as <code>file_list</code>. Each entry of
<code>sample_ids</code> is assigned as a sample ID to the sample in the
corresponding entry of <code>file_list</code>.</p>
<p><strong>Sample ID values should avoid using characters that are
invalid for use in file names, as well as forward slash and backslash
(<code>/</code> and <code>\</code>).</strong></p>
</div>
</div>
<div class="section level3">
<h3 class="tabset" id="network-settings-for-sample-networks">
<a name="network_settings"></a> Network Settings for
Sample Networks<a class="anchor" aria-label="anchor" href="#network-settings-for-sample-networks"></a>
</h3>
<p>The settings used to construct the network for each sample can be
customized using the arguments below.</p>
<div class="section level4">
<h4 id="distance-function">Distance Function<a class="anchor" aria-label="anchor" href="#distance-function"></a>
</h4>
<p>The default method for measuring the similarity between TCR/BCR
sequences is the Hamming distance, which is computed using
<code><a href="../reference/hamDistBounded.html">hamDistBounded()</a></code>. It calculates the number of differences
between two sequences of the same length. If the sequences have
different lengths, the shorter sequence is extended by adding
non-matching characters to make it the same length as the longer
sequence.</p>
<p>The Levenshtein distance, which is computed using
<code><a href="../reference/levDistBounded.html">levDistBounded()</a></code>, can be used as an alternative measurement
to determine the similarity between sequences. It calculates the minimum
number of single-character edits (insertions, deletions and
transformations) needed to transform one sequence into the other. This
method is particularly useful for comparing sequences of different
lengths and can account for insertions and deletions. When constructing
a network based on the similarity of CDR-3 nucleotide sequences, it is
preferable to use the Levenshtein distance instead of the default
Hamming distance by setting the argument
<code>dist_type = "levenshtein"</code>. However, the Levenshtein
distance requires significantly more computational time than the Hamming
distance, which may be challenging when working with large data sets
having long TCR/BCR sequences.</p>
</div>
<div class="section level4">
<h4 id="distance-cutoff">Distance Cutoff<a class="anchor" aria-label="anchor" href="#distance-cutoff"></a>
</h4>
<p>The distance function specified in the <code>dist_type</code>
argument (Hamming distance by default) is used to quantify the
similarity between TCR/BCR sequences. The chosen distance measurement
determines the distance between two nodes in the network graph.</p>
<p>By default, two nodes in the graph are connected by an edge if their
distance is at most 1. However, if users want to adjust this cutoff, the
<code>dist_cutoff</code> argument can be set to a different value.</p>
<p>For example, if <code>dist_cutoff = 2</code>, then two nodes will be
connected by an edge if their distance is at most 2. The cutoff value
controls the stringency of the network construction and affects the
number and density of edges in the network.</p>
</div>
<div class="section level4">
<h4 id="keepremove-isolated-nodes">Keep/Remove Isolated Nodes<a class="anchor" aria-label="anchor" href="#keepremove-isolated-nodes"></a>
</h4>
<p>By default, if a given node has no edges connecting it to any other
nodes in the network, it will be removed from the network graph and will
not be included in the output.</p>
<p>If desired, all nodes can be kept in the network, including those
that do not have any edge connections to other nodes. This is
accomplished by setting the <code>drop_isolated_nodes</code> argument to
<code>FALSE</code>.</p>
</div>
<div class="section level4">
<h4 id="clustering-algorithm-for-sample-networks">Clustering Algorithm for Sample Networks<a class="anchor" aria-label="anchor" href="#clustering-algorithm-for-sample-networks"></a>
</h4>
<p>By default, clustering within each sample’s network is performed
using the <code><a href="../reference/clustering_algorithms.html">cluster_fast_greedy()</a></code> algorithm from the
<code>igraph</code> package. A different clustering algorithm can be
specified using the <code>cluster_fun</code> argument. See <a href="buildRepSeqNetwork.html#cluster_alg">this section of the
buildRepSeqNetwork vignette</a> for details.</p>
</div>
</div>
<div class="section level3">
<h3 class="tabset" id="output-settings-for-step-1">Output Settings for Step 1<a class="anchor" aria-label="anchor" href="#output-settings-for-step-1"></a>
</h3>
<p><code><a href="../reference/findPublicClusters.html">findPublicClusters()</a></code> does not return any direct output.
Instead, it saves data for the selected clusters to files that will be
used as inputs in step 2. For each sample,
<code><a href="../reference/findPublicClusters.html">findPublicClusters()</a></code> saves two files, one containing <a href="#filtered_node_data">node-specific metadata</a> and the other
containing <a href="#filtered_cluster_data">cluster-specific
metadata</a>. The data for each sample is saved after filtering the
clusters, and thus includes data only for those clusters that are
selected based on the specified filter settings.</p>
<div class="section level4">
<h4 id="variables-to-keep-from-step-1">
<a name="subset_cols"></a> Variables to Keep From Step 1<a class="anchor" aria-label="anchor" href="#variables-to-keep-from-step-1"></a>
</h4>
<p>By default, the <a href="#filtered_node_data">node-specific
metadata</a> for each sample includes all of the variables (data
columns) from the sample’s original input data frame.</p>
<p>In some cases, the user may not wish for all of the original
variables to be included. In this case, the user can specify the
original data columns they wish to keep. This is done using the
<code>subset_cols</code> argument, which accepts a vector containing
either the column names or the column indices of the columns to be
kept.</p>
<p>Regardless of the value of <code>subset_cols</code>, the variable for
clone sequence (as specified by the <code>seq_col</code> argument) will
always be included, as will the variable <code>SampleID</code>. Also
included will be <a href="#filtered_node_data">additional variables for
node-level network properties within the sample network</a>.</p>
<p>Any variables included in the output can be <a href="#vis">used as
metadata to color the nodes</a> in graph plots of the global network
during <a href="#step2">step 2</a>.</p>
</div>
<div class="section level4">
<h4 id="output-directory-for-step-1">
<a name="dir_out1"></a> Output Directory for Step 1<a class="anchor" aria-label="anchor" href="#output-directory-for-step-1"></a>
</h4>
<p>By default, a new directory named <code>public_clusters</code> is
created within the current working directory, to be used as an output
directory for saving the function’s output. A different output directory
can be specified by providing a file path to the <code>output_dir</code>
argument. The specified directory will be created if it does not already
exist.</p>
<p>Within the output directory, two subdirectories are created. One
subdirectory, named <code>node_meta_data</code>, contains the node-level
data files for each sample, and the other, named
<code>cluster_meta_data</code>, contains the cluster-level data files.
Within each of these two subdirectories, the file for each sample is
named according to its <a href="#sample_ids">sample ID</a>.</p>
</div>
<div class="section level4">
<h4 id="output-file-format-for-step-1">Output File Format for Step 1<a class="anchor" aria-label="anchor" href="#output-file-format-for-step-1"></a>
</h4>
<p>By default, each file is saved as a RDS file. This can be changed
using the <code>output_type</code> argument. Other valid options include
<code>"rda"</code> and <code>"csv"</code>.</p>
<p><strong>Note regarding the Rdata format:</strong> For the
<code>"rda"</code> (Rdata) format, each data frame containing node-level
data will be named <code>ndat</code> in the R environment when its Rdata
file is loaded, while each data frame containing cluster-level data will
be named <code>cdat</code>.</p>
</div>
<div class="section level4">
<h4 id="optional-saving-data-for-unfiltered-sample-networks">
<a name="save_unfiltered"></a> (Optional) Saving Data for Unfiltered
Sample Networks<a class="anchor" aria-label="anchor" href="#optional-saving-data-for-unfiltered-sample-networks"></a>
</h4>
<p>By default, <code><a href="../reference/findPublicClusters.html">findPublicClusters()</a></code> saves data only for the
clusters that are selected based on the specified filter settings.</p>
<p>If desired, data for each sample’s full network can also be saved
prior to filtering the clusters. This is done by by providing a file
path to the <code>output_dir_unfiltered</code> argument, which specifies
a separate output directory for the full network data. Each sample’s
full network data prior to filtering the clusters will then be saved to
the directory specified by the <code>output_dir_unfiltered</code>
argument. This data is saved separately from, and in addition to, the
default data that is saved after filtering the clusters.</p>
<p>Note that the sequence-based filter settings specified by
<code>min_seq_length</code> and <code>drop_chars</code> still apply to
each sample’s full network, since the network is only constructed after
applying these filters. The full, pre-filtered network refers to the
network that contains all of the sample’s clusters, i.e., before the
public clusters are identified and other clusters are removed.</p>
<p>The full network data for each sample includes <a href="buildRepSeqNetwork.html#base_output">all of the objects contained
in the output of <code>buildRepSeqNetwork</code></a>, including cluster
metadata. By default, these are placed in a list named <code>net</code>
and saved to an RDS file whose file name is the <a href="#sample_ids">sample ID</a>. The RDA format can be used instead of
RDS by setting <code>output_type_unfiltered = "rda"</code>.
Alternatively, setting
<code>output_type_unfiltered = "individual"</code> saves each R object
using the <a href="buildRepSeqNetwork.html#output_format">default file
formats from <code>buildRepSeqNetwork</code></a>, where each file’s name
begins with the <a href="#sample_ids">sample ID</a>.</p>
</div>
<div class="section level4">
<h4 id="optional-visualization-of-sample-networks">(Optional) Visualization of Sample Networks<a class="anchor" aria-label="anchor" href="#optional-visualization-of-sample-networks"></a>
</h4>
<p>By default, <code><a href="../reference/findPublicClusters.html">findPublicClusters()</a></code> does not produce visual
plots when constructing the network for each sample. Instead,
visualization occurs after combining the data from all samples into a
single network in <a href="#step2">step 2</a>.</p>
<p>If desired, visual plots of the full network for each sample (prior
to filtering the clusters) can be enabled by setting
<code>plots = TRUE</code>. This does nothing on its own: by default,
these plots are neither printed nor saved. To print the plots in the R
plotting window, the user must additionally set
<code>print_plots = TRUE</code>. To save the plots, the user must <a href="#save_unfiltered">specify a directory to the
<code>output_type_unfiltered</code> argument</a>.</p>
<p>By default, the plot for each sample will color the nodes according
to the cluster to which they belong. The user can specify a different
variable for coloring the nodes using the <code>color_nodes_by</code>
argument <a href="network_visualization.html#node_colors">as detailed
here</a>. It is even possible to generate <a href="network_visualization.html#multiple_plots">multiple plots of each
sample’s network, with each plot colored according to a different
variable</a>.</p>
<p>If the user <a href="#save_unfiltered">provides a directory to the
<code>output_type_unfiltered</code> argument</a>, a pdf file will be
saved for each sample, containing all plots generated for the sample’s
network. Each pdf file’s name will be the <a href="#sample_ids">sample
ID</a>.</p>
<p><a href="network_visualization.html">This vignette</a> details
additional arguments that can be used to further customize the visual
plot for each sample.</p>
</div>
</div>
<div class="section level3">
<h3 id="demonstration-step-1">
<a name="demo1"></a> Demonstration (Step 1)<a class="anchor" aria-label="anchor" href="#demonstration-step-1"></a>
</h3>
<div class="section level4">
<h4 id="execute-findpublicclusters">Execute <code>findPublicClusters()</code><a class="anchor" aria-label="anchor" href="#execute-findpublicclusters"></a>
</h4>
<p>We execute <code><a href="../reference/findPublicClusters.html">findPublicClusters()</a></code> using the <a href="#files1">input we prepared earlier</a> for the
<code>file_list</code> argument.</p>
<p>The file path of our output directory is saved to the R variable
named <code>dir_filtered_samples</code> for later reference.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># directory to save filtered sample data</span></span>
<span><span class="va">dir_filtered_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"filtered_samples"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Filter Clusters Within Each Sample</span></span>
<span><span class="fu"><a href="../reference/findPublicClusters.html">findPublicClusters</a></span><span class="op">(</span>file_list <span class="op">=</span> <span class="va">input_files</span>, </span>
<span>                   input_type <span class="op">=</span> <span class="st">"rds"</span>,</span>
<span>                   seq_col <span class="op">=</span> <span class="st">"CloneSeq"</span>, </span>
<span>                   count_col <span class="op">=</span> <span class="st">"CloneCount"</span>,</span>
<span>                   min_seq_length <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>                   drop_matches <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                   top_n_clusters <span class="op">=</span> <span class="fl">3</span>, </span>
<span>                   min_node_count <span class="op">=</span> <span class="fl">5</span>, </span>
<span>                   min_clone_count <span class="op">=</span> <span class="fl">15000</span>,</span>
<span>                   output_dir <span class="op">=</span> <span class="va">dir_filtered_samples</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; &lt;&lt;&lt; Beginning search for public clusters &gt;&gt;&gt;</span></span>
<span><span class="co">#&gt; Processing sample 1 of 30: Sample1</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 30 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 7 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt;              </span></span>
<span><span class="co">#&gt; * 3 clusters (18 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 2 of 30: Sample2</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 28 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 6 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt;              </span></span>
<span><span class="co">#&gt; * 4 clusters (22 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 3 of 30: Sample3</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 29 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 6 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt;              </span></span>
<span><span class="co">#&gt; * 3 clusters (20 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 4 of 30: Sample4</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 28 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 6 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt;              </span></span>
<span><span class="co">#&gt; * 3 clusters (18 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 5 of 30: Sample5</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 27 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 6 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt;              </span></span>
<span><span class="co">#&gt; * 4 clusters (22 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 6 of 30: Sample6</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 29 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 9 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt;              </span></span>
<span><span class="co">#&gt; * 3 clusters (15 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 7 of 30: Sample7</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 26 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 8 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt;              </span></span>
<span><span class="co">#&gt; * 3 clusters (13 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 8 of 30: Sample8</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 26 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 9 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt;              </span></span>
<span><span class="co">#&gt; * 3 clusters (12 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 9 of 30: Sample9</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 26 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 8 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt;              </span></span>
<span><span class="co">#&gt; * 3 clusters (13 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 10 of 30: Sample10</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 28 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 9 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt;              </span></span>
<span><span class="co">#&gt; * 3 clusters (15 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 11 of 30: Sample11</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 26 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 8 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt;              </span></span>
<span><span class="co">#&gt; * 4 clusters (18 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 12 of 30: Sample12</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 29 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 10 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt;              </span></span>
<span><span class="co">#&gt; * 3 clusters (12 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 13 of 30: Sample13</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 27 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 8 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt;              </span></span>
<span><span class="co">#&gt; * 4 clusters (18 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 14 of 30: Sample14</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 26 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 7 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt;              </span></span>
<span><span class="co">#&gt; * 3 clusters (17 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 15 of 30: Sample15</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 29 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 8 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt;              </span></span>
<span><span class="co">#&gt; * 3 clusters (14 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 16 of 30: Sample16</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 26 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 6 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt;              </span></span>
<span><span class="co">#&gt; * 4 clusters (21 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 17 of 30: Sample17</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 29 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 7 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt;              </span></span>
<span><span class="co">#&gt; * 4 clusters (22 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 18 of 30: Sample18</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 28 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 9 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt;              </span></span>
<span><span class="co">#&gt; * 3 clusters (14 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 19 of 30: Sample19</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 28 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 8 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt;              </span></span>
<span><span class="co">#&gt; * 3 clusters (17 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 20 of 30: Sample20</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 26 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 8 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt;              </span></span>
<span><span class="co">#&gt; * 3 clusters (15 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 21 of 30: Sample21</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 28 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 8 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt;              </span></span>
<span><span class="co">#&gt; * 4 clusters (19 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 22 of 30: Sample22</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 28 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 8 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt;              </span></span>
<span><span class="co">#&gt; * 3 clusters (16 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 23 of 30: Sample23</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 27 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 7 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt;              </span></span>
<span><span class="co">#&gt; * 4 clusters (19 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 24 of 30: Sample24</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 27 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 8 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt;              </span></span>
<span><span class="co">#&gt; * 4 clusters (18 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 25 of 30: Sample25</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 28 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 9 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt;              </span></span>
<span><span class="co">#&gt; * 4 clusters (17 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 26 of 30: Sample26</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 28 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 7 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt;              </span></span>
<span><span class="co">#&gt; * 3 clusters (17 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 27 of 30: Sample27</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 25 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 5 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt;              </span></span>
<span><span class="co">#&gt; * 3 clusters (18 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 28 of 30: Sample28</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 26 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 7 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt;              </span></span>
<span><span class="co">#&gt; * 3 clusters (16 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 29 of 30: Sample29</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 29 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 7 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt;              </span></span>
<span><span class="co">#&gt; * 4 clusters (22 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 30 of 30: Sample30</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 28 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 6 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt;              </span></span>
<span><span class="co">#&gt; * 3 clusters (19 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; All samples complete. Filtered data is located in the following directory:</span></span>
<span><span class="co">#&gt;   /tmp/Rtmp9ZTzYj/filtered_samples</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="output-of-findpublicclusters">Output of <code>findPublicClusters()</code><a class="anchor" aria-label="anchor" href="#output-of-findpublicclusters"></a>
</h4>
<p>In practice, we can proceed to <a href="#step2">step 2</a> without
examining the output from step 1. Here we present some of the output
purely for demonstration.</p>
<div class="section level5">
<h5 id="node-metadata-for-filtered-sample-networks">
<a name="filtered_node_data"></a> Node Metadata for Filtered Sample
Networks<a class="anchor" aria-label="anchor" href="#node-metadata-for-filtered-sample-networks"></a>
</h5>
<p>For the sake of demonstration, we load the node-level metadata for
the first sample’s filtered network in order to show the variables it
contains.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Directory of filtered node data</span></span>
<span><span class="va">dir_filtered_samples_node</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir_filtered_samples</span>, <span class="st">"node_meta_data"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load first file </span></span>
<span><span class="va">sample1_filtered_node_data</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">dir_filtered_samples_node</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># Variables contained in each sample's filtered node-level data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">sample1_filtered_node_data</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "CloneSeq"                           "CloneFrequency"                    </span></span>
<span><span class="co">#&gt;  [3] "CloneCount"                         "SampleID"                          </span></span>
<span><span class="co">#&gt;  [5] "SampleLevelNetworkDegree"           "ClusterIDInSample"                 </span></span>
<span><span class="co">#&gt;  [7] "SampleLevelTransitivity"            "SampleLevelCloseness"              </span></span>
<span><span class="co">#&gt;  [9] "SampleLevelCentralityByCloseness"   "SampleLevelEigenCentrality"        </span></span>
<span><span class="co">#&gt; [11] "SampleLevelCentralityByEigen"       "SampleLevelBetweenness"            </span></span>
<span><span class="co">#&gt; [13] "SampleLevelCentralityByBetweenness" "SampleLevelAuthorityScore"         </span></span>
<span><span class="co">#&gt; [15] "SampleLevelCoreness"                "SampleLevelPageRank"</span></span></code></pre></div>
<p>Each row in the data frame corresponds to a node belonging to one of
the filtered clusters in the sample network.</p>
<p>Refer <a href="buildRepSeqNetwork.html#node_properties">here</a> for
more information about the node-level network properties.</p>
</div>
<div class="section level5">
<h5 id="cluster-metadata-for-filtered-sample-networks">
<a name="filtered_cluster_data"></a> Cluster Metadata for Filtered
Sample Networks<a class="anchor" aria-label="anchor" href="#cluster-metadata-for-filtered-sample-networks"></a>
</h5>
<p>We load the cluster-level metadata for the first sample’s filtered
network in order to show the variables it contains.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Directory of filtered cluster data</span></span>
<span><span class="va">dir_filtered_samples_cluster</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir_filtered_samples</span>, <span class="st">"cluster_meta_data"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load first file </span></span>
<span><span class="va">sample1_filtered_cluster_data</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">dir_filtered_samples_cluster</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># Variables contained in each sample's filtered node-level data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">sample1_filtered_cluster_data</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "ClusterIDInSample"           "node_count"                 </span></span>
<span><span class="co">#&gt;  [3] "mean_seq_length"             "mean_degree"                </span></span>
<span><span class="co">#&gt;  [5] "max_degree"                  "seq_w_max_degree"           </span></span>
<span><span class="co">#&gt;  [7] "agg_count"                   "max_count"                  </span></span>
<span><span class="co">#&gt;  [9] "seq_w_max_count"             "diameter_length"            </span></span>
<span><span class="co">#&gt; [11] "global_transitivity"         "assortativity"              </span></span>
<span><span class="co">#&gt; [13] "edge_density"                "degree_centrality_index"    </span></span>
<span><span class="co">#&gt; [15] "closeness_centrality_index"  "eigen_centrality_index"     </span></span>
<span><span class="co">#&gt; [17] "eigen_centrality_eigenvalue" "SampleID"</span></span></code></pre></div>
<p>Each row in the data frame corresponds to one of the filtered
clusters in the sample network.</p>
<p>Refer <a href="buildRepSeqNetwork.html#cluster_properties">here</a>
for more information about the cluster-level network properties.</p>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 class="tabset" id="step-2-global-network-of-public-clusters">
<a name="step2"></a> Step 2: Global Network of Public
Clusters<a class="anchor" aria-label="anchor" href="#step-2-global-network-of-public-clusters"></a>
</h2>
<p>In <a href="#step1">step 1</a>, we used
<code><a href="../reference/findPublicClusters.html">findPublicClusters()</a></code> to perform network analysis on each
sample individually and select clusters based on node count and clone
count.</p>
<p>Next, we will use <code><a href="../reference/buildPublicClusterNetwork.html">buildPublicClusterNetwork()</a></code> to combine
the filtered data from all samples into a single global network and
perform clustering analysis.</p>
<div class="section level3">
<h3 id="specify-input-data-for-step-2">Specify Input Data for Step 2<a class="anchor" aria-label="anchor" href="#specify-input-data-for-step-2"></a>
</h3>
<p>The output files created in the previous step contain the filtered
data for each sample. <a href="#dir_out1">As detailed earlier</a>, these
files are located in two separate subdirectories, one containing the
files for the node-level metadata and the other containing the files for
the cluster-level metadata.</p>
<p>To call <code><a href="../reference/buildPublicClusterNetwork.html">buildPublicClusterNetwork()</a></code>, we only need the
files containing the node-level metadata.</p>
<div class="section level4">
<h4 id="file-paths-of-filtered-node-data">
<a name="files2"></a> File Paths of Filtered Node Data<a class="anchor" aria-label="anchor" href="#file-paths-of-filtered-node-data"></a>
</h4>
<p>If <code><a href="../reference/findPublicClusters.html">findPublicClusters()</a></code> was called with the default
value of <code>output_dir</code>, then it is not necessary to specify
the input file paths when calling
<code><a href="../reference/buildPublicClusterNetwork.html">buildPublicClusterNetwork()</a></code>, <em>provided the current
working directory is the same as the previous working directory, and
provided <code><a href="../reference/findPublicClusters.html">findPublicClusters()</a></code> has not previously been
called with different input data while using this working
directory.</em></p>
<p>If <code><a href="../reference/findPublicClusters.html">findPublicClusters()</a></code> was called with a non-default
value for <code>output_dir</code>, then the node data files must be
specified using the <code>file_list</code> argument, which accepts a
vector of file paths. The node data files will be located in the
<code>node_meta_data</code> subdirectory of the <a href="#demo1">output
directory used in step 1</a>. They can be conveniently referenced using
<code><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files()</a></code> with <code>full.names = TRUE</code>, assuming
that <code><a href="../reference/findPublicClusters.html">findPublicClusters()</a></code> has not previously been called
with different input data and the same output directory.</p>
<p>Below, we create the vector of file paths to provide to the
<code>file_list</code> argument.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Directory of filtered node data</span></span>
<span><span class="va">dir_filtered_samples_node</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir_filtered_samples</span>, <span class="st">"node_meta_data"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Vector of file paths for filtered node data</span></span>
<span><span class="va">files_filtered_samples_node</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">dir_filtered_samples_node</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">files_filtered_samples_node</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "/tmp/Rtmp9ZTzYj/filtered_samples/node_meta_data/Sample1.rds" </span></span>
<span><span class="co">#&gt; [2] "/tmp/Rtmp9ZTzYj/filtered_samples/node_meta_data/Sample10.rds"</span></span>
<span><span class="co">#&gt; [3] "/tmp/Rtmp9ZTzYj/filtered_samples/node_meta_data/Sample11.rds"</span></span>
<span><span class="co">#&gt; [4] "/tmp/Rtmp9ZTzYj/filtered_samples/node_meta_data/Sample12.rds"</span></span>
<span><span class="co">#&gt; [5] "/tmp/Rtmp9ZTzYj/filtered_samples/node_meta_data/Sample13.rds"</span></span>
<span><span class="co">#&gt; [6] "/tmp/Rtmp9ZTzYj/filtered_samples/node_meta_data/Sample14.rds"</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="file-format-of-filtered-node-data">File Format of Filtered Node Data<a class="anchor" aria-label="anchor" href="#file-format-of-filtered-node-data"></a>
</h4>
<p>If <code><a href="../reference/findPublicClusters.html">findPublicClusters()</a></code> was called with the default
value of <code>output_type</code>, then it is not necessary to specify
the input type when calling
<code><a href="../reference/buildPublicClusterNetwork.html">buildPublicClusterNetwork()</a></code>.</p>
<p>If the files created by <code><a href="../reference/findPublicClusters.html">findPublicClusters()</a></code> were saved
using a different file format, the <code>input_type</code> argument can
be used to specify the format used. This argument has the <a href="#input_type">same behavior described earlier</a> in
<code><a href="../reference/findPublicClusters.html">findPublicClusters()</a></code>. The situational arguments
<code>header</code>, <code>sep</code> and <code>data_symbols</code> also
perform the same roles as before.</p>
<p><strong>Note about Rdata file format:</strong> If the
<code>"rda"</code> (Rdata) input file format is specified using
<code>input_type = "rda"</code>, then the <code>data_symbols</code>
argument is used to specify the name of each sample’s node-level data
frame within its respective Rdata file. If
<code><a href="../reference/findPublicClusters.html">findPublicClusters()</a></code> was called with
<code>output_type = "rda"</code>, then the node-level data frame for
each sample will be named <code>ndat</code>. Accordingly, the default
value of <code>data_symbols</code> is <code>ndat</code>, so the user
should not need to manually specify the value of
<code>data_symbols</code> unless the names within the files are
changed.</p>
</div>
<div class="section level4">
<h4 id="column-names-in-filtered-node-data">Column Names in Filtered Node Data<a class="anchor" aria-label="anchor" href="#column-names-in-filtered-node-data"></a>
</h4>
<p>The <code>seq_col</code> and <code>count_col</code> arguments are
used to specify the names of the columns containing the receptor
sequences and clone counts, respectively, in the <a href="#input_settings">same manner seen earlier</a> for
<code><a href="../reference/findPublicClusters.html">findPublicClusters()</a></code>.</p>
<p>Generally, users should provide the same values for these arguments
as they did in step 1.</p>
<p>The clone counts are used to compute some of the cluster-level
properties, which will be omitted if <code>count_col</code> is left
unspecified.</p>
</div>
</div>
<div class="section level3">
<h3 class="tabset" id="settings-for-global-network-analysis">Settings for Global Network Analysis<a class="anchor" aria-label="anchor" href="#settings-for-global-network-analysis"></a>
</h3>
<div class="section level4">
<h4 id="network-construction">Network Construction<a class="anchor" aria-label="anchor" href="#network-construction"></a>
</h4>
<p>The settings used to construct the network can be customized <a href="#network_settings">using the same arguments from
<code>findPublicClusters()</code></a>, shown below with their default
values for <code><a href="../reference/buildPublicClusterNetwork.html">buildPublicClusterNetwork()</a></code>:</p>
<ul>
<li><code>dist_type = "hamming"</code></li>
<li><code>dist_cutoff = 1</code></li>
<li><code>drop_isolated_nodes = FALSE</code></li>
</ul>
<p>Refer to <a href="buildRepSeqNetwork.html#network_settings">this
vignette</a> for details on their meaning and behavior.</p>
</div>
<div class="section level4">
<h4 id="clustering-algorithm-for-global-network">Clustering Algorithm for Global Network<a class="anchor" aria-label="anchor" href="#clustering-algorithm-for-global-network"></a>
</h4>
<p>After constructing the global network,
<code><a href="../reference/buildPublicClusterNetwork.html">buildPublicClusterNetwork()</a></code> performs cluster analysis on
the network nodes, partitioning the global network graph into
densely-connected subgraphs. These global clusters can contain nodes
from different samples.</p>
<p>By default, clustering within the global network is performed using
the <code><a href="../reference/clustering_algorithms.html">cluster_fast_greedy()</a></code> algorithm from the
<code>igraph</code> package. A different clustering algorithm can be
specified using the <code>cluster_fun</code> argument. See <a href="buildRepSeqNetwork.html#cluster_alg">this section of the
buildRepSeqNetwork vignette</a> for details.</p>
</div>
<div class="section level4">
<h4 id="visualization-of-global-network">
<a name="vis"></a> Visualization of Global Network<a class="anchor" aria-label="anchor" href="#visualization-of-global-network"></a>
</h4>
<p>By default, <code><a href="../reference/buildPublicClusterNetwork.html">buildPublicClusterNetwork()</a></code> produces a
visual plot of the global network graph. This step can be omitted by
setting <code>plots = FALSE</code>.</p>
<p>By default, the plot colors the nodes according to sample ID. This
can assist the user in identifying the public clusters.</p>
<p>If desired, a different variable can be used to color the nodes. This
is done using the <code>color_nodes_by</code> argument, which accepts a
character string specifying the variable to use. The value should be the
name of a <a href="#subset_cols">variable present in the filtered
node-level metadata for each sample</a> or a <a href="#node_variables">node-level network property that has been
computed for the global network</a>.</p>
<p>The <code>color_nodes_by</code> argument also accommodates a
character vector specifying multiple variables, in which case <a href="network_visualization.html#multiple_plots">one plot will be
created for each variable specified, with each plot coloring the nodes
according to the respective variable</a>. If the user wishes to include
the binary variable of interest as one of these variables, it is
contained in each neighborhood’s data frame as the variable
<code>"GroupID"</code>.</p>
<p><a href="network_visualization.html">This vignette</a> details
additional arguments that can be used to further customize the visual
plot, such as arguments to control the color scale, node size, title and
legend.</p>
</div>
</div>
<div class="section level3">
<h3 class="tabset" id="output-settings-for-step-2">Output Settings for Step 2<a class="anchor" aria-label="anchor" href="#output-settings-for-step-2"></a>
</h3>
<p>The output returned by <code><a href="../reference/buildPublicClusterNetwork.html">buildPublicClusterNetwork()</a></code>
follows the <a href="buildRepSeqNetwork.html#output_settings">same
format as the output of <code>buildRepSeqNetwork</code></a>. The
function returns a list containing the <a href="#node_variables">node-level</a> and <a href="#cluster_variables">cluster-level</a> metadata for the global
network, as well as any plots generated, in addition to the network
adjacency matrix and the <code>igraph</code> network edge list.</p>
<div class="section level4">
<h4 id="output-directory-for-step-2">Output Directory for Step 2<a class="anchor" aria-label="anchor" href="#output-directory-for-step-2"></a>
</h4>
<p>By default, a subdirectory named <code>public_clusters</code> is
created within the current working directory, and the contents of the
list returned by <code><a href="../reference/buildPublicClusterNetwork.html">buildPublicClusterNetwork()</a></code> are saved to
this subdirectory.</p>
<p>The output can be saved to a different directory by providing a file
path to the <code>output_dir</code> argument.</p>
<p>The user can also specify <code>output_dir = NULL</code> in order to
prevent the output from being saved.</p>
</div>
<div class="section level4">
<h4 id="output-file-format-for-step-2">Output File Format for Step 2<a class="anchor" aria-label="anchor" href="#output-file-format-for-step-2"></a>
</h4>
<p>By default, each element of the list returned by
<code><a href="../reference/buildPublicClusterNetwork.html">buildPublicClusterNetwork()</a></code> is saved as an individual file.
The file formats are the same default file formats used by <a href="buildRepSeqNetwork.html#output_settings"><code>buildRepSeqNetwork</code></a>.
In particular, the node-level and cluster-level metadata are saved as
csv files.</p>
<p>Alternatively, the user can save the entire output list to a single
compressed rds or rda file by setting <code>output_type = "rds"</code>
or <code>output_type = "rda"</code>, respectively.</p>
</div>
<div class="section level4">
<h4 id="pdf-of-global-network-graph-plots">PDF of Global Network Graph Plots<a class="anchor" aria-label="anchor" href="#pdf-of-global-network-graph-plots"></a>
</h4>
<ul>
<li>Regardless of the output file format specified, a separate pdf file
is created in <code>output_dir</code> containing all graph plots
generated for the global network. The dimensions (in inches) for the pdf
can be adjusted using the <code>plot_width</code> and
<code>plot_height</code> arguments, with the defaults being
<code>12</code> and <code>10</code>, respectively. The pdf file is not
created if no plots are generated (i.e., if the argument
<code>plots</code> is set to <code>FALSE</code>).</li>
<li>
<strong>Note:</strong> the <code>ggraph</code> object for each plot
is only saved if the user sets <code>output_type = "rds"</code> or
<code>output_type = "rda"</code>. Using one of these settings is
recommended used if the user wishes to modify any plots in the future.
Note, however, that plots can always be re-generated from the node-level
meta data using <code><a href="../reference/generateNetworkGraphPlots.html">generateNetworkGraphPlots()</a></code> as described <a href="network_visualization.html#new_plots">here</a>.</li>
</ul>
</div>
<div class="section level4">
<h4 id="output-file-names-for-step-2">Output File Name(s) for Step 2<a class="anchor" aria-label="anchor" href="#output-file-names-for-step-2"></a>
</h4>
<p>By default, the name of each saved file begins with
<code>PublicClusterNetwork</code>. This prefix can be changed using the
<code>output_name</code> argument, which accepts a character vector.</p>
<p>Refer to <a href="buildRepSeqNetwork.html#output_filenames">this
vignette</a> for further details on how the output files are named.</p>
</div>
</div>
<div class="section level3">
<h3 id="demonstration-step-2">
<a name="demo2"></a> Demonstration (Step 2)<a class="anchor" aria-label="anchor" href="#demonstration-step-2"></a>
</h3>
<div class="section level4">
<h4 id="execute-buildpublicclusternetwork">Execute <code>buildPublicClusterNetwork()</code><a class="anchor" aria-label="anchor" href="#execute-buildpublicclusternetwork"></a>
</h4>
<p>We execute <code><a href="../reference/buildPublicClusterNetwork.html">buildPublicClusterNetwork()</a></code> using the <a href="#files2">input we prepared earlier</a> for the
<code>file_list</code> argument:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dir_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"public_clusters"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Collect clones from all public clusters and perform network analysis</span></span>
<span><span class="va">public_clusters</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/buildPublicClusterNetwork.html">buildPublicClusterNetwork</a></span><span class="op">(</span>file_list <span class="op">=</span> <span class="va">files_filtered_samples_node</span>,</span>
<span>                            seq_col <span class="op">=</span> <span class="st">"CloneSeq"</span>, </span>
<span>                            count_col <span class="op">=</span> <span class="st">"CloneCount"</span>,</span>
<span>                            size_nodes_by <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                            output_dir <span class="op">=</span> <span class="va">dir_out</span></span>
<span>  <span class="op">)</span></span>
<span><span class="co">#&gt; Building network of public clusters:</span></span>
<span><span class="co">#&gt; Input data contains 517 rows.</span></span>
<span><span class="co">#&gt; Removing sequences with length fewer than 3 characters... Done. 517 rows remaining.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 517 nodes.</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 20 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; Generating graph plot with nodes colored by SampleID...</span></span>
<span><span class="co">#&gt;  Done.</span></span>
<span><span class="co">#&gt; Node-level meta-data saved to file:</span></span>
<span><span class="co">#&gt;   /tmp/Rtmp9ZTzYj/public_clusters/PublicClusterNetwork_NodeMetadata.csv</span></span>
<span><span class="co">#&gt; Cluster-level meta-data saved to file:</span></span>
<span><span class="co">#&gt;   /tmp/Rtmp9ZTzYj/public_clusters/PublicClusterNetwork_ClusterMetadata.csv</span></span></code></pre></div>
<p><img src="public_clusters_files/figure-html/unnamed-chunk-8-1.png" width="100%" style="display: block; margin: auto;"></p>
<pre><code><span><span class="co">#&gt; Network graph plots saved to file:</span></span>
<span><span class="co">#&gt;   /tmp/Rtmp9ZTzYj/public_clusters/PublicClusterNetwork.pdf</span></span>
<span><span class="co">#&gt; Network igraph saved in edgelist format to file:</span></span>
<span><span class="co">#&gt;   /tmp/Rtmp9ZTzYj/public_clusters/PublicClusterNetwork_EdgeList.txt</span></span>
<span><span class="co">#&gt; Adjacency matrix saved to file:</span></span>
<span><span class="co">#&gt;   /tmp/Rtmp9ZTzYj/public_clusters/PublicClusterNetwork_AdjacencyMatrix.mtx</span></span></code></pre>
<p>From the plot of the global network graph, where the nodes are
colored according to sample ID, we notice that many samples are
represented among the larger clusters. Hence <a href="#step3">these
clusters may be of particular interest</a> as public clusters.</p>
</div>
<div class="section level4">
<h4 id="output-of-buildpublicclusternetwork">Output of <code>buildPublicClusterNetwork()</code><a class="anchor" aria-label="anchor" href="#output-of-buildpublicclusternetwork"></a>
</h4>
<p>The default files saved by <code><a href="../reference/buildPublicClusterNetwork.html">buildPublicClusterNetwork()</a></code>
have the following names:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Files saved by buildPublicClusterNetwork</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">dir_out</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "PublicClusterNetwork_AdjacencyMatrix.mtx"</span></span>
<span><span class="co">#&gt; [2] "PublicClusterNetwork_ClusterMetadata.csv"</span></span>
<span><span class="co">#&gt; [3] "PublicClusterNetwork_EdgeList.txt"       </span></span>
<span><span class="co">#&gt; [4] "PublicClusterNetwork_NodeMetadata.csv"   </span></span>
<span><span class="co">#&gt; [5] "PublicClusterNetwork.pdf"</span></span></code></pre></div>
<p>The elements of the output list returned by
<code><a href="../reference/buildPublicClusterNetwork.html">buildPublicClusterNetwork()</a></code> have the following names:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># output returned by buildPublicClusterNetwork</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">public_clusters</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "igraph"           "adjacency_matrix" "node_data"        "cluster_data"    </span></span>
<span><span class="co">#&gt; [5] "plots"</span></span></code></pre></div>
<p>Below, we examine the data frames <code>node_data</code> and
<code>cluster_data</code>. For details on the other list elements, refer
to <a href="buildRepSeqNetwork.html#base_output">the output of the
<code>buildRepSeqNetwork</code> function</a>.</p>
<div class="section level5">
<h5 id="node-metadata-for-global-network">
<a name="node_variables"></a> Node Metadata for Global Network<a class="anchor" aria-label="anchor" href="#node-metadata-for-global-network"></a>
</h5>
<p>The list element <code>node_data</code> is a data frame containing
metadata for the network nodes, where each row corresponds to a node in
the global network graph. The following variables are present in this
data frame:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># variables in the node-level metadata</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">public_clusters</span><span class="op">$</span><span class="va">node_data</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "CloneSeq"                           "CloneFrequency"                    </span></span>
<span><span class="co">#&gt;  [3] "CloneCount"                         "SampleID"                          </span></span>
<span><span class="co">#&gt;  [5] "SampleLevelNetworkDegree"           "ClusterIDInSample"                 </span></span>
<span><span class="co">#&gt;  [7] "SampleLevelTransitivity"            "SampleLevelCloseness"              </span></span>
<span><span class="co">#&gt;  [9] "SampleLevelCentralityByCloseness"   "SampleLevelEigenCentrality"        </span></span>
<span><span class="co">#&gt; [11] "SampleLevelCentralityByEigen"       "SampleLevelBetweenness"            </span></span>
<span><span class="co">#&gt; [13] "SampleLevelCentralityByBetweenness" "SampleLevelAuthorityScore"         </span></span>
<span><span class="co">#&gt; [15] "SampleLevelCoreness"                "SampleLevelPageRank"               </span></span>
<span><span class="co">#&gt; [17] "PublicNetworkDegree"                "ClusterIDPublic"                   </span></span>
<span><span class="co">#&gt; [19] "PublicTransitivity"                 "PublicCloseness"                   </span></span>
<span><span class="co">#&gt; [21] "PublicCentralityByCloseness"        "PublicEigenCentrality"             </span></span>
<span><span class="co">#&gt; [23] "PublicCentralityByEigen"            "PublicBetweenness"                 </span></span>
<span><span class="co">#&gt; [25] "PublicCentralityByBetweenness"      "PublicAuthorityScore"              </span></span>
<span><span class="co">#&gt; [27] "PublicCoreness"                     "PublicPageRank"</span></span></code></pre></div>
<p>Notice that <a href="#subset_cols">by default</a>, all variables that
were present in each sample’s original data, such as
<code>"CloneFrequency"</code> and <code>"CloneCount"</code>, are
automatically carried over into this data.</p>
<p>Many <a href="buildRepSeqNetwork.html#node_properties">node-level
network properties</a> are present in the data. Some pertain to the
network for the individual sample from which each node originated, i.e.,
the networks constructed in <a href="#step1">step 1</a>. This includes
all variables that begin with <code>"SampleLevel"</code>, such as
<code>SampleLevelNetworkDegree</code>, as well as the variable
<code>ClusterIDInSample</code>.</p>
<p>The variables that begin with <code>"Public"</code>, such as
<code>PublicNetworkDegree</code>, pertain to the global network. In
particular, <code>ClusterIDPublic</code> indicates the ID of the global
cluster to which each node belongs.</p>
</div>
<div class="section level5">
<h5 id="cluster-metadata-for-global-network">
<a name="cluster_variables"></a> Cluster Metadata for Global
Network<a class="anchor" aria-label="anchor" href="#cluster-metadata-for-global-network"></a>
</h5>
<p>The list element <code>cluster_data</code> is a data frame containing
metadata for the public clusters, where each row corresponds to a
cluster in the global network. The following variables are present in
this data frame:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># variables in the node-level metadata</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">public_clusters</span><span class="op">$</span><span class="va">cluster_data</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "cluster_id"                  "node_count"                 </span></span>
<span><span class="co">#&gt;  [3] "mean_seq_length"             "mean_degree"                </span></span>
<span><span class="co">#&gt;  [5] "max_degree"                  "seq_w_max_degree"           </span></span>
<span><span class="co">#&gt;  [7] "agg_count"                   "max_count"                  </span></span>
<span><span class="co">#&gt;  [9] "seq_w_max_count"             "diameter_length"            </span></span>
<span><span class="co">#&gt; [11] "global_transitivity"         "assortativity"              </span></span>
<span><span class="co">#&gt; [13] "edge_density"                "degree_centrality_index"    </span></span>
<span><span class="co">#&gt; [15] "closeness_centrality_index"  "eigen_centrality_index"     </span></span>
<span><span class="co">#&gt; [17] "eigen_centrality_eigenvalue"</span></span></code></pre></div>
<p>Refer <a href="buildRepSeqNetwork.html#cluster_properties">here</a>
for more information about the cluster-level network properties.</p>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 class="tabset" id="step-3-additional-tasks">
<a name="step3"></a> Step 3: Additional Tasks<a class="anchor" aria-label="anchor" href="#step-3-additional-tasks"></a>
</h2>
<p>After calling <code><a href="../reference/buildPublicClusterNetwork.html">buildPublicClusterNetwork()</a></code>, the following
tasks can be performed using the returned output.</p>
<div class="section level3">
<h3 id="labeling-the-global-clusters">Labeling the Global Clusters<a class="anchor" aria-label="anchor" href="#labeling-the-global-clusters"></a>
</h3>
<p>In order to more easily cross-reference the clusters in the visual
plot with the clusters in the data, we can label the clusters with their
ID numbers.</p>
<p>This is accomplished using <code><a href="../reference/addClusterLabels.html">addClusterLabels()</a></code> <a href="network_visualization.html#labeling_clusters">as described
here</a>. Note that within the node-level metadata, the global cluster
ID is stored in the variable <code>ClusterIDPublic</code>, so we must
provide this column name to the <code>cluster_id_col</code>
argument.</p>
<p>Below, we label the six largest clusters in the plot with their
cluster IDs:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Modify plot to add labels to the clusters</span></span>
<span><span class="va">public_clusters</span><span class="op">$</span><span class="va">plots</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/addClusterLabels.html">addClusterLabels</a></span><span class="op">(</span>plot <span class="op">=</span> <span class="va">public_clusters</span><span class="op">$</span><span class="va">plots</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                   net <span class="op">=</span> <span class="va">public_clusters</span>,</span>
<span>                   top_n_clusters <span class="op">=</span> <span class="fl">6</span>,</span>
<span>                   cluster_id_col <span class="op">=</span> <span class="st">"ClusterIDPublic"</span>,</span>
<span>                   size <span class="op">=</span> <span class="fl">7</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># View modified plot</span></span>
<span><span class="va">public_clusters</span><span class="op">$</span><span class="va">plots</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#&gt; Warning: Removed 511 rows containing missing values (`geom_text()`).</span></span></code></pre></div>
<p><img src="public_clusters_files/figure-html/unnamed-chunk-13-1.png" width="100%" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="focusing-on-individual-clusters-of-interest">Focusing on Individual Clusters of Interest<a class="anchor" aria-label="anchor" href="#focusing-on-individual-clusters-of-interest"></a>
</h3>
<p>If we wish to focus on a particular cluster of interest within the
global network, we can build a network exclusively using the clones from
that cluster. This is accomplished using
<code><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork()</a></code>, where we subset our data according to
the value of the <code>ClusterIDPublic</code> column, which contains the
global cluster IDs.</p>
<p>Below, we focus on the first cluster, which in this case is also the
largest cluster by node count. In the plot, we color each node according
to the receptor sequence of its corresponding clone.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># focus on cluster 1</span></span>
<span><span class="fu"><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> </span>
<span>    <span class="va">public_clusters</span><span class="op">$</span><span class="va">node_data</span><span class="op">[</span><span class="va">public_clusters</span><span class="op">$</span><span class="va">node_data</span><span class="op">$</span><span class="va">ClusterIDPublic</span> <span class="op">==</span> <span class="fl">1</span>, <span class="op">]</span>,</span>
<span>  seq_col <span class="op">=</span> <span class="st">"CloneSeq"</span>, </span>
<span>  color_nodes_by <span class="op">=</span> <span class="st">"CloneSeq"</span>, </span>
<span>  color_scheme <span class="op">=</span> <span class="st">"plasma"</span>, </span>
<span>  size_nodes_by <span class="op">=</span> <span class="fl">3</span>, </span>
<span>  output_dir <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>  output_name <span class="op">=</span> <span class="st">"Cluster 1"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Input data contains 96 rows.</span></span>
<span><span class="co">#&gt; Removing sequences with length fewer than 3 characters... Done. 96 rows remaining.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 96 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Generating graph plot with nodes colored by CloneSeq...</span></span></code></pre></div>
<p><img src="public_clusters_files/figure-html/unnamed-chunk-14-1.png" width="100%" style="display: block; margin: auto;"></p>
<pre><code><span><span class="co">#&gt;  Done.</span></span></code></pre>
<p>Next, we do the same for the sixth largest cluster:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># focus on cluster 6</span></span>
<span><span class="fu"><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> </span>
<span>    <span class="va">public_clusters</span><span class="op">$</span><span class="va">node_data</span><span class="op">[</span><span class="va">public_clusters</span><span class="op">$</span><span class="va">node_data</span><span class="op">$</span><span class="va">ClusterIDPublic</span> <span class="op">==</span> <span class="fl">6</span>, <span class="op">]</span>,</span>
<span>  seq_col <span class="op">=</span> <span class="st">"CloneSeq"</span>, </span>
<span>  color_nodes_by <span class="op">=</span> <span class="st">"CloneSeq"</span>, </span>
<span>  color_scheme <span class="op">=</span> <span class="st">"plasma"</span>, </span>
<span>  size_nodes_by <span class="op">=</span> <span class="fl">3</span>, </span>
<span>  output_dir <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>  output_name <span class="op">=</span> <span class="st">"Cluster 6"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Input data contains 27 rows.</span></span>
<span><span class="co">#&gt; Removing sequences with length fewer than 3 characters... Done. 27 rows remaining.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 27 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Generating graph plot with nodes colored by CloneSeq...</span></span></code></pre></div>
<p><img src="public_clusters_files/figure-html/unnamed-chunk-15-1.png" width="100%" style="display: block; margin: auto;"></p>
<pre><code><span><span class="co">#&gt;  Done.</span></span></code></pre>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Brian Neal, Hai Yang, Phi Long Le, Li Zhang.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
