<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="NAIR">
<title>Searching for Public TCR/BCR Clusters • NAIR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Searching for Public TCR/BCR Clusters">
<meta property="og:description" content="NAIR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">NAIR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.9040</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/NAIR.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/associated_clusters.html">Searching for Associated TCR/BCR Clusters</a>
    <a class="dropdown-item" href="../articles/buildRepSeqNetwork.html">buildRepSeqNetwork</a>
    <a class="dropdown-item" href="../articles/dual_chain.html">Dual-Chain Network Analysis</a>
    <a class="dropdown-item" href="../articles/network_visualization.html">Network Visualization</a>
    <a class="dropdown-item" href="../articles/public_clusters.html">Searching for Public TCR/BCR Clusters</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Searching for Public TCR/BCR Clusters</h1>
            
      
      
      <div class="d-none name"><code>public_clusters.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The <code>NAIR</code> package includes a set of functions that
facilitate searching for public TCR/BCR clusters across multiple samples
of Adaptive Immune Receptor Repertoire Sequencing (AIRR-Seq) data.</p>
<p>In this context, a public cluster consists of similar TCR/BCR clones
(e.g., those whose CDR3 amino acid sequences differ by at most one amino
acid) that are shared across samples (e.g., across individuals or across
time points for a single individual).</p>
<p>We first provide a brief conceptual overview, followed by a
demonstration in which we explain the process in greater detail.</p>
<div class="section level3">
<h3 id="overview-of-process">Overview of Process<a class="anchor" aria-label="anchor" href="#overview-of-process"></a>
</h3>
<ol style="list-style-type: decimal">
<li>
<a href="#step1"><strong>Filter the clusters within each sample’s
network</strong></a>. For each sample, construct the repertoire network
and use cluster analysis to partition the nodes into clusters. Filter
the data, keeping only those clusters with sufficient node count or
clone count.</li>
<li>
<a href="#step2"><strong>Construct global network from the filtered
data and perform clustering.</strong></a> Combine the filtered data from
all samples into a single global network. Perform cluster analysis and
assign membership to the global clusters. Assess sample representation
within the clusters to identify public clusters.</li>
</ol>
</div>
<div class="section level3">
<h3 id="simulate-data-for-demonstration">Simulate Data for Demonstration<a class="anchor" aria-label="anchor" href="#simulate-data-for-demonstration"></a>
</h3>
<p>We simulate some toy data for demonstration.</p>
<p>We simulate a total of 30 samples, each containing 30
observations.</p>
<p>Some sequences are simulated with a tendency to appear in relatively
few samples, while others are simulated with a tendency to appear in
many samples.</p>
<p>Each sample is saved in a separate file using the .rds file format.
The files are named “<code>Sample1.rds</code>”,
“<code>Sample2.rds</code>”, etc. The file path of their directory is
saved to the R environment variable <code>dir_input_samples</code> for
later reference.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Use temp dir</span></span>
<span><span class="va">data_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Directory to store input files</span></span>
<span><span class="va">dir_input_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"input_samples"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">dir_input_samples</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fl">30</span></span>
<span><span class="va">sample_size</span> <span class="op">&lt;-</span> <span class="fl">30</span> <span class="co"># (seqs per sample)          </span></span>
<span></span>
<span><span class="va">base_seqs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"CASSIEGQLSTDTQYF"</span>, <span class="st">"CASSEEGQLSTDTQYF"</span>, <span class="st">"CASSSVETQYF"</span>,</span>
<span>  <span class="st">"CASSPEGQLSTDTQYF"</span>, <span class="st">"RASSLAGNTEAFF"</span>, <span class="st">"CASSHRGTDTQYF"</span>, <span class="st">"CASDAGVFQPQHF"</span>,</span>
<span>  <span class="st">"CASSLTSGYNEQFF"</span>, <span class="st">"CASSETGYNEQFF"</span>, <span class="st">"CASSLTGGNEQFF"</span>, <span class="st">"CASSYLTGYNEQFF"</span>,</span>
<span>  <span class="st">"CASSLTGNEQFF"</span>, <span class="st">"CASSLNGYNEQFF"</span>, <span class="st">"CASSFPWDGYGYTF"</span>, <span class="st">"CASTLARQGGELFF"</span>,</span>
<span>  <span class="st">"CASTLSRQGGELFF"</span>, <span class="st">"CSVELLPTGPLETSYNEQFF"</span>, <span class="st">"CSVELLPTGPSETSYNEQFF"</span>,</span>
<span>  <span class="st">"CVELLPTGPSETSYNEQFF"</span>, <span class="st">"CASLAGGRTQETQYF"</span>, <span class="st">"CASRLAGGRTQETQYF"</span>,</span>
<span>  <span class="st">"CASSLAGGRTETQYF"</span>, <span class="st">"CASSLAGGRTQETQYF"</span>, <span class="st">"CASSRLAGGRTQETQYF"</span>,</span>
<span>  <span class="st">"CASQYGGGNQPQHF"</span>, <span class="st">"CASSLGGGNQPQHF"</span>, <span class="st">"CASSNGGGNQPQHF"</span>, <span class="st">"CASSYGGGGNQPQHF"</span>,</span>
<span>  <span class="st">"CASSYGGGQPQHF"</span>, <span class="st">"CASSYKGGNQPQHF"</span>, <span class="st">"CASSYTGGGNQPQHF"</span>, </span>
<span>  <span class="st">"CAWSSQETQYF"</span>, <span class="st">"CASSSPETQYF"</span>, <span class="st">"CASSGAYEQYF"</span>, <span class="st">"CSVDLGKGNNEQFF"</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># relative generation probabilities</span></span>
<span><span class="va">pgen</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span></span>
<span>  <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/toeplitz.html" class="external-link">toeplitz</a></span><span class="op">(</span><span class="fl">0.6</span><span class="op">^</span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="op">(</span><span class="va">sample_size</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">1</span>, nrow <span class="op">=</span> <span class="va">samples</span>, ncol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">base_seqs</span><span class="op">)</span> <span class="op">-</span> <span class="va">samples</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Simulate the data</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mlizhangx.github.io/Network-Analysis-for-Repertoire-Sequencing-/">NAIR</a></span><span class="op">)</span></span>
<span><span class="va">foo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulateToyData.html">simulateToyData</a></span><span class="op">(</span>    </span>
<span>  samples <span class="op">=</span> <span class="va">samples</span>, sample_size <span class="op">=</span> <span class="va">sample_size</span>,</span>
<span>  prefix_length <span class="op">=</span> <span class="fl">1</span>, prefix_chars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">""</span>, <span class="st">""</span><span class="op">)</span>,</span>
<span>  prefix_probs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>, <span class="va">samples</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">samples</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  affixes <span class="op">=</span> <span class="va">base_seqs</span>, affix_probs <span class="op">=</span> <span class="va">pgen</span>, num_edits <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  output_dir <span class="op">=</span> <span class="va">dir_input_samples</span>, no_return <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>The first few rows of the data for the first sample appear as
follows:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># View first few rows of data for sample 1</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir_input_samples</span>, <span class="st">"Sample1.rds"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;           CloneSeq CloneFrequency CloneCount SampleID</span></span>
<span><span class="co">#&gt; 1 CASSIEGQLSTDTQYF     0.02606559       2832  Sample1</span></span>
<span><span class="co">#&gt; 2 CASSEEGQLSTDTQYF     0.03718396       4040  Sample1</span></span>
<span><span class="co">#&gt; 3      CASSSPETQYF     0.03182726       3458  Sample1</span></span>
<span><span class="co">#&gt; 4 CASSIEGQLSTDTQYF     0.04615781       5015  Sample1</span></span>
<span><span class="co">#&gt; 5      CAWSSQETQYF     0.06006498       6526  Sample1</span></span>
<span><span class="co">#&gt; 6 CASSEEGQLSTDTQYF     0.03363123       3654  Sample1</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 class="tabset" id="step-1-filter-clusters-within-each-sample">
<a name="step1"></a> Step 1: Filter Clusters Within
Each Sample<a class="anchor" aria-label="anchor" href="#step-1-filter-clusters-within-each-sample"></a>
</h2>
<p>First, we use the <code>findPublicClusters</code> function to perform
network analysis on each sample individually and select clusters based
on node count and clone count.</p>
<p>For each sample, the repertoire network is constructed and cluster
analysis is used to partition the network into clusters. The clusters
are then filtered according to node count and clone count based on
user-specified criteria. The node-level and cluster-level metadata for
the clusters that remain after filtering are saved as files to be used
as inputs for <a href="#step2">step 2</a>.</p>
<p>Below, we explain the function’s usage.</p>
<div class="section level3">
<h3 class="tabset" id="filter-settings">Filter Settings<a class="anchor" aria-label="anchor" href="#filter-settings"></a>
</h3>
<p>The <code>findPublicClusters</code> function has several parameters
that control the criteria used to filter the nodes and clusters in each
sample. These arguments are presented below.</p>
<div class="section level4">
<h4 id="top-n-clusters">Top <span class="math inline">\(n\)</span> Clusters<a class="anchor" aria-label="anchor" href="#top-n-clusters"></a>
</h4>
<p>Within each sample, the clusters are ranked by node count. The top
<span class="math inline">\(n\)</span> highest-ranking clusters (those
with the greatest node count) within each sample are automatically
retained. The default value of <span class="math inline">\(n\)</span> is
20. A different value of <span class="math inline">\(n\)</span> can be
specified using the <code>top_n_clusters</code> argument.</p>
<p>If more than one cluster is tied for <span class="math inline">\(n\)</span>-th place in the ranking by node count,
only one such cluster will be included in the top <span class="math inline">\(n\)</span> clusters. This ensures that no more
than <span class="math inline">\(n\)</span> clusters are selected from
each sample based on this criterion.</p>
<p>If fewer than <span class="math inline">\(n\)</span> clusters are
present in the network for a sample, then all of the clusters will be
retained.</p>
</div>
<div class="section level4">
<h4 id="minimum-node-count">Minimum Node Count<a class="anchor" aria-label="anchor" href="#minimum-node-count"></a>
</h4>
<p>In addition to retaining the top <span class="math inline">\(n\)</span> clusters from each sample, clusters
that contain a sufficient number of nodes will also be retained. By
default, any cluster containing at least ten nodes will be retained.
This value can be adjusted using the <code>min_node_count</code>
argument. For example, setting <code>min_clone_count = 30</code> will
retain all clusters containing at least 30 nodes.</p>
</div>
<div class="section level4">
<h4 id="minimum-clone-count">Minimum Clone Count<a class="anchor" aria-label="anchor" href="#minimum-clone-count"></a>
</h4>
<p>In addition to the clusters retained based on node count, clusters
with a sufficient aggregate clone count will also be retained. The
aggregate clone count of a cluster is the sum of the clone counts across
all nodes (clones) in the cluster.</p>
<p>By default, any cluster with an aggregate clone count of at least 100
will be retained. This value can be adjusted using the
<code>min_clone_count</code> argument. For example, setting
<code>min_clone_count = 500</code> will retain all clusters with an
aggregate clone count of at least 500.</p>
<p>Note that clusters are only selected according to this criterion if
clone counts are provided. This is done using the <code>count_col</code>
argument as explained in the <strong>Specifying the Count
Column</strong> subsection.</p>
</div>
<div class="section level4">
<h4 id="sequence-length">Sequence Length<a class="anchor" aria-label="anchor" href="#sequence-length"></a>
</h4>
<p>When building the network for each sample, only clones whose receptor
sequences are at least three characters in length will be included in
the network. This minimum value for sequence length can be adjusted by
setting the <code>min_seq_length</code> argument to a different value.
Setting the value to <code>NULL</code> bypasses this check.</p>
</div>
<div class="section level4">
<h4 id="sequence-content">Sequence Content<a class="anchor" aria-label="anchor" href="#sequence-content"></a>
</h4>
<p>When building the network for each sample, all clones whose receptor
sequences contain characters <code>*</code> or <code>_</code> will be
omitted from the network. This can be changed using the
<code>drop_matches</code> argument, which accommodates a character
string or regular expression specifying the pattern of content to search
for. The content of each clone’s sequence is checked for a match to this
pattern using the <code>grep</code> function from base R. If a match is
found, the clone is omitted from the network. Setting the value to
<code>NULL</code> bypasses this check.</p>
<p>For details on how the pattern matching is performed, please refer to
the base R documentation files for <code>regex</code> and
<code>grep</code>.</p>
</div>
</div>
<div class="section level3">
<h3 id="input-file-paths">Input File Paths<a class="anchor" aria-label="anchor" href="#input-file-paths"></a>
</h3>
<p>The main argument of the <code>findPublicClusters</code> function is
the <code>file_list</code> argument, which accepts a vector containing
file paths. Each path corresponds to a distinct AIRR-Seq data file
representing an individual sample.</p>
<p>Below, we prepare the vector <code>input_files</code> to be provided
to the <code>file_list</code> argument of
<code>findPublicClusters</code>:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># input files for step 1 (one per sample)</span></span>
<span><span class="va">input_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir_input_samples</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Sample"</span>, <span class="fl">1</span><span class="op">:</span><span class="va">samples</span>, <span class="st">".rds"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">input_files</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "/tmp/Rtmp4keLMW/input_samples/Sample1.rds"</span></span>
<span><span class="co">#&gt; [2] "/tmp/Rtmp4keLMW/input_samples/Sample2.rds"</span></span>
<span><span class="co">#&gt; [3] "/tmp/Rtmp4keLMW/input_samples/Sample3.rds"</span></span>
<span><span class="co">#&gt; [4] "/tmp/Rtmp4keLMW/input_samples/Sample4.rds"</span></span>
<span><span class="co">#&gt; [5] "/tmp/Rtmp4keLMW/input_samples/Sample5.rds"</span></span>
<span><span class="co">#&gt; [6] "/tmp/Rtmp4keLMW/input_samples/Sample6.rds"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="input-file-format">Input File Format<a class="anchor" aria-label="anchor" href="#input-file-format"></a>
</h3>
<p>The file format of the input files for
<code>findPublicClusters</code> is specified using the
<code>input_type</code> parameter. The supported formats include
<code>"rds"</code>, <code>"rda"</code>, <code>"csv"</code>, as well as
files that can be read using the <code>read.table</code> function, such
as <code>"tsv"</code> and <code>"txt"</code>.</p>
<p>For text formats such as <code>"csv"</code>, <code>"tsv"</code> and
<code>"txt"</code>, users can specify the separation option by utilizing
the <code>sep</code> argument. The default setting <code>sep = ""</code>
accommodates all forms of white space, i.e., one or more spaces, tabs,
newlines or carriage returns. In addition, it is important to note that
the first line of the data is assumed to be the header by default. To
disable this behavior and treat the first line as data, users must set
the <code>header</code> parameter to <code>FALSE</code>.</p>
<p>For the <code>"rda"</code> (Rdata) format, the
<code>data_symbols</code> argument is used to specify the name of each
sample’s AIRR-Seq data frame within the Rdata file (i.e., the name of
the data frame as it will appear in the R environment when the Rdata
file is loaded). The argument accepts a character string. This requires
that each sample’s data frame has the same name within its respective
Rdata file. For example, if each sample’s data frame is named
<code>mydataframe</code> and saved using the command
<code>save(mydataframe, file = myfilepath)</code>, where the file path
<code>myfilepath</code> is different for each sample, then we would set
<code>data_symbols = "mydataframe"</code>.</p>
<p>Our samples are stored in .rds files, so we use
<code>input_type = "rds"</code>.</p>
</div>
<div class="section level3">
<h3 id="specifying-the-sequence-column">Specifying the Sequence Column<a class="anchor" aria-label="anchor" href="#specifying-the-sequence-column"></a>
</h3>
<p>Each clone’s TCR/BCR sequence is used as the basis of similarity
between clones in the network.</p>
<p>The <code>seq_col</code> argument is used to specify the column
containing the clone sequences in the input data for each sample. The
argument accepts either the column name or column index.</p>
<p>In our simulated data, the column containing the clone sequences is
named <code>CloneSeq</code>.</p>
</div>
<div class="section level3">
<h3 id="specifying-the-count-column">Specifying the Count Column<a class="anchor" aria-label="anchor" href="#specifying-the-count-column"></a>
</h3>
<p>The clone count for each clone is used to filter the clusters within
each sample.</p>
<p>The <code>count_col</code> argument is used to specify the column
containing the clone counts in the input data for each sample. The
argument accepts either the column name or column index.</p>
<p>In our simulated data, the column containing the clone sequences is
named <code>CloneCount</code>.</p>
</div>
<div class="section level3">
<h3 id="assigning-custom-sample-ids">Assigning Custom Sample IDs<a class="anchor" aria-label="anchor" href="#assigning-custom-sample-ids"></a>
</h3>
<p>The sample ID for each clone will be included in the filtered data
saved by <code>findPublicClusters</code>. This will allow us to
distinguish the sample of origin for each clone after we combine the
filtered data from all samples in <a href="#step2">step 2</a>.</p>
<p>By default, the samples are labeled numerically according to the
order they appear in <code>file_list</code>. The <code>sample_ids</code>
argument allows for custom sample IDs to be assigned, if desired. The
argument accepts a vector of the same length as <code>file_list</code>.
Each entry of <code>sample_ids</code> is assigned as a sample ID to the
sample in the corresponding entry of <code>file_list</code>.</p>
</div>
<div class="section level3">
<h3 class="tabset" id="network-settings-per-sample">Network Settings Per Sample<a class="anchor" aria-label="anchor" href="#network-settings-per-sample"></a>
</h3>
<p>The settings used to construct the network for each sample can be
customized using the arguments below.</p>
<div class="section level4">
<h4 id="distance-function">Distance Function<a class="anchor" aria-label="anchor" href="#distance-function"></a>
</h4>
<p>The default method for measuring the similarity between TCR/BCR
sequences is the Hamming distance. It calculates the number of
differences between two sequences of the same length. If the sequences
have different lengths, the shorter sequence is extended by adding
non-matching characters to make it the same length as the longer
sequence.</p>
<p>The Levenshtein distance can be used as an alternative measurement to
determine the similarity between sequences. It calculates the minimum
number of single-character edits (insertions, deletions and
transformations) needed to transform one sequence into the other. This
method is particularly useful for comparing sequences of different
lengths and can account for insertions and deletions. When constructing
a network based on the similarity of CDR-3 nucleotide sequences, it is
preferable to use the Levenshtein distance instead of the default
Hamming distance by setting the argument
<code>dist_type = "levenshtein"</code>. However, the Levenshtein
distance requires significantly more computational time than the Hamming
distance, which may be challenging when working with large data sets
having long TCR/BCR sequences.</p>
</div>
<div class="section level4">
<h4 id="distance-cutoff">Distance Cutoff<a class="anchor" aria-label="anchor" href="#distance-cutoff"></a>
</h4>
<p>The distance function specified in the <code>dist_type</code>
argument (Hamming distance by default) is used to quantify the
similarity between TCR/BCR sequences. The chosen distance measurement
determines the distance between two nodes in the network graph.</p>
<p>By default, two nodes in the graph are connected by an edge if their
distance is at most 1. However, if users want to adjust this cutoff, the
<code>dist_cutoff</code> argument can be set to a different value.</p>
<p>For example, if <code>dist_cutoff = 2</code>, then two nodes will be
connected by an edge if their distance is at most 2. The cutoff value
controls the stringency of the network construction and affects the
number and density of edges in the network.</p>
</div>
<div class="section level4">
<h4 id="keepremove-isolated-nodes">Keep/Remove Isolated Nodes<a class="anchor" aria-label="anchor" href="#keepremove-isolated-nodes"></a>
</h4>
<p>By default, if a given node has no edges connecting it to any other
nodes in the network, it will be removed from the network graph and will
not be included in the output.</p>
<p>If desired, all nodes can be kept in the network, including those
that do not have any edge connections to other nodes. This is
accomplished by setting the <code>drop_isolated_nodes</code> argument to
<code>FALSE</code>.</p>
</div>
</div>
<div class="section level3">
<h3 id="clustering-algorithm">Clustering Algorithm<a class="anchor" aria-label="anchor" href="#clustering-algorithm"></a>
</h3>
<p>By default, clustering is performed using the
<code>cluster_fast_greedy</code> algorithm from the <code>igraph</code>
package. Other clustering algorithms from the <code>igraph</code>
package can be used instead of the default algorithm. The algorithm is
specified using the <code>cluster_fun</code> argument, which accepts one
of the following functions:</p>
<ul>
<li><code>cluster_fast_greedy</code></li>
<li><code>cluster_edge_betweenness</code></li>
<li><code>cluster_fluid_communities</code></li>
<li><code>cluster_infomap</code></li>
<li><code>cluster_label_prop</code></li>
<li><code>cluster_leading_eigen</code></li>
<li><code>cluster_leiden</code></li>
<li><code>cluster_louvain</code></li>
<li><code>cluster_optimal</code></li>
<li><code>cluster_spinglass</code></li>
<li><code>cluster_walktrap</code></li>
</ul>
<p>For example, setting <code>cluster_fun = cluster_leiden</code>
performs clustering using the <code>cluster_leiden</code> algorithm.</p>
<p>For more information about a particular algorithm, users can refer to
its help documentation file. For example, the command
<code><a href="https://r.igraph.org/reference/cluster_fast_greedy.html" class="external-link">?igraph::cluster_fast_greedy</a></code> loads the documentation file
for the <code>cluster_fast_greedy</code> algorithm.</p>
</div>
<div class="section level3">
<h3 class="tabset" id="output-settings">Output Settings<a class="anchor" aria-label="anchor" href="#output-settings"></a>
</h3>
<p>The <code>findPublicClusters</code> function does not return any
direct output. Instead, it saves the network metadata for the selected
clusters to files that will be used as inputs in step 2.</p>
<p>For each sample, <code>findPublicClusters</code> saves two files, one
containing the node-specific network metadata and the other containing
the cluster-specific network metadata. The data for each sample is saved
after filtering the clusters, and thus includes only the clusters
selected based on the specified filter settings.</p>
<div class="section level4">
<h4 id="subsetting-columns-from-the-input-data">Subsetting Columns From the Input Data<a class="anchor" aria-label="anchor" href="#subsetting-columns-from-the-input-data"></a>
</h4>
<p>The node-specific network metadata for each sample includes all of
the data columns that were present in the sample’s original input data
frame.</p>
<p>In some cases, the user may not wish for all columns to be included.
In this case, the user can specify the columns to be included using the
<code>subset_cols</code> argument, which accepts a vector containing
either the column names or the column indices of the columns to be
kept.</p>
<p>The column specified by the <code>seq_col</code> argument will
automatically be included regardless of the value of
<code>subset_cols</code>.</p>
</div>
<div class="section level4">
<h4 id="output-directory">Output Directory<a class="anchor" aria-label="anchor" href="#output-directory"></a>
</h4>
<p>The file path for the output directory is specified using the
<code>output_dir</code> argument. The output directory specified by the
<code>output_dir</code> argument will be created if it does not already
exist.</p>
<p>Within the output directory, two subdirectories are created. One
subdirectory, named <code>node_meta_data</code>, contains the node-level
data files for each sample, and the other, named
<code>cluster_meta_data</code>, contains the cluster-level data files.
Within each of these two subdirectories, the file for each sample is
named according to its sample ID as specified by the
<code>sample_ids</code> argument (see the <strong>Assigning Custom
Sample IDs</strong> subsection above).</p>
</div>
<div class="section level4">
<h4 id="output-file-format">Output File Format<a class="anchor" aria-label="anchor" href="#output-file-format"></a>
</h4>
<p>By default, each file is saved as a RDS file. This can be changed
using the <code>output_type</code> argument. Other valid options include
<code>"rda"</code> and <code>"csv"</code>. For the <code>"rda"</code>
(Rdata) format, each data frame containing node-level data will be named
<code>ndat</code> in the R environment when its Rdata file is loaded,
while each data frame containing cluster-level data will be named
<code>cdat</code>.</p>
</div>
<div class="section level4">
<h4 id="saving-unfiltered-network-data">Saving Unfiltered Network Data<a class="anchor" aria-label="anchor" href="#saving-unfiltered-network-data"></a>
</h4>
<p>By default, the <code>findPublicClusters</code> function saves the
network metadata only for the clusters selected based on the specified
filter settings.</p>
<p>If desired, the network metadata for each sample’s full network can
also be saved prior to filtering the clusters. This is done by by
providing a file path to the <code>output_dir_unfiltered</code>
argument, which specifies a separate output directory for the full
network data. Each sample’s full network data prior to filtering the
clusters will then be saved to the directory specified by the
<code>output_dir_unfiltered</code> argument. This data is saved
separately from, and in addition to, the default data that is saved
after filtering the clusters.</p>
<p>Note that the sequence-based filter settings specified by
<code>min_seq_length</code> and <code>drop_chars</code> still apply to
each sample’s full network, since the network is only constructed after
applying these filters. The full, pre-filtered network refers to the
network that contains all of the sample’s clusters, i.e., before the
public clusters are identified and other clusters are removed.</p>
<p>The full network data for each sample includes node-level metadata,
cluster-level metadata, the <code>igraph</code> edge list, the network
adjacency matrix, as well as any plots generated (see next subsection).
By default, the R objects for these files are saved into a single RDS
file whose file name (excluding the .rds file extension) is the sample
ID (as specified by the <code>sample_ids</code> argument). The RDA file
format can be used instead of RDS by setting
<code>output_type_unfiltered = "rda"</code>. If desired, each R object
can be saved to a separate file by setting
<code>output_type_unfiltered = "individual"</code>. This saves each
object according to the <a href="buildRepSeqNetwork.html#output_settings">default file format used
by <code>buildRepSeqNetwork</code></a>. When saving objects
individually, the sample ID is used as the common file name prefix for
the files from each sample.</p>
</div>
</div>
<div class="section level3">
<h3 id="optional-network-visualization-per-sample">(Optional) Network Visualization Per Sample<a class="anchor" aria-label="anchor" href="#optional-network-visualization-per-sample"></a>
</h3>
<p>By default, <code>findPublicClusters</code> does not produce visual
plots when constructing the network for each sample. Instead,
visualization occurs after combining the data from all samples into a
single network in <a href="#step2">step 2</a>.</p>
<p>If desired, visual plots of the full network for each sample (prior
to filtering the clusters) can be produced by setting
<code>plots = TRUE</code>. The plots can be printed in R by additionally
setting <code>print_plots = TRUE</code>. Furthermore, if the user has
specified to save the full network data for each sample (prior to
filtering the clusters) by providing a file path to the
<code>output_type_unfiltered</code> argument, these plots will be saved
along with the rest of the full network data for each sample.</p>
<p>Note that the plots are not saved along with the default data that is
saved after filtering the clusters. Therefore, if no file path is
provided to the <code>output_type_unfiltered</code> argument, the plots
will not be saved at all. If, in addition, the argument
<code>print_plots</code> is set to <code>FALSE</code> (the default),
setting <code>plots = TRUE</code> will have no effect.</p>
<p>If the user sets <code>plots</code> to <code>TRUE</code>, then by
default, the plot for each sample will color the nodes according to
their cluster membership. If desired, a different variable can be used
to color the nodes. This is done by specifying the column for the
variable to the <code>color_nodes_by</code> argument, which accepts a
column name or column index.</p>
<p><a href="network_visualization.html">This vignette</a> details
additional arguments that can be used to further customize the visual
plot for each sample.</p>
</div>
<div class="section level3">
<h3 id="example-function-call">Example Function Call<a class="anchor" aria-label="anchor" href="#example-function-call"></a>
</h3>
<p>We execute the <code>findPublicClusters</code> function using the
input we prepared earlier for the <code>file_list</code> argument.</p>
<p>The file path of our output directory is saved to the R variable
named <code>dir_filtered_samples</code> for later reference.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 1. Filter Clusters Within Each Sample</span></span>
<span><span class="va">dir_filtered_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"filtered_samples"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/findPublicClusters.html">findPublicClusters</a></span><span class="op">(</span></span>
<span>  file_list <span class="op">=</span> <span class="va">input_files</span>, input_type <span class="op">=</span> <span class="st">"rds"</span>,</span>
<span>  sample_ids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Sample"</span>, <span class="fl">1</span><span class="op">:</span><span class="va">samples</span><span class="op">)</span>,</span>
<span>  seq_col <span class="op">=</span> <span class="st">"CloneSeq"</span>, count_col <span class="op">=</span> <span class="st">"CloneCount"</span>,</span>
<span>  min_seq_length <span class="op">=</span> <span class="cn">NULL</span>, drop_matches <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  top_n_clusters <span class="op">=</span> <span class="fl">3</span>, min_node_count <span class="op">=</span> <span class="fl">5</span>, min_clone_count <span class="op">=</span> <span class="fl">15000</span>,</span>
<span>  output_dir <span class="op">=</span> <span class="va">dir_filtered_samples</span><span class="op">)</span></span>
<span><span class="co">#&gt; &lt;&lt;&lt; Beginning search for public clusters &gt;&gt;&gt;</span></span>
<span><span class="co">#&gt; Processing sample 1 of 30: Sample1</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 30 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 7 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt; * 3 clusters (18 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 2 of 30: Sample2</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 28 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 6 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt; * 4 clusters (22 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 3 of 30: Sample3</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 29 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 6 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt; * 3 clusters (20 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 4 of 30: Sample4</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 28 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 6 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt; * 3 clusters (18 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 5 of 30: Sample5</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 27 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 6 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt; * 4 clusters (22 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 6 of 30: Sample6</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 29 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 9 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt; * 3 clusters (15 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 7 of 30: Sample7</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 26 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 8 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt; * 3 clusters (13 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 8 of 30: Sample8</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 26 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 9 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt; * 3 clusters (12 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 9 of 30: Sample9</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 26 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 8 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt; * 3 clusters (13 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 10 of 30: Sample10</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 28 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 9 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt; * 3 clusters (15 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 11 of 30: Sample11</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 26 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 8 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt; * 4 clusters (18 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 12 of 30: Sample12</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 29 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 10 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt; * 3 clusters (12 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 13 of 30: Sample13</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 27 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 8 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt; * 4 clusters (18 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 14 of 30: Sample14</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 26 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 7 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt; * 3 clusters (17 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 15 of 30: Sample15</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 29 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 8 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt; * 3 clusters (14 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 16 of 30: Sample16</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 26 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 6 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt; * 4 clusters (21 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 17 of 30: Sample17</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 29 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 7 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt; * 4 clusters (22 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 18 of 30: Sample18</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 28 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 9 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt; * 3 clusters (14 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 19 of 30: Sample19</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 28 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 8 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt; * 3 clusters (17 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 20 of 30: Sample20</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 26 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 8 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt; * 3 clusters (15 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 21 of 30: Sample21</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 28 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 8 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt; * 4 clusters (19 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 22 of 30: Sample22</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 28 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 8 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt; * 3 clusters (16 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 23 of 30: Sample23</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 27 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 7 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt; * 4 clusters (19 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 24 of 30: Sample24</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 27 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 8 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt; * 4 clusters (18 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 25 of 30: Sample25</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 28 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 9 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt; * 4 clusters (17 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 26 of 30: Sample26</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 28 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 7 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt; * 3 clusters (17 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 27 of 30: Sample27</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 25 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 5 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt; * 3 clusters (18 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 28 of 30: Sample28</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 26 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 7 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt; * 3 clusters (16 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 29 of 30: Sample29</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 29 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 7 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt; * 4 clusters (22 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; Processing sample 30 of 30: Sample30</span></span>
<span><span class="co">#&gt; Input data contains 30 rows.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 28 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 6 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Filtering clusters in the current sample... Done.</span></span>
<span><span class="co">#&gt; * 3 clusters (19 nodes) remain. Saving results... Done.</span></span>
<span><span class="co">#&gt; ----------------------------------------------------------------------</span></span>
<span><span class="co">#&gt; All samples complete. Filtered data is located in the following directory:</span></span>
<span><span class="co">#&gt;   /tmp/Rtmp4keLMW/filtered_samples</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 class="tabset" id="step-2-global-network-of-public-clusters">
<a name="step2"></a> Step 2: Global Network of Public
Clusters<a class="anchor" aria-label="anchor" href="#step-2-global-network-of-public-clusters"></a>
</h2>
<p>Next, we use the <code>buildPublicClusterNetwork</code> function to
combine the filtered data from all samples into a single global network
and perform clustering analysis.</p>
<div class="section level3">
<h3 id="input-file-paths-1">Input File Paths<a class="anchor" aria-label="anchor" href="#input-file-paths-1"></a>
</h3>
<p>The files created by <code>findPublicClusters</code> in the previous
step contain the filtered data for each sample. As detailed earlier,
these files are located in two separate subdirectories, one containing
the files for the node-level metadata and the other containing the files
for the cluster-level metadata.</p>
<p>For this step, we require only the files containing the node-level
metadata. These files are provided to
<code>buildPublicClusterNetwork</code> by supplying a character vector
of file paths to the <code>file_list</code> argument. We create this
vector below.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Node-level metadata for each sample's filtered clusters</span></span>
<span><span class="va">dir_filtered_samples_node</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir_filtered_samples</span>, <span class="st">"node_meta_data"</span><span class="op">)</span></span>
<span><span class="va">files_filtered_samples_node</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">dir_filtered_samples_node</span>, </span>
<span>                                          full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">files_filtered_samples_node</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "/tmp/Rtmp4keLMW/filtered_samples/node_meta_data/Sample1.rds" </span></span>
<span><span class="co">#&gt; [2] "/tmp/Rtmp4keLMW/filtered_samples/node_meta_data/Sample10.rds"</span></span>
<span><span class="co">#&gt; [3] "/tmp/Rtmp4keLMW/filtered_samples/node_meta_data/Sample11.rds"</span></span>
<span><span class="co">#&gt; [4] "/tmp/Rtmp4keLMW/filtered_samples/node_meta_data/Sample12.rds"</span></span>
<span><span class="co">#&gt; [5] "/tmp/Rtmp4keLMW/filtered_samples/node_meta_data/Sample13.rds"</span></span>
<span><span class="co">#&gt; [6] "/tmp/Rtmp4keLMW/filtered_samples/node_meta_data/Sample14.rds"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="input-file-format-1">Input File Format<a class="anchor" aria-label="anchor" href="#input-file-format-1"></a>
</h3>
<p>By default, <code>findPublicClusters</code> saves its output files
using the RDS file format. By default,
<code>buildPublicClusterNetwork</code> assumes that the input files are
in the RDS file format.</p>
<p>If the files created by <code>findPublicClusters</code> were saved
using a different file format, the <code>input_type</code> argument can
be used to specify the format used. This argument has the same behavior
as seen in <code>findPublicClusters</code>.</p>
<p>For the <code>"rda"</code> (Rdata) file format, the
<code>data_symbols</code> argument is used to specify the name of each
sample’s node-level data frame within its respective Rdata file. If
<code>findPublicClusters</code> was called with
<code>output_type = "rda"</code>, then the node-level data frame for
each sample will be named <code>ndat</code>. Accordingly, the default
value of <code>data_symbols</code> is <code>ndat</code>, so the user
should not need to manually specify the value of
<code>data_symbols</code> unless the names within the files are
changed.</p>
</div>
<div class="section level3">
<h3 id="specifying-the-sequence-and-count-columns">Specifying the Sequence and Count Columns<a class="anchor" aria-label="anchor" href="#specifying-the-sequence-and-count-columns"></a>
</h3>
<p>The <code>seq_col</code> and <code>count_col</code> arguments are
used to specify the columns containing the receptor sequences and clone
counts, respectively, in the same manner seen earlier for
<code>findPublicClusters</code>.</p>
<p>The clone counts are used to compute some of the cluster-level
properties, which will be omitted if <code>count_col</code> is left
unspecified.</p>
</div>
<div class="section level3">
<h3 id="network-settings">Network Settings<a class="anchor" aria-label="anchor" href="#network-settings"></a>
</h3>
<p>The settings used to construct the network can be customized <a href="buildRepSeqNetwork.html#network_settings">using the same arguments
from <code>findPublicClusters</code></a>, shown below with their default
values:</p>
<ul>
<li><code>dist_type = "hamming"</code></li>
<li><code>dist_cutoff = 1</code></li>
<li><code>drop_isolated_nodes = FALSE</code></li>
</ul>
<p>Refer to <a href="buildRepSeqNetwork.html#network_settings">this
vignette</a> for details on their meaning and behavior.</p>
</div>
<div class="section level3">
<h3 id="clustering-algorithm-1">Clustering Algorithm<a class="anchor" aria-label="anchor" href="#clustering-algorithm-1"></a>
</h3>
<p>After constructing the global network,
<code>buildPublicClusterNetwork</code> performs cluster analysis on the
network nodes, partitioning the global network graph into
densely-connected subgraphs. These global clusters can contain nodes
from different samples.</p>
<p>By default, clustering is performed using the
<code>cluster_fast_greedy</code> algorithm from the <code>igraph</code>
package. Other clustering algorithms from the <code>igraph</code>
package can be used instead of the default algorithm. The algorithm is
specified using the <code>cluster_fun</code> argument, which accepts one
of the following functions:</p>
<ul>
<li><code>cluster_fast_greedy</code></li>
<li><code>cluster_edge_betweenness</code></li>
<li><code>cluster_fluid_communities</code></li>
<li><code>cluster_infomap</code></li>
<li><code>cluster_label_prop</code></li>
<li><code>cluster_leading_eigen</code></li>
<li><code>cluster_leiden</code></li>
<li><code>cluster_louvain</code></li>
<li><code>cluster_optimal</code></li>
<li><code>cluster_spinglass</code></li>
<li><code>cluster_walktrap</code></li>
</ul>
<p>For example, setting <code>cluster_fun = cluster_leiden</code>
performs clustering using the <code>cluster_leiden</code> algorithm.</p>
<p>For more information about a particular algorithm, users can refer to
its help documentation file. For example, the command
<code><a href="https://r.igraph.org/reference/cluster_fast_greedy.html" class="external-link">?igraph::cluster_fast_greedy</a></code> loads the documentation file
for the <code>cluster_fast_greedy</code> algorithm.</p>
</div>
<div class="section level3">
<h3 id="customization-of-visual-plot">Customization of Visual Plot<a class="anchor" aria-label="anchor" href="#customization-of-visual-plot"></a>
</h3>
<p>By default, the network graph plot produced by
<code>buildPublicClusterNetwork</code> colors the nodes according to
sample ID. This can assist the user in identifying the public clusters.
If desired, a different variable can be used to color the nodes. This is
done by specifying the column for the variable to the
<code>color_nodes_by</code> argument, which accepts a column name or
column index.</p>
<p><a href="network_visualization.html">This vignette</a> details
additional arguments that can be used to further customize the visual
plot.</p>
</div>
<div class="section level3">
<h3 class="tabset" id="output-settings-1">Output Settings<a class="anchor" aria-label="anchor" href="#output-settings-1"></a>
</h3>
<p>The output returned by <code>buildPublicClusterNetwork</code> follows
the <a href="buildRepSeqNetwork.html#output_settings">same format as the
output of <code>buildRepSeqNetwork</code></a>. The function returns a
list containing the node-level and cluster-level metadata for the global
network, as well as any plots generated, in addition to the network
adjacency matrix and the <code>igraph</code> network edge list.</p>
<div class="section level4">
<h4 id="output-directory-1">Output Directory<a class="anchor" aria-label="anchor" href="#output-directory-1"></a>
</h4>
<p>By default, a subdirectory named <code>public_clusters</code> is
created within the current working directory, and the contents of the
list returned by <code>buildPublicClusterNetwork</code> are saved to
this subdirectory.</p>
<p>The output can be saved to a different directory by providing a file
path to the <code>output_dir</code> argument.</p>
<p>The user can also specify <code>output_dir = NULL</code> in order to
prevent the output from being saved.</p>
</div>
<div class="section level4">
<h4 id="output-file-format-1">Output File Format<a class="anchor" aria-label="anchor" href="#output-file-format-1"></a>
</h4>
<p>By default, each element of the list returned by
<code>buildPublicClusterNetwork</code> is saved as an individual file.
The file formats are the same default file formats used by <a href="buildRepSeqNetwork.html#output_settings"><code>buildRepSeqNetwork</code></a>.
In particular, the node-level and cluster-level metadata are saved as
csv files.</p>
<p>Alternatively, the user can save the entire output list to a single
compressed rds or rda file by setting <code>output_type = "rds"</code>
or <code>output_type = "rda"</code>, respectively.</p>
</div>
<div class="section level4">
<h4 id="pdf-of-plots">PDF of Plots<a class="anchor" aria-label="anchor" href="#pdf-of-plots"></a>
</h4>
<ul>
<li>Regardless of the value of <code>output_type</code>, a separate pdf
file containing the graph plot(s) is created in <code>output_dir</code>.
The dimensions (in inches) for the pdf can be adjusted using the
<code>plot_width</code> and <code>plot_height</code> arguments, with the
defaults being <code>12</code> and <code>10</code>, respectively. The
pdf file is not created if no plots are generated (i.e., if the argument
<code>plots</code> is set to <code>FALSE</code>).</li>
<li>
<strong>Note:</strong> the <code>ggraph</code> object for each plot
is only saved if the user sets <code>output_type = "rds"</code> or
<code>output_type = "rda"</code>. Using one of these settings is
recommended used if the user wishes to modify any plots in the future.
Note, however, that plots can always be re-generated from the node-level
meta data using the <code>generateGraphPlot</code> function as described
<a href="network_visualization.html#new_plots">here</a>.</li>
</ul>
</div>
<div class="section level4">
<h4 id="output-file-names">Output File Name(s)<a class="anchor" aria-label="anchor" href="#output-file-names"></a>
</h4>
<ul>
<li>By default, the name of each saved file starts with
<code>PublicClusterNetwork</code>. This prefix can be changed by
providing a character string to the <code>output_name</code>
argument.</li>
<li>If <code>output_type</code> is set to <code>"rds"</code> or
<code>"rda"</code>, then the name of the file will be the value of the
<code>output_name</code> argument followed by the appropriate file
extension (either <code>.rds</code> or <code>.rda</code>).</li>
<li>The name of the pdf file for the graph plot will be the value of the
<code>output_name</code> argument followed by the <code>.pdf</code> file
extension.</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 class="tabset" id="example-call-and-output">Example Call and Output<a class="anchor" aria-label="anchor" href="#example-call-and-output"></a>
</h3>
<p>We execute the <code>buildPublicClusterNetwork</code> function using
the input we prepared earlier for the <code>file_list</code>
argument:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dir_out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">data_dir</span>, <span class="st">"public_clusters"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Collect clones from all public clusters and perform network analysis</span></span>
<span><span class="va">public_clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/buildPublicClusterNetwork.html">buildPublicClusterNetwork</a></span><span class="op">(</span></span>
<span>  file_list <span class="op">=</span> <span class="va">files_filtered_samples_node</span>,</span>
<span>  seq_col <span class="op">=</span> <span class="st">"CloneSeq"</span>, count_col <span class="op">=</span> <span class="st">"CloneCount"</span>,</span>
<span>  size_nodes_by <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  output_dir <span class="op">=</span> <span class="va">dir_out</span><span class="op">)</span></span>
<span><span class="co">#&gt; Building network of public clusters:</span></span>
<span><span class="co">#&gt; Input data contains 517 rows.</span></span>
<span><span class="co">#&gt; Removing sequences with length fewer than 3 characters... Done. 517 rows remaining.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 517 nodes.</span></span>
<span><span class="co">#&gt; Computing cluster membership within the network... Done.</span></span>
<span><span class="co">#&gt; Computing node-level network statistics... Done.</span></span>
<span><span class="co">#&gt; Computing statistics for the 20 clusters in the network... Done.</span></span>
<span><span class="co">#&gt; Generating graph plot with nodes colored by SampleID...</span></span>
<span><span class="co">#&gt;  Done.</span></span>
<span><span class="co">#&gt; Node-level meta-data saved to file:</span></span>
<span><span class="co">#&gt;   /tmp/Rtmp4keLMW/public_clusters/PublicClusterNetwork_NodeMetadata.csv</span></span>
<span><span class="co">#&gt; Cluster-level meta-data saved to file:</span></span>
<span><span class="co">#&gt;   /tmp/Rtmp4keLMW/public_clusters/PublicClusterNetwork_ClusterMetadata.csv</span></span></code></pre></div>
<p><img src="public_clusters_files/figure-html/unnamed-chunk-6-1.png" width="100%" style="display: block; margin: auto;"></p>
<pre><code><span><span class="co">#&gt; Network graph plots saved to file:</span></span>
<span><span class="co">#&gt;   /tmp/Rtmp4keLMW/public_clusters/PublicClusterNetwork.pdf</span></span>
<span><span class="co">#&gt; Network igraph saved in edgelist format to file:</span></span>
<span><span class="co">#&gt;   /tmp/Rtmp4keLMW/public_clusters/PublicClusterNetwork_EdgeList.txt</span></span>
<span><span class="co">#&gt; Adjacency matrix saved to file:</span></span>
<span><span class="co">#&gt;   /tmp/Rtmp4keLMW/public_clusters/PublicClusterNetwork_AdjacencyMatrix.mtx</span></span></code></pre>
<p>The default files saved by <code>buildPublicClusterNetwork</code>
have the following names:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Files saved by buildPublicClusterNetwork</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">dir_out</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "PublicClusterNetwork_AdjacencyMatrix.mtx"</span></span>
<span><span class="co">#&gt; [2] "PublicClusterNetwork_ClusterMetadata.csv"</span></span>
<span><span class="co">#&gt; [3] "PublicClusterNetwork_EdgeList.txt"       </span></span>
<span><span class="co">#&gt; [4] "PublicClusterNetwork_NodeMetadata.csv"   </span></span>
<span><span class="co">#&gt; [5] "PublicClusterNetwork.pdf"</span></span></code></pre></div>
<p>The elements of the output list returned by
<code>buildPublicClusterNetwork</code> have the following names:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># output returned by buildPublicClusterNetwork</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">public_clusters</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "igraph"           "adjacency_matrix" "node_data"        "cluster_data"    </span></span>
<span><span class="co">#&gt; [5] "plots"</span></span></code></pre></div>
<p>Below, we describe each element of the above list.</p>
<div class="section level4">
<h4 id="node_data">
<code>node_data</code><a class="anchor" aria-label="anchor" href="#node_data"></a>
</h4>
<p>The <code>node_data</code> data frame contained in the output list
contains one row for each node in the network, with the following
variables:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># variables in the node-level metadata</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">public_clusters</span><span class="op">$</span><span class="va">node_data</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "CloneSeq"                           "CloneFrequency"                    </span></span>
<span><span class="co">#&gt;  [3] "CloneCount"                         "SampleID"                          </span></span>
<span><span class="co">#&gt;  [5] "SampleLevelNetworkDegree"           "ClusterIDInSample"                 </span></span>
<span><span class="co">#&gt;  [7] "SampleLevelTransitivity"            "PublicCloseness"                   </span></span>
<span><span class="co">#&gt;  [9] "SampleLevelCentralityByCloseness"   "SampleLevelEigenCentrality"        </span></span>
<span><span class="co">#&gt; [11] "SampleLevelCentralityByEigen"       "SampleLevelBetweenness"            </span></span>
<span><span class="co">#&gt; [13] "SampleLevelCentralityByBetweenness" "SampleLevelAuthorityScore"         </span></span>
<span><span class="co">#&gt; [15] "SampleLevelCoreness"                "SampleLevelPageRank"               </span></span>
<span><span class="co">#&gt; [17] "PublicNetworkDegree"                "ClusterIDPublic"                   </span></span>
<span><span class="co">#&gt; [19] "PublicTransitivity"                 "PublicCentralityByCloseness"       </span></span>
<span><span class="co">#&gt; [21] "PublicEigenCentrality"              "PublicCentralityByEigen"           </span></span>
<span><span class="co">#&gt; [23] "PublicBetweenness"                  "PublicCentralityByBetweenness"     </span></span>
<span><span class="co">#&gt; [25] "PublicAuthorityScore"               "PublicCoreness"                    </span></span>
<span><span class="co">#&gt; [27] "PublicPageRank"</span></span></code></pre></div>
<p>Notice that by default, all variables that were present in each
sample’s original data, such as <code>"CloneFrequency"</code> and
<code>"CloneCount"</code>, are automatically carried over into this
data.</p>
<p>Many node-level network properties are present in the data. Some
pertain to the network for the individual sample from which each node
originated, i.e., the networks constructed in <a href="#step1">step
1</a>. This includes all variables that begin with
<code>"SampleLevel"</code>, such as
<code>SampleLevelNetworkDegree</code>, as well as the variable
<code>ClusterIDInSample</code>.</p>
<p>The variables that begin with <code>"Public"</code>, such as
<code>PublicNetworkDegree</code>, pertain to the global network. In
particular, <code>ClusterIDPublic</code> indicates the ID of the global
cluster to which each node belongs.</p>
</div>
<div class="section level4">
<h4 id="cluster_data">
<code>cluster_data</code><a class="anchor" aria-label="anchor" href="#cluster_data"></a>
</h4>
<p>The <code>cluster_data</code> data frame contained in the output list
contains one row for each cluster in the network, with the following
variables representing the cluster-level network properties:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># variables in the node-level metadata</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">public_clusters</span><span class="op">$</span><span class="va">cluster_data</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "cluster_id"                  "node_count"                 </span></span>
<span><span class="co">#&gt;  [3] "mean_seq_length"             "mean_degree"                </span></span>
<span><span class="co">#&gt;  [5] "max_degree"                  "seq_w_max_degree"           </span></span>
<span><span class="co">#&gt;  [7] "agg_count"                   "max_count"                  </span></span>
<span><span class="co">#&gt;  [9] "seq_w_max_count"             "diameter_length"            </span></span>
<span><span class="co">#&gt; [11] "global_transitivity"         "assortativity"              </span></span>
<span><span class="co">#&gt; [13] "edge_density"                "degree_centrality_index"    </span></span>
<span><span class="co">#&gt; [15] "closeness_centrality_index"  "eigen_centrality_index"     </span></span>
<span><span class="co">#&gt; [17] "eigen_centrality_eigenvalue"</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="plots">
<code>plots</code><a class="anchor" aria-label="anchor" href="#plots"></a>
</h4>
<p>The item <code>plots</code> is a list containing each of the plot(s)
created by <code>buildPublicClusterNetwork</code>. Each plot is an
object of class <code>ggraph</code> (which inherits from the
<code>ggplot</code> class), and is named according to the variable used
to color the nodes.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">public_clusters</span><span class="op">$</span><span class="va">plots</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "SampleID"</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">public_clusters</span><span class="op">$</span><span class="va">plots</span><span class="op">$</span><span class="va">SampleID</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "ggraph" "gg"     "ggplot"</span></span></code></pre></div>
<p>The plots can be modified after their creation as described in <a href="network_visualization.html#modifying_plots">this vignette</a>.</p>
<p>New plots can be generated from the output as described <a href="network_visualization.html#new_plots">here</a>.</p>
</div>
<div class="section level4">
<h4 id="igraph">
<code>igraph</code><a class="anchor" aria-label="anchor" href="#igraph"></a>
</h4>
<p>The item <code>igraph</code> is an object of class
<code>igraph</code>, which contains the list of nodes and edges that
comprise the network graph.</p>
<p>It can be used to <a href="network_visualization.html#new_plots">generate additional plots of
the network graph</a>.</p>
</div>
<div class="section level4">
<h4 id="adjacency_matrix">
<code>adjacency_matrix</code><a class="anchor" aria-label="anchor" href="#adjacency_matrix"></a>
</h4>
<p>The item <code>adjacency_matrix</code> is an object of class
<code>dgCMatrix</code>, which is a sparse matrix format from the
<code>Matrix</code> package.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">public_clusters</span><span class="op">$</span><span class="va">adjacency_matrix</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "dgCMatrix"</span></span>
<span><span class="co">#&gt; attr(,"package")</span></span>
<span><span class="co">#&gt; [1] "Matrix"</span></span></code></pre></div>
<p>The network adjacency matrix encodes the graph (i.e., the set of
nodes and edges) as a matrix. It is symmetric, with each row
(equivalently, each column) corresponding to a distinct network node.
The network nodes follow the same order in the rows as they do in the
columns. Each column name is the receptor sequence corresponding to the
node for that column. Each row name is the row index of the original
input data that corresponds to the node for that row.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># First 5 rows and columns of the network adjacency matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">public_clusters</span><span class="op">$</span><span class="va">adjacency_matrix</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; 5 x 5 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">#&gt;               </span></span>
<span><span class="co">#&gt; [1,] 1 1 . 1 .</span></span>
<span><span class="co">#&gt; [2,] 1 1 . 1 .</span></span>
<span><span class="co">#&gt; [3,] . . 1 . .</span></span>
<span><span class="co">#&gt; [4,] 1 1 . 1 .</span></span>
<span><span class="co">#&gt; [5,] . . . . 1</span></span></code></pre></div>
<p>Each matrix entry has a value of 0 or 1, with 1 indicating an edge
connecting the node for that row to the node for that column, and 0
indicating no edge connection. Although the diagonal entries are 1,
their values are not meaningful, since an edge is not drawn from a given
node to itself.</p>
<p>Note that the rows (nodes) in the adjacency matrix correspond to the
rows (nodes) in the <code>node_data</code> data frame. Both have the
same number of rows and the same row names.</p>
</div>
</div>
<div class="section level3">
<h3 id="labeling-the-global-clusters">Labeling the Global Clusters<a class="anchor" aria-label="anchor" href="#labeling-the-global-clusters"></a>
</h3>
<p>In the plot generated by <code>buildPublicClusterNetwork</code>,
there appear to be six clusters with more than three samples
represented.</p>
<p>In order to reference these clusters within the data, we can <a href="network_visualization.html#labeling_clusters">label the six
largest clusters in the plot with their cluster IDs</a> using the
<code>addClusterLabels</code> function. Note that within the node-level
metadata, the global cluster ID is stored in the variable
<code>ClusterIDPublic</code>, so we must provide this column name to the
<code>cluster_id_col</code> argument.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Modify plot to add labels to the clusters</span></span>
<span><span class="va">public_clusters</span><span class="op">$</span><span class="va">plots</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="../reference/addClusterLabels.html">addClusterLabels</a></span><span class="op">(</span></span>
<span>    plot <span class="op">=</span> <span class="va">public_clusters</span><span class="op">$</span><span class="va">plots</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    net <span class="op">=</span> <span class="va">public_clusters</span>,</span>
<span>    top_n_clusters <span class="op">=</span> <span class="fl">6</span>,</span>
<span>    cluster_id_col <span class="op">=</span> <span class="st">"ClusterIDPublic"</span>,</span>
<span>    size <span class="op">=</span> <span class="fl">7</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># View modified plot</span></span>
<span><span class="va">public_clusters</span><span class="op">$</span><span class="va">plots</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#&gt; Warning: Removed 511 rows containing missing values (`geom_text()`).</span></span></code></pre></div>
<p><img src="public_clusters_files/figure-html/unnamed-chunk-15-1.png" width="100%" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="focusing-on-individual-clusters-of-interest">Focusing on Individual Clusters of Interest<a class="anchor" aria-label="anchor" href="#focusing-on-individual-clusters-of-interest"></a>
</h3>
<p>If we wish to focus on a particular cluster of interest within the
global network, we can build a network exclusively using the clones from
that cluster. This is accomplished using the
<code>buildRepSeqNetwork</code> function, where we subset our data
according to the value of the <code>ClusterIDPublic</code> column, which
contains the global cluster IDs.</p>
<p>Below, we focus on the first cluster, which in this case is also the
largest cluster by node count. In the plot, we color each node according
to the receptor sequence of its corresponding clone.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># focus on cluster 1</span></span>
<span><span class="fu"><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> </span>
<span>    <span class="va">public_clusters</span><span class="op">$</span><span class="va">node_data</span><span class="op">[</span><span class="va">public_clusters</span><span class="op">$</span><span class="va">node_data</span><span class="op">$</span><span class="va">ClusterIDPublic</span> <span class="op">==</span> <span class="fl">1</span>, <span class="op">]</span>,</span>
<span>  seq_col <span class="op">=</span> <span class="st">"CloneSeq"</span>, </span>
<span>  color_nodes_by <span class="op">=</span> <span class="st">"CloneSeq"</span>, color_scheme <span class="op">=</span> <span class="st">"plasma"</span>, </span>
<span>  size_nodes_by <span class="op">=</span> <span class="fl">3</span>, </span>
<span>  output_dir <span class="op">=</span> <span class="cn">NULL</span>, output_name <span class="op">=</span> <span class="st">"Cluster 1"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Input data contains 96 rows.</span></span>
<span><span class="co">#&gt; Removing sequences with length fewer than 3 characters... Done. 96 rows remaining.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 96 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Generating graph plot with nodes colored by CloneSeq...</span></span></code></pre></div>
<p><img src="public_clusters_files/figure-html/unnamed-chunk-16-1.png" width="100%" style="display: block; margin: auto;"></p>
<pre><code><span><span class="co">#&gt;  Done.</span></span></code></pre>
<p>Below, we do the same for the sixth largest cluster:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># focus on cluster 6</span></span>
<span><span class="fu"><a href="../reference/buildRepSeqNetwork.html">buildRepSeqNetwork</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> </span>
<span>    <span class="va">public_clusters</span><span class="op">$</span><span class="va">node_data</span><span class="op">[</span><span class="va">public_clusters</span><span class="op">$</span><span class="va">node_data</span><span class="op">$</span><span class="va">ClusterIDPublic</span> <span class="op">==</span> <span class="fl">6</span>, <span class="op">]</span>,</span>
<span>  seq_col <span class="op">=</span> <span class="st">"CloneSeq"</span>, </span>
<span>  color_nodes_by <span class="op">=</span> <span class="st">"CloneSeq"</span>, color_scheme <span class="op">=</span> <span class="st">"plasma"</span>, </span>
<span>  size_nodes_by <span class="op">=</span> <span class="fl">3</span>, </span>
<span>  output_dir <span class="op">=</span> <span class="cn">NULL</span>, output_name <span class="op">=</span> <span class="st">"Cluster 6"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Input data contains 27 rows.</span></span>
<span><span class="co">#&gt; Removing sequences with length fewer than 3 characters... Done. 27 rows remaining.</span></span>
<span><span class="co">#&gt; Computing network edges based on a max hamming distance of 1... Done.</span></span>
<span><span class="co">#&gt; Network contains 27 nodes (after removing isolated nodes).</span></span>
<span><span class="co">#&gt; Generating graph plot with nodes colored by CloneSeq...</span></span></code></pre></div>
<p><img src="public_clusters_files/figure-html/unnamed-chunk-17-1.png" width="100%" style="display: block; margin: auto;"></p>
<pre><code><span><span class="co">#&gt;  Done.</span></span></code></pre>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Brian Neal, Hai Yang, Le Phi Long, Li Zhang.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
